#!/usr/bin/env snakemake -s
##
## Izaskun Mallona
## GPLv3

"""
Repeats in scRNA-seq main Snakefile
"""

import os.path as op
from glob import glob
import re

configfile: "config.yaml"

include: "utils.snmk"
include: "simulations.snmk"
include: "starsolo.snmk"
#include: 'counting.snmk'
include: "data_acquisition.snmk"

print(config)

for item in ['annotations', 'references', 'simulations', 'tmp', 'logs', 'benchmarks']:
    try:
        if not op.exists(op.dirname(op.join(config['base'], item))):
            os.makedirs(op.join(config['base'], item))
    except OSError as err:
        print(err)

## also downloading some data, even if related to the non-simulations usecase
rule all:
    input:
        op.join(config['base'], 'simulations', 'cdna.fastq.gz'),
        expand(op.join(config['base'], 'data', 'colon_cancer_lines', '{srr}_1.fastq.gz'),
               srr = ['SRR10974769', 'SRR10974767']),
        expand(op.join(config['base'], 'starsolo', '{sample}', '{multimappers}',
                       'Aligned.sortedByCoord.out.bam'),
               sample = config['sample'],
               multimappers = ['multimappers', 'unique_reads'])
        # expand(op.join(config['base'], 'starsolo', '{sample}', '{multimappers}',
        #                'counts.tsv.gz'),
        #        sample = config['sample'],
        #        multimappers = ['multimappers', 'unique_reads'])

