---
title: "Within sample report: bulk monocytes"
author: "Izaskun Mallona at Robinsonlab"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    code_download: true
    number_sections: true
    df_print: kable
    theme: lumen
    self_contained: yes
params:
  identifier: ""   
  repeats_bowtie_unique: ""
  repeats_bowtie_multi: ""
  repeats_star_unique: ""
  repeats_star_multi: ""
  repeats_only_unique: ""
  repeats_only_multi: ""
  genes_star: ""
  de_output: ""
  config_fn: ""
---

# Config


## Parameters

Rendered for sample:

```{r}
print(params)
```

## Libraries

```{r}
suppressPackageStartupMessages({
library(Matrix)

library(knitr)
## library(Seurat)
library(dplyr)
## library(scater)
library(data.table)
## library(SingleCellExperiment)
## library(scran)
## library(org.Mm.eg.db)
## suppressPackageStartupMessages(library(dplyr))
## suppressPackageStartupMessages(library(DT))
## library(pheatmap)
## library(Cairo)
## library(RColorBrewer)
## library(corrplot)
library(ggplot2)
## library(xtable)
library(mclust)
library(pheatmap)
## library(SCINA)
library(DT)
library(biomaRt)
library(edgeR)
})
```


```{r}
opts_chunk$set(fig.width = 5,
               fig.height = 5,
               dev = 'png',
               cache = FALSE,
               include = TRUE,
               cache.lazy = FALSE,
               warning = TRUE,
               echo = TRUE,
               message = FALSE,
               knitr.table.format = "html")


options(bitmapType='cairo')
getOption('bitmapType')
```


## Functions

```{r functions}
## add alpha
ac <- function(col, alpha=1){
  apply(sapply(col, col2rgb)/255, 2, 
                     function(x) 
                       rgb(x[1], x[2], x[3], alpha=alpha))  
}

## coef_var <- function(x){
##         (sd(x)/mean(x))*100
## }

```


```{r}

read_featurecounts <- function(fn, identifier, add_gene_symbols = FALSE, remove_dirname = TRUE) {
    ## Featurecounts-based repeats
    counts <- data.table::fread(fn, sep = '\t', header = TRUE,
                             drop = c('Chr', 'Start', 'End', 'Strand', 'Length'))
    rownames(counts) <- counts$Geneid ## here

    colnames(counts) <- gsub(".*:([A-Z]+)-1.bam","\\1", colnames(counts))
    ## counts <- counts[,7:ncol(counts)]

    counts <- as.data.frame(counts)
    rownames(counts) <- counts$Geneid
    counts <- counts[,-1]

    if (add_gene_symbols) {
        if (!file.exists('feature_info_human.RData')) {
            
            annot <- list()

            annot$ids <- rownames(counts)

            annot$filters = "ensembl_gene_id_version"
            annot$attributes = c(annot$filters, "external_gene_name", "chromosome_name", 
                               "gene_biotype", "start_position", "end_position")
            annot$biomart = "ENSEMBL_MART_ENSEMBL" 
            annot$dataset = "hsapiens_gene_ensembl"
            annot$host = "asia.ensembl.org"
            annot$bmart <- biomaRt::useMart(biomart = annot$biomart, dataset = annot$dataset, 
                                          host = annot$host)

            feature_info <- biomaRt::getBM(attributes = annot$attributes,
                                           filters = annot$filters, 
                                           values = annot$ids, mart = annot$bmart)

            ## dim(feature_info)
            save(feature_info, file = 'feature_info_human.RData')
        } else {
            load('feature_info_human.RData')
        }

        tmp <- merge(data.frame(ensembl_gene_id_version = rownames(counts)),
                       by = 'ensembl_gene_id_version', feature_info, all.x = TRUE)

        rownames(tmp) <- tmp$ensembl_gene_id_version
        tmp <- tmp[rownames(counts),]

        stopifnot(rownames(counts) == tmp$ensembl_gene_id_version)

        rownames(counts) <- sprintf('%s_%s', tmp$ensembl_gene_id_version, tmp$external_gene_name)
    }

    if (remove_dirname) {
        colnames(counts) <- basename(colnames(counts))
        colnames(counts) <- gsub('.bam', '', colnames(counts))
        colnames(counts) <- gsub('_Aligned.sortedByCoord.out', '', colnames(counts))
    }
    return(counts)
}
```


# Data load

```{r, eval = FALSE}
params <- list(identifier = "bulk_GSE51984", 
          genes_star = "/home/imallona/repeats_sc/runs/bulk_GSE51984/star_transcriptome/genes/bulk_GSE51984_star_genes.counts.gz", 
          repeats_star_unique = "/home/imallona/repeats_sc/runs/bulk_GSE51984/star_transcriptome/repeats/unique_reads/bulk_GSE51984_star_transcriptome_repeats.counts.gz", 
          repeats_star_multi = "/home/imallona/repeats_sc/runs/bulk_GSE51984/star_transcriptome/repeats/multimappers/bulk_GSE51984_star_transcriptome_repeats.counts.gz", 
          repeats_bowtie_unique = "/home/imallona/repeats_sc/runs/bulk_GSE51984/bowtie_repeatome/unique_reads/bulk_GSE51984_bowtie_repeats.counts.gz", 
          repeats_bowtie_multi = "/home/imallona/repeats_sc/runs/bulk_GSE51984/bowtie_repeatome/multimappers/bulk_GSE51984_bowtie_repeats.counts.gz", 
          repeats_only_unique = "/home/imallona/repeats_sc/runs/bulk_GSE51984/star_transcriptome/repeats_no_overlap/unique_reads/bulk_GSE51984_repeats_no_overlap_genes.counts.gz",
          repeats_only_multi = "/home/imallona/repeats_sc/runs/bulk_GSE51984/star_transcriptome/repeats_no_overlap/multimappers/bulk_GSE51984_repeats_no_overlap_genes.counts.gz",
          seurat_output = "/home/imallona/repeats_sc/runs/bulk_GSE51984/bulk_GSE51984_pmbc_smartseq2_regress_nCount_FALSE_nFeature_TRUE.rds", 
          aris_output = "/home/imallona/repeats_sc/runs/bulk_GSE51984/bulk_GSE51984_aris_regress_nCount_FALSE_nFeature_TRUE.rds", 
          cobra_output = "/home/imallona/repeats_sc/runs/bulk_GSE51984/bulk_GSE51984_cobra_regress_nCount_FALSE_nFeature_TRUE.rds",
          config_fn = '/home/imallona/repeats_sc/data/bulk_GSE51984/bulk_GSE51984.conf',
          regress_genes_nCount = "FALSE",
          regress_genes_nFeature = "TRUE")


```

Annotation for monocytes

```{r}
annot <- read.table(params$config_fn,
                     header = FALSE, sep = ',',
                     stringsAsFactors = FALSE)

colnames(annot)[1] <- c('srr')
colnames(annot)[8] <- 'sample'

rownames(annot) <- annot$srr
```

```{r}
table(annot$sample)
annot$sample <- gsub('-', '_', annot$sample)
annot$sample <- gsub(' ', '_', annot$sample)
annot$sample <- gsub('+', '', annot$sample, fixed = TRUE)
table(annot$sample)

annot$sample_reduced <- ifelse(annot$sample == 'Monocytes', 'monocytes', 'non_monocytes')

```


```{r}
d <- list()


for (item in setdiff(names(params), c('identifier', 'de_output'))){

    d[[item]] <- read_featurecounts(
    params[[item]],
    identifier = sprintf('%s_%s', params$identifier, params[[item]]),
    add_gene_symbols = grepl('genes', item))

    d[[item]] <- DGEList(d[[item]], group = annot$sample_reduced)
}

str(d)

```


# Standard RNA-seq processing

Normalization factors

```{r edgeR-dge-generate}
d <- lapply(d,  function(x) edgeR::calcNormFactors(x))

```

## logCPMs

```{r logcpm}
logcpms <- lapply(d, function(x) edgeR::cpm(x, log = TRUE))
```

# design

```{r}


des <- model.matrix(as.formula('~ 0 + sample_reduced'), data = annot)
print(des)
```

Mind the re-calculation of normalization factors

```{r}
for (item in names(d)) {
    keep <- edgeR::filterByExpr(d[[item]], design = des)
    d[[item]] <- d[[item]][keep, ]
    d[[item]] <- calcNormFactors(d[[item]])
}
```

# Estimate dispersion and fit QL model


```{r }

d <- lapply(d, function(x) estimateDisp(x, design = des))

fit <- list()
fit <- lapply(d, function(x) glmQLFit(x, design = des))

lapply(d, function(x) plotBCV(x))
```

# Define contrasts 

Monocytes vs all

```{r edgeR-define-contrasts}
(conts <- as.data.frame(
   makeContrasts(
       contrasts = 'sample_reducedmonocytes-sample_reducednon_monocytes',
       levels = des)))


```

# DGE

```{r dge}
res <- list()
for (item in names(d)) {
    res[[item]] <- lapply(conts, function(x) {
        qlf <- glmQLFTest(fit[[item]], contrast = x)
        tt <- topTags(qlf, n = Inf, sort.by = "PValue")$table
    })
}
```

```{r}
lapply(res, function(x) head(x[[1]], 10)) 
```


# Trace

## Save outputs


```{r}
saveRDS(object = res, file = params$de_output)
```

## Session info

```{r}
devtools::session_info()

date()
```

