---
title: "PBMCs preliminary"
author: "Izaskun Mallona"
output:
  html_document:
    keep_md: false
    toc: true
    toc_float: true
    toc_depth: 4
params:
  seed: 11
---


@todo regress out mitochondrial contribution

similar to https://satijalab.org/seurat/v3.0/pbmc3k_tutorial.html

```{r}
## stop('todo: get top quantiles for chipseq signals and run pagoda on these')

```

# Aim

Exploratory analysis on Hubrecht plates Tina 

Tip for transparency: `.Rprofile` with

```{r}
setHook(packageEvent("grDevices", "onLoad"),
function(...) grDevices::X11.options(type='cairo'))
options(device='x11')

```
# Preprocessing

```{r settings, include = TRUE}

library(knitr)
library(Seurat)
library(dplyr)
library(scater)
library(dplyr)
## library(data.table)
## library(SingleCellExperiment)
## library(scran)
## library(org.Mm.eg.db)
## suppressPackageStartupMessages(library(dplyr))
## suppressPackageStartupMessages(library(DT))
## library(pheatmap)
## library(Cairo)
## library(RColorBrewer)
## library(corrplot)
library(ggplot2)
library(reshape2)
library(xtable)

opts_chunk$set(fig.width = 8,
               fig.height = 8,
               cache = TRUE,
               include = TRUE,
               cache.lazy = FALSE,
               warning = TRUE,
               message = TRUE)


options(bitmapType='cairo')
getOption('bitmapType')
```

```{r}
COUNTS_PATH='/home/Shared_s3it/imallona/repeats_sc/data2/featurecounts'
```

```{r functions}
## add alpha
ac <- function(col, alpha=1){
  apply(sapply(col, col2rgb)/255, 2, 
                     function(x) 
                       rgb(x[1], x[2], x[3], alpha=alpha))  
}

```

Data load. We'll focus on TranscriptCounts, but read them all just in case we needed to compare summarizing ratios.

# Expression

```{r}


# this is wget http://cf.10xgenomics.com/samples/cell-exp/3.0.2/5k_pbmc_v3/5k_pbmc_v3_filtered_feature_bc_matrix.tar.gz
pbmc.data <- Read10X(data.dir = "../data/filtered_feature_bc_matrix/")
# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc5k", min.cells = 3,
                           min.features = 200)

rm(pbmc.data)
pbmc
```


```{r}
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot2
```

```{r}
pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500)
```

```{r}
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000)
```

```{r}
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(pbmc), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(pbmc)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
CombinePlots(plots = list(plot1, plot2))
```

```{r}
pbmc <- ScaleData(pbmc)
```

```{r}
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
```

```{r}
DimPlot(pbmc, reduction = "pca")
```

```{r}
ElbowPlot(pbmc)
```

```{r}
pbmc <- FindNeighbors(pbmc, dims = 1:16)
pbmc <- FindClusters(pbmc, resolution = 0.15)
```

```{r}
pbmc <- RunUMAP(pbmc, dims = 1:16)
                
DimPlot(pbmc, reduction = "umap")
```


```{r}
pbmc <- RunTSNE(pbmc, dims = 1:16)
                
DimPlot(pbmc, reduction = "tsne")
```

```{r}
# find markers for every cluster compared to all remaining cells, report only the positive ones
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
pbmc.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC)
```

```{r}
top10 <- pbmc.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
DoHeatmap(pbmc, features = top10$gene) + NoLegend()
```


```
	IL7R, CCR7	Naive CD4+ T
	IL7R, S100A4	Memory CD4+
	CD14, LYZ	CD14+ Mono
	MS4A1	B
	CD8A	CD8+ T
	FCGR3A, MS4A7	FCGR3A+ Mono
	GNLY, NKG7	NK
	FCER1A, CST3	DC
	PPBP	Platelet
    ```
    
```
0	IL7R, CCR7	Naive CD4+ T
1 IL7R, S100A4	Memory CD4+
2 MS4A1	B
3 CD8?
4 GNLY, NKG7	NK
5	CD14, LYZ	CD14+ Mono
6 FCER1A, CST3	DC
7 Platelet
```


```{r}
DoHeatmap(pbmc, features = c('IL7R', 'CCR7', 'S100A4', 'CD14', 'LYZ',
                             'MS4A1', 'CD8A', 'FCGR3A', 'MS4A7',
                             'GNLY', 'NKG7',
                             'FCER1A', 'CST3',
                             'PPBP')) + NoLegend()
```

Adding new labels

```{r, eval = TRUE, fig.width =4, fig.height = 4}
new.cluster.ids <- c("Naive CD4 T", "Memory CD4 T",
                     "B",  'NK', 'CD14 Mono',
                     'DC', 'Platelet')

names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = "tsne", label = TRUE, pt.size = 1.5) + NoLegend()

```


# Repeats

@todo fix this is untested, indicate the path

```{r}

if (!file.exists(file.path('data', 'repeats_counts_01.RData'))) {
## setwd(COUNTS_PATH)
    fns <- list.files(path = COUNTS_PATH, pattern = '*.counts$', recursive = TRUE)

    d <- list()
    for (fn in fns) {
        print(fn)
        ## d[[gsub('_counts', '', basename(fn))]] <- read.table(
        ##     file.path(COUNTS_PATH, fn),
        ##     header = TRUE,
        ##     sep = "\t",
        ##     colClasses = c(rep("NULL", 6), "integer"),
        ##     skip = 1)

        d[[gsub('_counts', '', basename(fn))]] <- read.table(
            pipe(sprintf('cut -f7 %s', file.path(COUNTS_PATH, fn))),            
            header = TRUE,
            sep = "\t",
            skip = 1)
        
    }



    str(d)

    ## last item, to get the annotation
    ## annot <-  read.table(file.path(COUNTS_PATH, fn),
    ##                      header = TRUE,
    ##                      sep = "\t",
    ##                      skip = 1)[,1]

    annot <-read.table(
            pipe(sprintf('cut -f1 %s', file.path(COUNTS_PATH, fn))),            
            header = TRUE,
            sep = "\t",
            skip = 1)


    d_bak <- d
    ## d <- cbind(annot, d)
    d <- do.call(cbind.data.frame, d)
    rownames(d) <- annot[,1]
    str(d)
    colnames(d) <- gsub('.1.bam', '', colnames(d))
    colnames(d) <- gsub('..foo_', '', colnames(d))
    
} else {
    load(file.path('data', 'repeats_counts_01.RData'))
}
```

```{r save_repeats_object}
if (!file.exists(file.path('data', 'repeats_counts_01.RData'))){
    dir.create('data')
    save(d, file = file.path('data', 'repeats_counts_01.RData'))
}
    
```

## Repeats basic stats


Highlighting the repeats expression of the RNAseq clusters? Maybe adding the features as metadata?

```{r}
print('todo')
```

Overall expression rates of some repeat families

How many repeat families are detected? Which percentage is it?
```{r}

detected <- sum(apply(d, 1, function(x) sum(x) > 1))
detected/nrow(d) * 100

nrow(d)

```

<!-- 33.20062 -->


Getting the top most variable repeat instances, which is their variation range?

```{r}
vars <- apply(d, 1, var)
names(vars) <- rownames(d)

selected <- names(vars[vars > quantile(vars, 0.999)])

apply(d[selected,], 1, mean)
apply(d[selected,], 1, var)


```


## Repeats Seurat




```{r}
reps <- CreateSeuratObject(counts = d,
                           project = "pbmc5k",
                           min.cells = 3,
                           min.features = 50)


```


```{r}
plot2 <- FeatureScatter(reps, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot2
```

```{r}
reps <- subset(reps, subset = nFeature_RNA > 200 & nFeature_RNA < 2500)
```

```{r}
reps <- NormalizeData(reps, normalization.method = "LogNormalize", scale.factor = 10000)
```


```{r}
reps <- FindVariableFeatures(reps, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(reps), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(reps)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
CombinePlots(plots = list(plot1, plot2))
```

```{r}
reps <- ScaleData(reps)
```

```{r}
reps <- RunPCA(reps, features = VariableFeatures(object = reps))
```

```{r}
DimPlot(reps, reduction = "pca")
```

```{r}
ElbowPlot(reps)
```

```{r, cache = FALSE}
reps <- FindNeighbors(reps, dims = 1:10)
reps <- FindClusters(reps, resolution = 0.1)

```

```{r}
reps <- RunUMAP(reps, dims = 1:10)
                
DimPlot(reps, reduction = "umap")
```


```{r}
reps <- RunTSNE(reps, dims = 1:10)
                
DimPlot(reps, reduction = "tsne")
```

```{r}
# find markers for every cluster compared to all remaining cells, report only the positive ones
reps.markers <- FindAllMarkers(reps, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
reps.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC)
```

```{r}
top10 <- reps.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
DoHeatmap(reps, features = top10$gene) + NoLegend()
```

## Integrative (according to markers)


Repeats in accordance to the gene-coding rnaseq clustering

These we cannot explore much due to the abundance of NAs


```{r}
reps_markers_char <- as.data.frame(reps.markers %>% group_by(cluster) %>%
                                   top_n(n = 2, wt = avg_logFC))$gene
```

```{r}
## from_rna_meta <- na.omit(pbmc@meta.data[colnames(reps),])
## nrow(na.omit(from_rna))

tmp <- as.data.frame(log10(t(d) +1))
## tmp$cell <- rownames(tmp)

tmp <- merge(pbmc@meta.data, tmp, by.x='row.names', by.y = 'row.names', all.x = TRUE)
## colnames(tmp)[1:5]  <- sprintf('genes_%s', colnames(tmp)[1:5])
pbmc@meta.data <- tmp
rownames(pbmc@meta.data) <- tmp$Row.names

rm(tmp)

```

Getting the top variable repeats
```{r, fig.height = 6, fig.width = 6}
## summary(pbmc@meta.data[selected,])

for (feature in reps_markers_char){
    p <- FeaturePlot(pbmc, features = feature, reduction = 'tsne',
                     shape.by = 'RNA_snn_res.0.15',
                     pt.size = 2)
    print(p + ggtitle(sprintf('log10(x + 1) %s counts', feature)))
}
```

Log10 raw repeat counts by RNA (genes) clustering

```{r}

target <- pbmc@meta.data[,c('RNA_snn_res.0.15', reps_markers_char)]

foo <- melt(target, id.vars='RNA_snn_res.0.15')
colnames(foo) <- c('genes_clust', 'repfamily', 'log10_counts')
ggplot(foo,
       aes(x=genes_clust, y=log10_counts)) + 
    geom_boxplot(outlier.size=1) +
    facet_wrap(~ repfamily, ncol = 3) +
    theme(axis.text.x = element_text(angle = 90))


ggplot(foo,
       aes(x= repfamily, y=log10_counts)) + 
    geom_boxplot(outlier.size=1) +
    facet_wrap(~genes_clust, ncol = 3) +
    theme(axis.text.x = element_text(angle = 90))

```

Quality checks i.e. num genes/library size vs number repeats

```{r, eval = FALSE}
for (feature in reps_markers_char) {
    par(mfrow = c(1,2), pty = 's')
    plot(y = pbmc@meta.data[,feature],
         x = pbmc@meta.data$nFeature_RNA,
         main = feature,
         pch =19,
         xlab = 'number of features (genes)',
         ylab = 'number of repeats [log10(x+1)]',
         col = ac('black', 0.5))
    
     plot(y = pbmc@meta.data[,feature],
          x = pbmc@meta.data$nCount_RNA,
          main = feature,
          xlab = 'number of reads (genes)',
         ylab = 'number of repeats [log10(x+1)]',
         pch =19,
         col = ac('black', 0.5))
}
```

Quality checks i.e. num genes/library size vs number repeats

```{r}
for (feature in reps_markers_char) {
    par(mfrow = c(1,3), pty = 's')
    plot(y = pbmc@meta.data[,feature],
         x = pbmc@meta.data$nFeature_RNA,
         main = feature,
         pch =19,
         xlab = 'number of features (genes)',
         ylab = 'number of repeats [log10(x+1)]',
         col = ac(as.numeric(pbmc@meta.data$RNA_snn_res.0.15), 0.5))
    
     plot(y = pbmc@meta.data[,feature],
          x = pbmc@meta.data$nCount_RNA,
          main = feature,
          xlab = 'number of reads (genes)',
         ylab = 'number of repeats [log10(x+1)]',
         pch =19,
         col = ac(as.numeric(pbmc@meta.data$RNA_snn_res.0.15), 0.5))

    
    plot.new()
    legend('topleft',
           legend= levels(pbmc@meta.data$RNA_snn_res.0.15),
           pch = 19,
           col = 1: max(as.numeric(pbmc@meta.data$RNA_snn_res.0.15)),
           title = 'genes-based clustering (SNN)')
}
```

## Integrative (according to variance)



```{r}
vars <- apply(d, 1, var)
names(vars) <- rownames(d)

selected <- names(vars[vars > quantile(vars, 0.999)])

apply(d[selected,], 1, mean)
apply(d[selected,], 1, var)


```


```{r}

target <- pbmc@meta.data[,c('RNA_snn_res.0.15', selected)]

foo <- melt(target, id.vars='RNA_snn_res.0.15')
colnames(foo) <- c('genes_clust', 'repfamily', 'log10_counts')

ggplot(foo,
       aes(x=genes_clust, y=log10_counts)) + 
    geom_boxplot(outlier.size=1) +
    facet_wrap(~ repfamily, ncol = 3) +
    theme(axis.text.x = element_text(angle = 90))


ggplot(foo,
       aes(x= repfamily, y=log10_counts)) + 
    geom_boxplot(outlier.size=1) +
    facet_wrap(~genes_clust, ncol = 3) +
    theme(axis.text.x = element_text(angle = 90))
```

```{r, eval = FALSE}
for (feature in selected) {
    par(mfrow = c(1,2), pty = 's')
    plot(y = pbmc@meta.data[,feature],
         x = pbmc@meta.data$nFeature_RNA,
         main = feature,
         pch =19,
         xlab = 'number of features (genes)',
         ylab = 'number of repeats [log10(x+1)]',
         col = ac('black', 0.5))
    
     plot(y = pbmc@meta.data[,feature],
          x = pbmc@meta.data$nCount_RNA,
          main = feature,
          xlab = 'number of reads (genes)',
         ylab = 'number of repeats [log10(x+1)]',
         pch =19,
         col = ac('black', 0.5))
}
```

Quality checks i.e. num genes/library size vs number repeats

```{r}
for (feature in selected) {
    par(mfrow = c(1,3), pty = 's')
    plot(y = pbmc@meta.data[,feature],
         x = pbmc@meta.data$nFeature_RNA,
         main = feature,
         pch =19,
         xlab = 'number of features (genes)',
         ylab = 'number of repeats [log10(x+1)]',
         col = ac(as.numeric(pbmc@meta.data$RNA_snn_res.0.15), 0.5))
    
     plot(y = pbmc@meta.data[,feature],
          x = pbmc@meta.data$nCount_RNA,
          main = feature,
          xlab = 'number of reads (genes)',
         ylab = 'number of repeats [log10(x+1)]',
         pch =19,
         col = ac(as.numeric(pbmc@meta.data$RNA_snn_res.0.15), 0.5))

    
    plot.new()
    legend('topleft',
           legend= levels(pbmc@meta.data$RNA_snn_res.0.15),
           pch = 19,
           col = 1: max(as.numeric(pbmc@meta.data$RNA_snn_res.0.15)),
           title = 'genes-based clustering (SNN)')
}
```

# Overlap gene clusters and repeat clusters

```{r}
tryCatch({
    foo <- data.frame(genes = pbmc@meta.data[,'RNA_snn_res.0.15'],
                      cell = row.names(pbmc@meta.data))

    bar <- data.frame(repeats = reps@meta.data[,'RNA_snn_res.0.15'],
                      cell = row.names(reps@meta.data))

    tmp <- merge(foo, bar, by = 'cell')

    table(tmp$genes, tmp$repeats)
}, error = function(x) print(x))


```

# Repeats and gene markers


```{r}

vars <- apply(d, 1, var)
names(vars) <- rownames(d)

selected <- names(vars[vars > quantile(vars, 0.999)])

apply(d[selected,], 1, mean)

avgs <- data.frame('repeat' = as.character(selected),
                   mean = as.character(round(apply(d[selected,], 1, mean),2)),
                   sd = as.character(round(apply(d[selected,], 1, sd),2)))


print.xtable(xtable(avgs, include.rownames = FALSE))
## print(avgs)
```


AluSq and (GCCCC)n

```
0	IL7R, CCR7	Naive CD4+ T
1	IL7R, S100A4	Memory CD4+ + CD8+
2   IL7R, CCR7, S100A4 T cells?  	
3	GNLY, NKG7	NK
4	MS4A1	B
5   FCER1A, CST3	DC
6   MS4A1  B
7   CD8 CD8?
8	PPBP	Platelet
```

```{r wrong1, eval =FALSE}

pbmc@meta.data$literate <- 
for (feature in c('AluY', '(GGCCAC)n')) {
    par(mfrow = c(2,2), pty = 's')
    plot(y = pbmc@meta.data[,feature],
         x = pbmc@meta.data$nFeature_RNA,
         main = feature,
         pch =19,
         xlab = 'number of features (genes)',
         ylab = 'number of repeats [log10(x+1)]',
         col = ac(as.numeric(pbmc@meta.data$RNA_snn_res.0.1), 0.5))
    
    plot(y = pbmc@meta.data[,feature],
         x = pbmc@meta.data$nCount_RNA,
         main = feature,
         xlab = 'number of reads (genes)',
         ylab = 'number of repeats [log10(x+1)]',
         pch =19,
         col = ac(as.numeric(pbmc@meta.data$RNA_snn_res.0.1), 0.5))

    foo <- na.omit(data.frame(repfam = pbmc@meta.data[,feature],
                              gene = FetchData(pbmc, 'IL7R',
                                               cells = rownames(pbmc@meta.data[,feature]),
                                               slot = "data")))
    
    plot(foo,
         main = feature,
         xlab = 'PPBP (normalized expression)',
         ylab = 'number of repeats [log10(x+1)]',
         pch =19,
         col = ac(as.numeric(pbmc@meta.data$RNA_snn_res.0.1), 0.5))
    
   foo <- na.omit(data.frame(repfam = pbmc@meta.data[,feature],
                              gene = FetchData(pbmc, 'CCR7',
                                               cells = rownames(pbmc@meta.data[,feature]),
                                               slot = "data")))

    plot(foo,
         main = feature,
         xlab = 'FCER1A (normalized expression)',
         ylab = 'number of repeats [log10(x+1)]',
         pch =19,
         col = ac(as.numeric(pbmc@meta.data$RNA_snn_res.0.1), 0.5))
    
    ## plot.new()
    ## legend('topleft',
    ##        legend= levels(pbmc@meta.data$RNA_snn_res.0.1),
    ##        pch = 19,
    ##        col = 1: max(as.numeric(pbmc@meta.data$RNA_snn_res.0.1)),
    ##        title = 'genes-based clustering (SNN)')
}


```

```{r, fig.width=6, fig.height=4}

new.cluster.ids <- c("Naive CD4 T", "Memory CD4 T",
                     "B",  'NK', 'CD14 Mono',
                     'DC', 'Platelet')

names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)

selected <- c('AluY', 'AluSx', 'MER49', '(GGCCAC)n', 'MIR', 'L2a', 'L2b', 'L1PA11')
target <- pbmc@meta.data[, selected]

target$literate_ids <- Idents(pbmc, cells = rownames(target))

rep_based <- melt(target, id.vars='literate_ids')
colnames(rep_based) <- c('genes_clust', 'repfamily', 'log10_counts')

ggplot(rep_based,
       aes(x=genes_clust, y=log10_counts)) + 
    geom_boxplot(outlier.size=1) +
    facet_wrap(~ repfamily, ncol = 4) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    ylab('repeat expression [log10(x + 1) counts]') +
    xlab('gene-based clustering')



## cell_type_features <- c('IL7R', 'CCR7', 'S100A4', 'GNLY', 'PPBP', 'MS4A1', 'CD8A')


## target <- FetchData(pbmc, cell_type_features, slot = 'data')

## target$literate_ids <- Idents(pbmc, cells = rownames(target))

## gene_based <- melt(target, id.vars='literate_ids')
## colnames(gene_based) <- c('genes_clust', 'repfamily', 'log10_counts')

## ggplot(gene_based,
##        aes(x=genes_clust, y=log10_counts)) + 
##     geom_boxplot(outlier.size=1) +
##     facet_wrap(~ repfamily, ncol = 5) +
##     theme(axis.text.x = element_text(angle = 90))

```




# Timestamp

```{r sessionInfo, cache = FALSE}
date()
sessionInfo()
devtools::session_info()

```
