---
title: "Across samples report"
author: "Izaskun Mallona at Robinsonlab"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    code_download: true
    number_sections: true
    df_print: kable
    theme: lumen
    always_allow_html: yes
params:
  chromium1: ""
  chromium2: ""
  chromium3: ""
  chromium4: ""
  smartseq1: ""
  celseq2: ""
  dropseq: ""
  seqwell: ""
  bulk_gse80095: ""
  bulk_gse94820: ""
  zheng_truth: ""
---



Rendered for samples:

```{r}
## print(params)
```
# Config


```{r}
suppressPackageStartupMessages({
library(Matrix)

library(knitr)
library(Seurat)
library(dplyr)
## library(scater)
library(data.table)
## library(SingleCellExperiment)
## library(scran)
## library(org.Mm.eg.db)
## suppressPackageStartupMessages(library(dplyr))
## suppressPackageStartupMessages(library(DT))
library(pheatmap)
## library(Cairo)
## library(RColorBrewer)
## library(corrplot)
library(ggplot2)
library(reshape2)
## library(xtable)
## library(mclust)
library(viridis)
library(UpSetR)
library(ggplot2)
library(grid)
library(gridExtra)
library(scales)
library(tidyr)
library(Wind) # https://github.com/haowulab/Wind
library(gridExtra)
library(scales)
library(edgeR)
## devtools::install_github("fionarhuang/TreeHeatmap")
library(ggtree)
library(TreeHeatmap)
library(ggnewscale)
})

```

```{r}
BASE <- file.path('/home', 'imallona', 'repeats_sc')
```

```{r}
opts_chunk$set(fig.width = 5,
               fig.height = 5,
               dev = 'png',
               cache = TRUE,
               echo = TRUE,
               include = TRUE,
               cache.lazy = FALSE,
               warning = TRUE,
               error = TRUE,
               message = FALSE)


options(bitmapType='cairo')
## getOption('bitmapType')
```

```{r}
theme_set(theme_classic()) 

## theme_update(panel.background = element_blank(),
##              ## panel.border = element_blank(),
##              panel.grid.major = element_blank(),
##              panel.grid.minor = element_blank(),
##              axis.line = element_line(colour = "black"))

theme_update(panel.spacing = unit(.05, "lines"),
             panel.border = element_rect(color = "gray20", fill = NA, size = 0.1), 
             strip.background = element_rect(color = "gray20", size = 0.1),
             text = element_text(size = 15))

```


```{r functions}
## add alpha
ac <- function(col, alpha=1){
  apply(sapply(col, col2rgb)/255, 2, 
                     function(x) 
                       rgb(x[1], x[2], x[3], alpha=alpha))  
}

load_standard_cellranger_genes <- function(run_name) {
    ## barcodes_fn <- params$genes_cellranger ## should end with `barcodes.tsv.gz`
    barcodes_fn <- file.path(BASE, 'runs', run_name, 'cellranger_standard', 'outs',
                             'filtered_feature_bc_matrix', 'barcodes.tsv.gz')
    
    path <- dirname(barcodes_fn)

    genes <- CreateSeuratObject(counts = Read10X(data.dir = path),
                                project = 'cellranger_standard',
                                min.cells = 0,
                                min.features = 0)
    return(genes)
}

load_cellranger_repeats <-  function(run_name) {
    barcodes_fn <- file.path(BASE, 'runs', run_name, 'cellranger_repeats', 'outs',
                             'filtered_feature_bc_matrix', 'barcodes.tsv.gz')

    path <- dirname(barcodes_fn)

    repeats <- CreateSeuratObject(counts = Read10X(data.dir = path),
                                project = 'cellranger_repeats',
                                min.cells = 0,
                                min.features = 0)
    return(repeats)
}

load_featurecounts_repeats <-  function(run_name) {

    repeats <- file.path(BASE, "runs", run_name, 'count_repeats_on_cellranger_standard',
                                paste0(run_name, "_repeats.counts.gz"))

    tmp <- data.table::fread(repeats, sep = '\t', header = TRUE,
                             drop = c('Chr', 'Start', 'End', 'Strand', 'Length'))
    rownames(tmp) <- tmp$Geneid ## here

    colnames(tmp) <- gsub(".*:([A-Z]+)-1.bam","\\1", colnames(tmp))

    tmp <- as.data.frame(tmp)
    rownames(tmp) <- tmp$Geneid
    tmp <- tmp[,-1]

    featurecounts <- CreateSeuratObject(
        tmp,
        project = "seurat",
        assay = "RNA",
        min.cells = 0,
        min.features = 0,
        names.field = 1,
        names.delim = "_",
        meta.data = NULL)

    return(featurecounts)
}

coef_var <- function(x){
        (sd(x)/mean(x))*100
}


```

```{r justfordebugging, eval = FALSE}
## params <- list(genes_cellranger = "/home/imallona/repeats_sc/runs/frozen_pbmc_donor_a/cellranger_standard/outs/filtered_feature_bc_matrix/barcodes.tsv.gz",
##                repeats_cellranger= "/home/imallona/repeats_sc/runs/frozen_pbmc_donor_a/cellranger_repeats/outs/filtered_feature_bc_matrix/barcodes.tsv.gz",
##                repeats_featurecounts= "/home/imallona/repeats_sc/runs/frozen_pbmc_donor_a/count_repeats_on_cellranger_standard/frozen_pbmc_donor_a_repeats.counts.gz")

## SRR10974769

## params <- list(run1 = '/home/imallona/repeats_sc/runs/SRR10974769/SRR10974769_pmbc_cellranger_regress_nCount_FALSE_nFeature_FALSE.rds',
##                run2 = '/home/imallona/repeats_sc/runs/5k_pbmc_v3/5k_pbmc_v3_pmbc_cellranger_regress_nCount_FALSE_nFeature_FALSE.rds',
##                run3 = '/home/imallona/repeats_sc/runs/frozen_pbmc_donor_a/frozen_pbmc_donor_a/summarize_cellranger_run_regress_nCount_FALSE_nFeature_FALSE.rds',
##                run4 = '/home/imallona/repeats_sc/runs/5k_pbmc_v3/5k_pbmc_v3_pmbc_cellranger_regress_nCount_FALSE_nFeature_FALSE.rds')

params <- list( chromium1 = '/home/imallona/repeats_sc/runs/pbmc8k/pbmc8k_pmbc_chromium_regress_nCount_FALSE_nFeature_FALSE.rds',
        chromium2 = '/home/imallona/repeats_sc/runs/5k_pbmc_v3/5k_pbmc_v3_pmbc_chromium_regress_nCount_FALSE_nFeature_FALSE.rds',
        chromium3 = '/home/imallona/repeats_sc/runs/pbmc_10k_v3/pbmc_10k_v3_pmbc_chromium_regress_nCount_FALSE_nFeature_FALSE.rds',
        smartseq1 = '/home/imallona/repeats_sc/runs/pbmcs_smartseq2_ding/pbmcs_smartseq2_ding_pmbc_smartseq2_regress_nCount_FALSE_nFeature_FALSE.rds',
        chromium4 = '/home/imallona/repeats_sc/runs/frozen_pbmc_donor_a/frozen_pbmc_donor_a_pmbc_chromium_regress_nCount_FALSE_nFeature_FALSE.rds',
        dropseq = "/home/imallona/repeats_sc/runs/ding_dropseq/ding_dropseq_pmbc_dropseq_regress_nCount_FALSE_nFeature_FALSE.rds",
        celseq2 = "/home/imallona/repeats_sc/runs/ding_celseq2/ding_celseq2_pmbc_celseq2_regress_nCount_FALSE_nFeature_FALSE.rds",
        seqwell = "/home/imallona/repeats_sc/runs/ding_seqwell/ding_seqwell_pmbc_seqwell_regress_nCount_FALSE_nFeature_FALSE.rds",
        bulk_gse80095 = "/home/imallona/repeats_sc/runs/bulk_GSE80095/bulk_GSE80095_expression.rds",
        bulk_gse94820 = "/home/imallona/repeats_sc/runs/bulk_GSE51984/bulk_GSE51984_expression.rds",
        zheng_truth = "/home/imallona/repeats_sc/runs/zheng_truth/zheng_truth_pbmc_zheng_truth_regress_nCount_FALSE_nFeature_FALSE.rds")

```


```{r}
annot <- read.table(text="id,type,origin
pbmc8k,chromium_v2,tenx
5k_pbmc_v3,chromium_v3,tenx
pbmc_10k_v3,chromium_v3,tenx
pbmcs_smartseq2_ding,smartseq2,ding
frozen_pbmc_donor_a,chromium_v1,tenx
ding_dropseq,dropseq,ding
ding_celseq2,celseq2,ding
ding_seqwell,seqwell,ding
zheng_truth,chromium_v1,zheng", sep = ",", header = TRUE)

rownames(annot) <- annot$id

```


```{r}
d <- list()
```

```{r}
## cr cellranger
## fc featurecounts
for (rds in params) {
    ## only reading single cell files here
    if (grepl('regress', rds)) {
        ## d[[run]] <- list()
        ## d[[run]]$genes_cr <- load_standard_cellranger_genes(run_name = run)
        ## d[[run]]$reps_cr <- load_cellranger_repeats(run_name = run)
        ## d[[run]]$reps_fc <- load_featurecounts_repeats(run_name = run)

        tryCatch({
            d[[sapply(strsplit(dirname(rds), .Platform$file.sep),
                      function(x) return(x[6]))]] <- readRDS(file = rds)
        },
        error = function(x) print(sprintf('error detected and ignored loading data: \n%s\n', x)))
    }
}
```

# Number of cells, flavours table  {.tabset .tabset-fade .tabset-pills}

```{r}
## name harmonization
for (dataset in names(d)) {
    d[[dataset]]$repeats_star_unique <- d[[dataset]]$repeats_featurecounts_unique
    d[[dataset]]$repeats_featurecounts_unique <- NULL
    d[[dataset]]$repeats_star_multi <- d[[dataset]]$repeats_featurecounts_multi
     d[[dataset]]$repeats_featurecounts_multi <- NULL
    d[[dataset]]$repeats_only_unique <- d[[dataset]]$control_repeats_only_unique
    d[[dataset]]$control_repeats_only_unique <- NULL
    d[[dataset]]$repeats_only_multi <- d[[dataset]]$control_repeats_only_multi
    d[[dataset]]$control_repeats_only_multi <- NULL
     d[[dataset]]$genes_star <- d[[dataset]]$genes_featurecounts
    d[[dataset]]$genes_featurecounts <- NULL
}
```


```{r}

fd <- list()

i <- 1
for (dataset in names(d)) {
    flavours <- names(d[[dataset]])
    for (flavour in flavours) {
        ncells <- ncol(d[[dataset]][[flavour]])
        avg_reads <- mean(d[[dataset]][[flavour]]@meta.data$nCount_RNA)
        avg_features <- mean(d[[dataset]][[flavour]]@meta.data$nFeature_RNA)
        fd[[i]] <- c(dataset, flavour, ncells, avg_reads, avg_features)
        i <- i + 1
    }
}

fd <- do.call(rbind.data.frame, fd)
colnames(fd) <- c('dataset',
                  'flavour',
                  'num_cells',
                  'avg_num_reads_per_cell',
                  'avg_num_genes_per_cell')

for (item in c('num_cells',
               'avg_num_reads_per_cell',
               'avg_num_genes_per_cell')) {
    fd[,item] <- as.numeric(as.character(fd[,item]))
}

```


```{r}

fd <- merge(fd, annot, by.x = 'dataset', by.y = 'id', all = TRUE)

fd$flavour <- factor(fd$flavour,
                        levels = c('genes_cellranger', 'genes_alevin',
                                   'genes_star',
                                   'both_unique', 'both_multi', 'both_alevin', 'both_cellranger',
                                   'repeats_cellranger', 'repeats_alevin',
                                   'repeats_bowtie_unique', 'repeats_bowtie_multi',
                                   'repeats_star_unique', 'repeats_star_multi',
                                   'repeats_only_unique', 'repeats_only_multi'))




```

```{r, fig.width = 10, fig.height = 5, results = 'asis'}

cat('## Num cells\n\n')

ggplot(fd, aes( y=num_cells, x=flavour, color = type)) + 
    geom_point() +
    scale_y_continuous(trans='log10') +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    facet_wrap(~dataset, ncol = 3, nrow = 3)

cat('\n\n')
```


```{r, fig.width =10, fig.height = 5, results = 'asis'}
cat('## Num reads per cell\n\n')

ggplot(fd, aes( y=avg_num_reads_per_cell, x=flavour, color = type)) + 
    geom_point() +
    scale_y_continuous(trans='log10') +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    facet_wrap(~dataset, ncol = 3, nrow = 3)

cat('\n\n')
```

```{r, fig.width = 10, fig.height = 5, results = 'asis'}
cat('## Avg genes per cell\n\n')

ggplot(fd, aes( y=avg_num_genes_per_cell, x=flavour, color = type)) + 
    geom_point() +
    ylab('avg_num_features_per_cell') +
    scale_y_continuous(trans='log10') +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    facet_wrap(~dataset, ncol = 3, nrow = 3)

cat('\n\n')
```


```{r}
pal <- read.table(text = 'identity,color
CD4+_T_cell,"#A6CEE3"
Cytotoxic_T_cell,"#1F78B4"
B_cell,"#B2DF8A"
Natural_killer_cell,"#33A02C"
CD14+_monocyte,"#FB9A99"
CD16+_monocyte,"#E31A1C"
Dendritic_cell,"#FDBF6F"
Plasmacytoid_dendritic_cell,"#FF7F00"
Plasma_cell,"#CAB2D6"
Platelet,"#6A3D9A"
unknown,"#C0C0C0"', header = TRUE, sep = ',', stringsAsFactors = FALSE)

rownames(pal) <- pal$identity
```

```{r}


zheng_pal <- read.table(text = 'truth,identity,color
memory_t,Memory_T,"#4c6085"
naive_cytotoxic,Naive_cytotoxic_T_cell,"#1F78B4"
regulatory_t,Regulatory_T,"#2549d9"
naive_t,Naive_T,"#a6cee3"
cd4_t_helper,CD4+_T_cell,"#A6CEE3"
cytotoxic_t,Cytotoxic_T_cell,"#1F78B4"
b_cells,B_cell,"#B2DF8A"
cd56_nk,Natural_killer_cell,"#33A02C"
cd14_monocytes,CD14+_monocyte,"#FB9A99"
missing_cd16mono,CD16+_monocyte,"#E31A1C"
missing_dendritic,Dendritic_cell,"#FDBF6F"
missing_plasmacytoid,Plasmacytoid_dendritic_cell,"#FF7F00"
missing_plasma,Plasma_cell,"#CAB2D6"
missing_platelet,Platelet,"#6A3D9A"
cd34,CD34+_monocyte,"#FB9A99"
', header = TRUE, sep = ',', stringsAsFactors = FALSE)

rownames(zheng_pal) <- zheng_pal$truth

## with greys for missing celltypes (when called by scina)
zheng_pal_grey <- read.table(text = 'truth,identity,color
cd4_t_helper,CD4+_T_cell,"#A6CEE3"
cytotoxic_t,Cytotoxic_T_cell,"#1F78B4"
b_cells,B_cell,"#B2DF8A"
cd56_nk,Natural_killer_cell,"#33A02C"
cd14_monocytes,CD14+_monocyte,"#FB9A99"
missing_cd16mono,CD16+_monocyte,"#E31A1C"
missing_dendritic,Dendritic_cell,"#FDBF6F"
missing_plasmacytoid,Plasmacytoid_dendritic_cell,"#FF7F00"
missing_plasma,Plasma_cell,"#CAB2D6"
missing_platelet,Platelet,"#6A3D9A"
cd34,CD34+_monocyte,"#FB9A99"
memory_t,Memory_T,"#C0C0C0"
naive_cytotoxic,Naive_cytotoxic_T_cell,"#1F78B4"
regulatory_t,Regulatory_T,"#C0C0C0"
naive_t,Naive_T,"#C0C0C0"', header = TRUE, sep = ',', stringsAsFactors = FALSE)

rownames(zheng_pal_grey) <- zheng_pal_grey$truth

```

# Zheng's {.tabset .tabset-fade .tabset-pills}

Using Zheng's dataset

```{r, results = 'asis'}
dataset <- 'zheng_truth'
for (flavour in names(d[[dataset]])) {
    cat('## ', flavour, ' {.tabset .tabset-fade .tabset-pills} \n\n')

    stopifnot(all(d[[dataset]][[flavour]]@meta.data$truth %in% zheng_pal$truth))

    for (dimred in c('pca', 'tsne', 'umap')) {
        cat('### ', dimred, ' \n\n')

        print(DimPlot(d[[dataset]][[flavour]], reduction = dimred, group.by = 'truth',
                      cols = setNames(zheng_pal$color, zheng_pal$truth)) +
              theme(legend.position="right",
                    aspect.ratio = 1,
                    plot.title = element_text(size=14),
                    axis.text=element_text(size=12),
                    axis.title=element_text(size=12)) +
              ggtitle(sprintf('%s\n%s', dataset, flavour)))

        cat('\n\n')
    }
    cat('\n\n')
}
```



# SCINA proportions



```{r}

scinas <- list()
i <- 1
for (dataset in names(d)) {
     flavours <- grep('genes', names(d[[dataset]]), value = TRUE)
     for (flavour in flavours) {
         for (identity in unique(as.character(d[[dataset]][[flavour]]@meta.data$scina))) {
             scinas[[i]] <- c(dataset, flavour, identity,
                              sum(d[[dataset]][[flavour]]@meta.data$scina == identity))
             i <- i + 1
         }
     }
}

scinas <- do.call(rbind.data.frame, scinas)
colnames(scinas) <- c('dataset', 'flavour', 'identity', 'count')
scinas$count <- as.numeric(as.character(scinas$count))
```


```{r, fig.width = 8, fig.height = 5}
print(ggplot(scinas, aes(y = count, x = flavour, fill = identity)) + 
    geom_bar(position="fill", stat="identity") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    ylab('proportion of cells') +
    facet_wrap(~dataset, ncol = 4) +
    scale_fill_manual(values = setNames(pal$color, as.character(pal$identity))))

```


# Dimensionality reductions (all datasets)  {.tabset .tabset-fade .tabset-pills} 


```{r, fig.width = 10, fig.height = 10, results = 'asis'}
for (flavour in unique(unlist(sapply(d, names)))){
    cat(sprintf('## %s  {.tabset .tabset-fade .tabset-pills} \n\n', flavour))
    p <- list()

    for (dimred in c('pca', 'tsne', 'umap')) {

        cat(sprintf('### %s \n\n', dimred))
        for (dataset in names(d)) {

            if (dataset != 'zheng_truth') {
                p[[dataset]] <- tryCatch({
                    DimPlot(d[[dataset]][[flavour]], reduction = dimred, group.by = 'scina',
                            cols = setNames(pal$color, pal$identity)) +
                        theme(aspect.ratio = 1,
                              legend.position="none", plot.title = element_text(size=14),
                              axis.text=element_text(size=12),
                              axis.title=element_text(size=12)) +
                        ggtitle(sprintf('%s\n%s', dataset, flavour))
                }, error = function(x) return(ggplot() + theme_void())
                )
            }
            else {
                ## zheng's truth has another palette
                p[[dataset]] <- tryCatch({
                    DimPlot(d[[dataset]][[flavour]], reduction = dimred, group.by = 'truth',
                            cols = setNames(zheng_pal_grey$color, zheng_pal_grey$truth)) +
                        theme(aspect.ratio = 1,
                              legend.position="none", plot.title = element_text(size=14),
                              axis.text=element_text(size=12),
                              axis.title=element_text(size=12)) +
                        ggtitle(sprintf('%s\n%s', dataset, flavour))
                }, error = function(x) return(ggplot() + theme_void())
                )
                
            }
            cat(sprintf('\n\n'))
        }
        do.call("grid.arrange", c(p, ncol = 3, nrow = 3))
        cat(sprintf('\n\n'))
    }
}

```

# F1s  {.tabset .tabset-fade .tabset-pills} 

```{r}
cobral <- list()


for (rds in params) {
    fn <- list.files(dirname(rds), pattern = '.*cobra.*TRUE.*TRUE.*')

    tryCatch({
        cobral[[sapply(strsplit(dirname(rds), .Platform$file.sep),
                      function(x) return(x[6]))]] <- readRDS(file = file.path(dirname(rds), fn))

    }, error = function(x) sprintf('caught error %s\n', x)) 
}

cobra_truth <- cobral$zheng_truth
cobral$zheng_truth <- NULL
    
cobra <- do.call(rbind.data.frame, cobral)
```

```{r}
## some classes are equivalent
cobra$flavour <- as.character(cobra$flavour)

cobra$flavour[cobra$flavour == 'genes_featurecounts'] <- 'genes_star'
cobra$flavour[cobra$flavour == 'repeats_featurecounts_unique'] <- 'repeats_star_unique'
cobra$flavour[cobra$flavour == 'repeats_featurecounts_multi'] <- 'repeats_star_multi'

cobra$flavour[cobra$flavour == 'control_repeats_only_multi'] <- 'repeats_only_multi'
cobra$flavour[cobra$flavour == 'control_repeats_only_unique'] <- 'repeats_only_unique'

table(cobra$flavour)

cobra$res <- gsub('RNA_snn_res.', 'snn_', cobra$res)
```



```{r, fig.width = 18, fig.height = 10, results = 'asis'}

cat('## By flavour \n\n')

ggplot(data = cobra,      
    aes(x = res, y =  F1, group=flavour, col = cell)) +
    geom_point() +
    facet_grid(id~flavour) +    
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    scale_colour_manual(values = setNames(pal$color, as.character(pal$identity))) 

```

```{r, fig.width = 11, fig.height = 6}
cat('## By flavour - only alevin \n\n')

ggplot(data = cobra[cobra$flavour %in% c('genes_alevin', 'repeats_alevin', 'both_alevin'),],      
    aes(x = res, y =  F1, group=flavour, col = cell)) +
    geom_point() +
    facet_grid(flavour~id) +
    scale_y_continuous(trans = 'atanh') +
    ylab('F1 score') +
    xlab('SNN resolution') +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    scale_colour_manual(values = setNames(pal$color, as.character(pal$identity))) 

```

Add whether the analyte is repeat, genes, or both


```{r}
cobra$analyte <- NA
cobra[grepl('genes', cobra$flavour), 'analyte'] <- 'genes'
cobra[grepl('repeats', cobra$flavour), 'analyte'] <- 'repeats'
cobra[grepl('both', cobra$flavour), 'analyte'] <- 'both'


table(cobra$analyte, useNA = 'always')

```


Add whether multimappers are allowed


```{r}
cobra$multimapper <- NA
cobra[grepl('alevin', cobra$flavour), 'multimapper'] <- 'alevin'
cobra[grepl('multi', cobra$flavour), 'multimapper'] <- 'multi'
cobra[grepl('unique', cobra$flavour), 'multimapper'] <- 'unique'

table(cobra$multimapper, useNA = 'always')
```

Add whether repeats are overlapping genes or not

```{r}
cobra$overlap <- 'overlap_allowed'
cobra[grepl('repeats_alevin', cobra$flavour), 'overlap'] <- 'alevin_decoy'
cobra[grepl('genes|both', cobra$flavour), 'overlap'] <- 'genes'
cobra[grepl('only', cobra$flavour), 'overlap'] <- 'no_overlap'


table(cobra$overlap, useNA = 'always')
```



```{r, fig.width = 12, fig.height = 6, results = 'asis', eval = FALSE}

cat('## By flavour \n\n')
ggplot(data = cobra,      
    aes(x = res, y =  F1, group=flavour, col = flavour)) +
    geom_point() +
    facet_grid(id~cell) + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    xlab('unsup. clust. resolution') +
    ylab('F1 score')

```


```{r, fig.width = 12, fig.height = 6, results = 'asis'}
cat('## By analyte \n\n')

ggplot(data = cobra,      
    aes(x = res, y =  F1, group=flavour, col = analyte)) +
    geom_point(aes(shape= analyte, col = analyte)) +
    facet_grid(id~cell) + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    xlab('unsup. clust. resolution') +
    ylab('F1 score')

```


```{r, fig.width = 12, fig.height = 6, results = 'asis'}

cat('## By multimapping \n\n')

ggplot(data = cobra,      
       aes(x = res, y =  F1, group=flavour, col = multimapper)) +
    geom_point(aes(shape= multimapper, col = multimapper)) +
    facet_grid(id~cell) + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    xlab('unsup. clust. resolution') +
    ylab('F1 score')

```


```{r, fig.width = 12, fig.height = 6, results = 'asis'}

cat('## By overlap with genes \n\n')

ggplot(data = cobra,      
    aes(x = res, y =  F1, group=flavour, col = overlap)) +
    geom_point(aes(shape= overlap, col = overlap)) +
    facet_grid(id~cell) + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    xlab('unsup. clust. resolution') +
    ylab('F1 score')


```

# F1's on Zheng's truth  {.tabset .tabset-fade .tabset-pills} 

```{r, results = 'asis', fig.width =10, fig.height = 5}

cat('## By flavour \n\n')

ggplot(data = cobra_truth,      
    aes(x = res, y =  F1, group=flavour, col = flavour)) +
    geom_point() +
    facet_grid(id~cell) + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    xlab('unsup. clust. resolution') +
    ylab('F1 score')
```


```{r, results = 'asis', fig.width = 10, fig.height = 5}
cat('## By identity \n\n')

ggplot(data = cobra_truth,      
    aes(x = res, y =  F1, group=flavour, col = cell)) +
    geom_point() +
    facet_grid(id~flavour) +    
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    scale_colour_manual(values = setNames(zheng_pal$color, as.character(zheng_pal$truth))) 

```



F1 scores gains/losses both genes vs genes only (alevin)


```{r, results = 'asis', fig.width =12, fig.height = 8}
delta <- list()

## celltypes <- levels(cobra$cell)
## resolutions <- levels(cobra$res)

for (run in names(cobral)) {
    delta[[run]] <- list()
    for (score in c('F1', 'pr', 're')) {
    delta[[run]][[score]] <- list()

        ## delta[[run]] <- matrix(NA, ncol = length(celltypes), nrow = length(resolutions))

    if ('genes_alevin' %in% cobral[[run]]$flavour){

        label <- 'alevin'
        genes <- cobral[[run]][cobral[[run]]$flavour == 'genes_alevin',] %>%
            select(cell, res, score) %>% spread(res, score)
   

        both <- cobral[[run]][cobral[[run]]$flavour ==  'both_alevin',] %>%
            select(cell, res, score) %>%
            spread(res, score)
    } else {
        tryCatch({
        label <- 'star'
        genes <- cobral[[run]][cobral[[run]]$flavour == 'genes_star',] %>%
            select(cell, res, score) %>% spread(res, score)
   
        both <- cobral[[run]][cobral[[run]]$flavour ==  'both_star',] %>%
            select(cell, res, score) %>%
            spread(res, score)
        }, error = function(x) print(x))
        
    }

    tryCatch({
        both_vs_genes <- data.frame(cell = genes$cell,
                                    both - genes)[,-2]
        rownames(both_vs_genes) <- genes$cell

        head(both_vs_genes)
        head(both_vs_genes  %>%  pivot_longer(!cell, names_to = "resolution", values_to = score))
        
        delta[[run]][[score]] <- data.frame(
            both_vs_genes  %>%
            pivot_longer(!cell, names_to = "resolution", values_to = score),
            score = score,
            run = run,
            type = 'both__minus__genes',
            mapper = label)
    }, error = function(x) print(x))

    }
}

f1s <- do.call(rbind.data.frame, sapply(delta, function(x) return(x['F1'])))
f1s$res <- as.numeric(as.character(gsub('RNA_snn_res.', '', f1s$resolution)))

```

```{r, eval = FALSE}
ggplot(data = f1s,      
    aes(x = res, y =  F1, group= run, col = run)) +
    geom_point(aes(shape= run)) +
    facet_grid(~cell) + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    xlab('unsup. clust. resolution') +
    ylab('F1 score with repeats vs only genes') +
    geom_smooth()

```


```{r, eval = FALSE}


ggplot(data = f1s,      
    aes(x = res, y =  F1,  col = cell)) +
    geom_point() +
    facet_grid(~run) + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    xlab('unsup. clust. resolution') +
    ylab('F1 score with repeats vs only genes') +
    scale_colour_manual(values = setNames(pal$color, as.character(pal$identity))) +
    geom_smooth()


```

Identical Louvain resolutions have very different meanings in different datasets. 

Wind approach








<!-- ## T cells vs Monocytes only -->

<!-- ```{r, eval = FALSE} -->
<!-- ggplot(data = cobra[cobra$cell %in% c('CD14+_monocyte', 'CD16+_monocyte', 'Cytotoxic_T_cell', -->
<!--                                       'CD4+_T_cell', 'Plasmacytoid_dendritic_cell'),],       -->
<!--     aes(x = res, y =  F1, group=flavour, col = cell)) + -->
<!--     geom_point() + -->
<!--     facet_grid(id~flavour) +  -->
<!--     theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--     scale_colour_manual(values = setNames(pal$color, as.character(pal$identity)))  -->
<!-- ``` -->


#  max F1s *resolution-agnostic* {.tabset .tabset-fade .tabset-pills} 

```{r, message = FALSE, warning = FALSE}

maxF1 <- list()
cobra$id <- as.factor(cobra$id)
cobra$flavour <- as.factor(cobra$flavour)
cobra$cell <- as.factor(cobra$cell)
tryCatch({
    ## cobra <- cobra[order(cobra$F1, cobra$cell, decreasing = TRUE),]
    ## maxf1 <- cobra[0,]

    for (run in levels(cobra$id)) {
        for (flavour in levels(cobra$flavour)) {
            for (cell in levels(cobra$cell)) {
                curr <- max(cobra[cobra$flavour == flavour &
                                                  cobra$cell == cell &
                                                  cobra$id == run, 'F1'])
                if (!is.finite(curr)) curr <- NA
                maxF1[[paste(run, flavour, cell)]] <- c(
                    run, flavour, cell, curr)
            }
        }
    }


    maxF1 <- do.call(rbind.data.frame, maxF1)
    colnames(maxF1) <- c('run', 'flavour', 'cell', 'max_F1')
    maxF1$max_F1 <- as.numeric(as.character(maxF1$max_F1))
    
    maxF1$cell <- factor(as.character(maxF1$cell),
                         levels = unique(as.character(maxF1$cell[order(maxF1$max_F1)])))
}, error = function(x) print(x))
```


```{r, warning = FALSE}

max_genes <- list()

for (run in levels(cobra$id)) {
    ## for (flavour in levels(cobra$flavour)) {
    for (cell in levels(cobra$cell)) {
        genes_flavour <-  'genes_alevin'

        curr <- cobra[cobra$flavour == genes_flavour &
                      cobra$cell == cell &
                      cobra$id == run,]
        
        if (is.finite(max(curr$F1))) {
            genes <- head(curr[order(curr$F1, decreasing = TRUE),], 1)
            max_genes[[paste(run, genes_flavour, cell)]] <- c(
                run, genes_flavour, cell, genes$F1, genes$res, genes$pr, genes$re )
        
            

            for (flavour in grep('repeats|both', levels(cobra$flavour), value = TRUE)) {
                
                other <- cobra[cobra$flavour == flavour &
                               cobra$cell == cell &
                               cobra$id == run &
                               cobra$res == genes$res,]

                if(nrow(other) > 0) {
                
                ## print(sprintf('%s %s', run, cell, flavourdim(other))
                
                    max_genes[[paste(run, flavour, cell)]] <- c(
                        run, flavour, cell, other$F1, other$res, other$pr, other$re )
                }
            }
            
        }
    }
}

## str(max_genes)

## table(sapply(max_genes, length))

max_genes <- do.call(rbind.data.frame, max_genes)

colnames(max_genes) <- c('run', 'flavour', 'cell', 'max_F1', 'res', 'pr', 're')
## table(max_genes$flavour)
max_genes$max_F1 <- as.numeric(as.character(max_genes$max_F1))
    
max_genes$cell <- factor(as.character(max_genes$cell),
                         levels = unique(as.character(max_genes$cell[order(max_genes$max_F1)])))

```

## Scatter

```{r, results = 'asis', fig.width =10, fig.height = 6}
## head(max_genes)

ggplot(data = max_genes, aes(x= flavour, y = max_F1, col = cell, group = cell)) +
    geom_point() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    facet_wrap(~run) +
    scale_colour_manual(values = setNames(pal$color, as.character(pal$identity)))


```

## Heatmap

<!-- Subtract the genes value to F1s? -->

```{r, results = 'asis', fig.width =10, fig.height = 6}
## max_genes %>%  pivot_wider(id_cols = c('cell', 'run', 'res'),
##                            names_from = 'flavour',
##                            values_from = 'max_F1')

max_genes %>%
    ggplot(aes(flavour, run,  fill = max_F1)) +
    geom_tile(aes(fill=max_F1)) +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    facet_wrap(~cell, ncol = 5) +
    ylab('dataset') +
    xlab('method') +
    scale_fill_gradientn(colours = c("white",  "blue", "red"),
                         values = c(0, 0.5, 1)) 

## geom_text(data=maxF1, aes(flavour, run, label = max_F1))
tmp <- max_genes %>%  pivot_wider(id_cols = c('cell', 'run', 'res'),
                                  names_from = 'flavour',
                                  values_from = 'max_F1')

## apply (tmp[]
```




Harmonize/sort flavour levels


```{r}
## tryCatch({
## sorting in a more sensible way
stopifnot(length(setdiff(levels(maxF1$flavour),
                         c('genes_cellranger', 'genes_alevin',
                           'genes_star',
                           'both_unique', 'both_multi', 'both_alevin',
                           'repeats_cellranger', 'repeats_alevin',
                           'repeats_bowtie_unique', 'repeats_bowtie_multi',
                           'repeats_star_unique', 'repeats_star_multi',
                           'repeats_only_unique', 'repeats_only_multi'))) == 0)

maxF1$flavour <- factor(maxF1$flavour,
                        levels = c('genes_cellranger', 'genes_alevin',
                                   'genes_star',
                                   'both_unique', 'both_multi', 'both_alevin',
                                   'repeats_cellranger', 'repeats_alevin',
                                   'repeats_bowtie_unique', 'repeats_bowtie_multi',
                                   'repeats_star_unique', 'repeats_star_multi',
                                   'repeats_only_unique', 'repeats_only_multi'))

maxF1$run <- as.character(maxF1$run)
maxF1$run[maxF1$run == 'Drop-seq'] <- 'ding_dropseq'
maxF1$run[maxF1$run == 'CEL-Seq2'] <- 'ding_celseq2'
maxF1$run[maxF1$run == 'Seq-Well'] <- 'ding_seqwell'

## table(maxF1$run)

maxF1$run <-  factor(maxF1$run,
                     levels = c("frozen_pbmc_donor_a",
                                "pbmc8k",
                                "5k_pbmc_v3",
                                "pbmc_10k_v3",   
                                "ding_celseq2",
                                "ding_dropseq",
                                "ding_seqwell",                                     
                                "pbmcs_smartseq2_ding"))

table(maxF1$run, useNA = 'always')
## head(maxF1)

```

Add annotations

```{r}

maxF1$analyte <- 'NA'
maxF1[grepl('genes', maxF1$flavour), 'analyte'] <- 'genes'
maxF1[grepl('repeats', maxF1$flavour), 'analyte'] <- 'repeats'
maxF1[grepl('both', maxF1$flavour), 'analyte'] <- 'both'

maxF1$analyte <- factor(maxF1$analyte, levels = c('genes', 'both', 'repeats'))

maxF1$multimapper <- 'NA'
maxF1[grepl('multi', maxF1$flavour), 'multimapper'] <- 'multi'
maxF1[grepl('alevin', maxF1$flavour), 'multimapper'] <- 'deduplicated'
maxF1[grepl('genes|cellranger|unique', maxF1$flavour), 'multimapper'] <- 'unique'



## table(maxF1$analyte, maxF1$multimapper, useNA = 'always')

maxF1$overlap <- 'overlap_allowed'
maxF1[grepl('repeats_alevin', maxF1$flavour), 'overlap'] <- 'alevin_decoy'
maxF1[grepl('genes|both', maxF1$flavour), 'overlap'] <- 'genes'
maxF1[grepl('only', maxF1$flavour), 'overlap'] <- 'no_overlap'

## head(maxF1)
```


<!-- ```{r, fig.width = 10, fig.height = 6} -->

<!-- ggplot(data = maxF1[maxF1$analyte %in% c('genes', 'repeats'),], -->
<!--        aes(x= overlap, y = max_F1, col = cell, group = cell)) + -->
<!--     geom_point() + -->
<!--     theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--     facet_grid(cell~run) + -->
<!--     scale_colour_manual(values = setNames(pal$color, as.character(pal$identity))) -->

<!-- ``` -->

## By mapping

```{r, fig.width = 10, fig.height = 4}

ggplot(data = maxF1[maxF1$analyte %in% c('genes', 'repeats'),],
       aes(x= overlap, y = max_F1, col = cell, group = cell)) +
    geom_point() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    facet_grid(~run) +
    scale_colour_manual(values = setNames(pal$color, as.character(pal$identity)))

```

## By overlap and analyte

```{r, fig.width = 12, fig.height = 6}

ggplot(data = maxF1, aes(x= run, y = max_F1, col = as.factor(analyte),
                         shape = as.factor(overlap), group = cell)) +
    geom_point() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    facet_grid( multimapper ~ cell)

```

## By overlap

```{r, fig.width = 10, fig.height = 5}

ggplot(data = maxF1, aes(x= run, y = max_F1, col = as.factor(cell), group = cell)) +
    geom_point() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    facet_grid( multimapper ~ overlap) +
    scale_colour_manual(values = setNames(pal$color, as.character(pal$identity))) 

```

## By dataset

```{r, fig.width = 9, fig.height = 8}
ggplot(data = maxF1[maxF1$analyte %in% c('genes', 'repeats'),],
        aes(x= overlap, y = max_F1, col = cell, group = cell)) +
    geom_point() +
     theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
     facet_grid(cell~run) +
     scale_colour_manual(values = setNames(pal$color, as.character(pal$identity)))
```

## By dataset and analyte

```{r, fig.width = 7, fig.height = 5}
ggplot(data = maxF1,
       aes(x= overlap, y = max_F1, col = analyte, group = cell,
           shape = analyte)) +
    geom_point() +
     theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
     facet_grid(run~cell)
```

# F1s scatterplot and heatmap {.tabset .tabset-fade .tabset-pills} 


```{r}

## text_size <- 1
axels_col <- 'black'
axes_col <- ifelse(grepl('both', levels(maxF1$flavour)), 'red', 'black')

```

## Scatter

```{r, fig.width = 10, fig.height = 5}
## help(trans_new)

ggplot(data = maxF1,
       aes(x= flavour, y = max_F1, col = cell, group = cell, shape = analyte)) +
    facet_wrap(~run, ncol = 4, nrow = 2) +
    geom_point(size = 2) +
    scale_y_continuous(trans = 'atanh') +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, colour = axes_col)) +
    scale_colour_manual(values = setNames(pal$color, as.character(pal$identity))) 

```

## Heatmap


```{r, fig.width =9, fig.height = 5}

maxF1 %>%
    ggplot(aes(flavour, run,  fill = max_F1)) +
    geom_tile(aes(fill=max_F1)) +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, colour = axes_col)) +
    facet_wrap(~cell, ncol = 5) +
    ylab('dataset') +
    xlab('method') +
    scale_fill_gradientn(colours = c("white",  "blue", "red"),
                         values = c(0, 0.5, 1)) 
    ## geom_text(data=maxF1, aes(flavour, run, label = max_F1))

```




# Weighted RIs (Wind) 


```{r}


nmis_outer <- list()
aris_outer <- list()
weights_outer <- list()

for (run in names(d)) {
    ## print(run)

    if (! 'genes_alevin' %in% names(d[[run]])) next
    ## only for the common cells not labelled 'unknown'
    common_cells <- Reduce(intersect, sapply(d[[run]], function (x) return(x@meta.data$cell)))

    common_cells <- intersect(
        common_cells,
        rownames(d[[run]]$genes_alevin@meta.data[d[[run]]$genes_alevin@meta.data$scina != 'unknown',]))

    nmis <- list()
    aris <- list()

    for (method in names(d[[run]])) {
        tryCatch({
            ## print(sprintf('%s %s', run, method))
            Y <- as.matrix(GetAssayData(d[[run]][[method]],
                                        slot = 'counts')[,common_cells])
            
            trueclass <- d[[run]]$genes_alevin@meta.data[common_cells, 'scina']
            ctStruct <- createRef(Y, trueclass)


            clusters <- list()

            for (res in grep('snn', colnames(d[[run]][[method]]@meta.data), value = TRUE)) {
                ## clusters <- sapply(d, function(x) return(x@meta.data[common_cells, res]))
                clusters[[res]] <- d[[run]][[method]]@meta.data[common_cells, res]
            }

            ## str(    clusters)
            resolutions <- names(clusters)

            ## NMI
            allNMI = matrix(0, nrow=length(resolutions), ncol=2)
            rownames(allNMI) = resolutions
            colnames(allNMI) = c("NMI", "wNMI")

            for(i in 1:length(clusters)) {
                allNMI[i,1] = wNMI(ctStruct, trueclass, clusters[[i]], FALSE)
                allNMI[i,2] = wNMI(ctStruct, trueclass, clusters[[i]])
            }

            nmis[[method]] <- allNMI
            
            ## ARI
            weights <- createWeights(Y, trueclass)
            ## str(weights)
            
            allRI = matrix(0, nrow=length(resolutions), ncol=6)
            rownames(allRI) = resolutions
            allRI

            colnames(allRI) = c("RI", "NI1","NI2","wRI","wNI1","wNI2")
            for(i in 1:length(clusters)) {
                allRI[i,1:3] = wRI(as.character(trueclass), as.numeric(clusters[[i]])) [1:3]
                allRI[i,4:6] = wRI(as.character(trueclass), as.numeric(clusters[[i]]), 
                                   weights$W0, weights$W1)[1:3]
            }

            aris[[method]] <- allRI
            weights_outer[[paste(run, method)]] <- weights
        }, error = function(x) print(x))
    }

    aris_outer[[run]] <- aris
    nmis_outer[[run]] <- nmis

    
}
```

```{r}
saveRDS(object = aris_outer, file = '/home/imallona/pbmc_aris_outer.rds')
saveRDS(object = nmis_outer, file = '/home/imallona/pbmc_nmis_outer.rds')
saveRDS(object = weights_outer, file = '/home/imallona/pbmc_weights_outer.rds')
```

# Plots wRI / wNMI  {.tabset .tabset-fade .tabset-pills} 



```{r, fig.width = 8, fig.height = 4, results = 'asis'}

for (dataset in names(nmis_outer)) {
    cat('## ', dataset, ' {.tabset .tabset-fade .tabset-pills} \n\n')
    for (method in names(nmis_outer[[dataset]])) {
        cat('### ', method, ' \n\n')
        par(mfrow = c(1,2))
        barplot(t(aris_outer[[dataset]][[method]][,c(1,4)]), beside=TRUE, ylim=c(0,1.05), 
            legend.text=TRUE, xpd=FALSE, las =2, main = sprintf('%s %s', dataset, method))
    

        barplot(t(nmis_outer[[dataset]][[method]]), beside=TRUE, ylim=c(0,1.05), 
                legend.text=TRUE, xpd=FALSE, las = 2, main = sprintf('%s %s', dataset, method))
        cat('\n\n')
    }
    cat('\n\n')
}

```

# Weight plots  {.tabset .tabset-fade .tabset-pills} 


```{r fig.width = 16, fig.height = 5, results = 'asis'}
for (id in names(weights_outer)) {

    cat('## ', id, ' \n\n')
    pheatmaps <- list()

    pheatmaps[[1]] <- pheatmap(weights_outer[[id]]$W1, main = sprintf('W1 %s', id),
                               silent = TRUE)[[4]]

    pheatmaps[[2]] <- pheatmap(weights_outer[[id]]$W0, main = sprintf('W0 %s', id),
                               silent = TRUE)[[4]]

    pheatmaps[[3]] <- pheatmap(weights_outer[[id]]$W1/weights_outer[[id]]$W0,
                               main = sprintf('W1/W0 %s', id),
                               silent = TRUE)[[4]]

    print(grid.arrange(arrangeGrob(grobs= pheatmaps, ncol=3)))

     cat(sprintf('\n\n'))
}

```




## ARIs  {.tabset .tabset-fade .tabset-pills} 


```{r fig.width = 5, fig.heigh = 5, results = 'asis'}
dataset <- 'pbmc8k'
## par(mfrow = c(1,2))
for (dataset in names(nmis_outer)) {
    cat ('### ', dataset, ' {.tabset .tabset-fade .tabset-pills}  \n\n')
    for (method in names(nmis_outer[[dataset]])) {
        cat ('#### ', method, ' \n\n')
        barplot(t(aris_outer[[dataset]][[method]][,c(1,4)]), beside=TRUE, ylim=c(0,1.05), 
                legend.text=TRUE, xpd=FALSE, las =2, main = sprintf('%s %s', dataset, method))

        cat('\n\n')
    }
    cat('\n\n')

    ## barplot(t(nmis_outer[[dataset]][[method]]), beside=TRUE, ylim=c(0,1.05), 
    ##         legend.text=TRUE, xpd=FALSE, las = 2, main = sprintf('%s %s', dataset, method))
}


```

## NMIS {.tabset .tabset-fade .tabset-pills} 


```{r fig.width = 5, fig.heigh = 5, results = 'asis'}

## par(mfrow = c(1,2))
for (dataset in names(nmis_outer)) {
    cat ('### ', dataset, ' {.tabset .tabset-fade .tabset-pills}  \n\n')
    for (method in names(nmis_outer[[dataset]])) {
        cat ('#### ', method, ' \n\n')
        barplot(t(nmis_outer[[dataset]][[method]][,c(1,4)]), beside=TRUE, ylim=c(0,1.05), 
                legend.text=TRUE, xpd=FALSE, las =2, main = sprintf('%s %s', dataset, method))

        cat('\n\n')
    }
    cat('\n\n')

    ## barplot(t(nmis_outer[[dataset]][[method]]), beside=TRUE, ylim=c(0,1.05), 
    ##         legend.text=TRUE, xpd=FALSE, las = 2, main = sprintf('%s %s', dataset, method))
}


```

<!-- # PBMC 8k in depth -->


<!-- ```{r fig.width =20, fig.heigh = 5} -->
<!-- dataset <- 'pbmc8k' -->
<!-- par(mfrow = c(1,2)) -->
<!-- for (method in names(nmis_outer[[dataset]])) { -->
<!--     barplot(t(aris_outer[[dataset]][[method]][,c(1,4)]), beside=TRUE, ylim=c(0,1.05),  -->
<!--             legend.text=TRUE, xpd=FALSE, las =2, main = sprintf('%s %s', dataset, method)) -->
    

<!--     barplot(t(nmis_outer[[dataset]][[method]]), beside=TRUE, ylim=c(0,1.05),  -->
<!--             legend.text=TRUE, xpd=FALSE, las = 2, main = sprintf('%s %s', dataset, method)) -->
<!-- } -->


<!-- ``` -->

# Markers {.tabset .tabset-fade .tabset-pills} 

For each sample, each flavour, and each celltype, get the significant markers. Then, look for overlaps.


```{r}
markers <- list()

for (rds in params) {    
    ## the params are RDS objects from Seurat
    ## aris[[sapply(strsplit(dirname(rds), .Platform$file.sep),
    ##       function(x) return(x[6]))]] <- list()
    fns <- list.files(dirname(rds), pattern = 'markers.*TRUE.*TRUE.*rds')
    for (fn in fns) {
        id <- gsub('_markers_regress_nCount_TRUE_nFeature_TRUE.rds', '', fn)
        curr_nest <- readRDS(file.path(dirname(rds), fn))
        for (item in names(curr_nest)) {
            curr_nest[[item]]$flavour <- item
            curr_nest[[item]]$dataset <- sapply(strsplit(dirname(rds), '/'),
                                                function(x) return(x[6]))
        }

        curr_nest <- lapply(curr_nest, function(x) x %>%
                                                   group_by(cluster))

        markers[[id]] <- do.call(rbind.data.frame, curr_nest)
        markers[[id]] <- markers[[id]][markers[[id]]$p_val_adj < 0.05,]
    }
}

markers_truth <- markers$zheng_truth

markers$zheng_truth <- NULL

markers <- do.call(rbind.data.frame, markers)


## colnames(markers)
## table(markers$cluster)
## table(markers$flavour)
## table(markers$dataset)
```


```{r, fig.width = 6, fig.height=5, results = 'asis'}

markers <- as.data.frame(markers)


for (celltype in unique(markers$cluster)) {
    cat('## ', celltype, '\n\n')
    ul <- list()
    for (dataset in unique(markers[markers$cluster == celltype, 'dataset'])) {

        for (flavour in grep('repeats', unique(markers[ markers$cluster == celltype &
                                                        markers$dataset == dataset, 'flavour']),
                             value = TRUE)) {
                    
            ul[[paste(celltype, dataset, flavour, sep = "_")]] <- markers[
                markers$cluster == celltype &
                markers$dataset == dataset &
                markers$flavour == flavour, 'gene']
        }
    }
    print(upset(fromList(ul), nsets = 10,  text.scale = 0.8))
    grid.text(paste(celltype) ,x = 0.65, y=0.95, gp=gpar(fontsize = 10))
    
    ## rm(ul)
    cat('\n\n')

}


## str(ul)


```

```{r}
head(markers)

markers_truth <- as.data.frame(markers_truth)
markers_truth$cluster <- sapply(strsplit(as.character(markers_truth$cluster),
                                         split = '\n', fixed = TRUE),
                                function(x) return(x[1]))
head(markers_truth)
```

# Markers per cell type, concordance across methods {.tabset .tabset-fade .tabset-pills} 


```{r, fig.width =8, fig.height = 4, results = 'asis'}

ul <- list()
for (celltype in unique(markers$cluster)) {
    cat('## ', celltype, ' {.tabset .tabset-fade .tabset-pills} \n\n')

   
    for (flavour in unique(markers[markers$cluster == celltype, 'flavour'])) {
        cat('### ', flavour, ' \n\n')
    
        ul <- list()
        for (dataset in unique(markers[markers$cluster == celltype &
                                     markers$flavour == flavour, 'dataset'])) {

            ul[[paste(dataset, flavour, sep = "_")]] <- markers[
                markers$flavour == flavour  &
                markers$cluster == celltype &
                markers$dataset == dataset, 'gene']
   
        
        }
        ## str(ul)
        
        print(upset(fromList(ul), nsets = length(ul)))
        grid.text(paste(celltype, flavour) ,x = 0.65, y=0.95, gp=gpar(fontsize = 10))

    cat('\n\n')

    }
    cat('\n\n')
}


```

# Per celltype plus truth concordance  {.tabset .tabset-fade .tabset-pills} 

```{r}

## dict <- zheng_pal[!grepl('missing', zheng_pal$truth),]

dict <- structure(list(
    truth = c("cd4_t_helper", "cytotoxic_t", "b_cells", 
              "cd56_nk", "cd14_monocytes", "cd34", "memory_t", "naive_cytotoxic", 
              "regulatory_t", "naive_t"),
    identity = c("CD4+_T_cell", "Cytotoxic_T_cell", 
                 "B_cell", "Natural_killer_cell", "CD14+_monocyte", "CD34+_monocyte", 
                 "Memory_T", "Naive_cytotoxic_T_cell", "Regulatory_T", "Naive_T"),
    flattened = c("CD4+_T_cell", "Cytotoxic_T_cell", 
                 "B_cell", "Natural_killer_cell", "CD14+_monocyte", "CD16+_monocyte", 
                 "Other", "Other", "Other", "Other"),
    color = c("#A6CEE3", "#1F78B4", "#B2DF8A", "#33A02C", "#FB9A99", 
              "#FB9A99", "#C0C0C0", "#1F78B4", "#C0C0C0", "#C0C0C0")),
    row.names = c("cd4_t_helper", "cytotoxic_t", "b_cells", "cd56_nk", "cd14_monocytes", "cd34", 
                  "memory_t", "naive_cytotoxic", "regulatory_t", "naive_t"),
    class = "data.frame")

markers_both <- rbind(markers, markers_truth)

markers_both <- markers_both[markers_both$cluster %in% c(levels(markers$cluster), dict$truth),]
markers_both$cluster <- as.character(markers_both$cluster)
for (zheng_id in dict$truth) {
    markers_both$cluster[markers_both$cluster == zheng_id] <- dict[dict$truth == zheng_id,
                                                                   'flattened']
}

## table(markers_both$cluster, markers_both$flavour)
markers_both <- markers_both[markers_both$cluster != 'Other',]

```

Adding the `both_cellranger` as `both_multi`


```{r}

markers_both$flavour[markers_both$flavour == 'both_cellranger'] <- 'both_multi'
```

```{r, fig.width =8, fig.height=4, results = 'asis'}

flavour <- 'repeats_cellranger'
celltype <- "CD14+_monocyte"


for (flavour in c('repeats_cellranger', 'genes_cellranger')) {
    cat('## ', flavour, ' \n\n')
    ul <- list()
    for (dataset in unique(markers_both[markers_both$cluster == celltype &
                                        markers_both$flavour == flavour, 'dataset'])) {

        ul[[paste(dataset, flavour, sep = "_")]] <- markers_both[
            markers_both$flavour == flavour  &
            markers_both$cluster == celltype &
            markers_both$dataset == dataset, 'gene']
        
        
    }
    ## str(ul)

    print(upset(fromList(ul), nsets = length(ul)))
    grid.text(paste(celltype, flavour) ,x = 0.65, y=0.95, gp=gpar(fontsize = 10))
    cat('\n\n')
}


```


# Markers III By celltype, no matter which method  {.tabset .tabset-fade .tabset-pills} 



```{r, fig.width =6, fig.height=4, results = 'asis'}


ul <- list()
for (celltype in unique(markers$cluster)) {
    cat('## ', celltype, ' \n\n')
    
    ul[[celltype]] <- NULL
    for (dataset in unique(markers[markers$cluster == celltype, 'dataset'])) {

        ## print(dataset)
            
        ul[[dataset]] <- unique(c(ul[[celltype]], markers[ grepl('repeats', markers$flavour) &
                                                             markers$cluster == celltype &
                                                             markers$dataset == dataset, 'gene']))

         
    }
    
   
    print(upset(fromList(ul), nsets  = length(ul)))
    grid.text(paste(celltype) ,x = 0.65, y=0.95, gp=gpar(fontsize = 10))
    cat('\n\n')
}




```

# Markers III with added truth {.tabset .tabset-fade .tabset-pills} 

```{r, results = 'asis'}

ul <- list()
for (celltype in unique(markers_both$cluster)) {
    cat('## ', celltype, ' \n\n')
    
    ul[[celltype]] <- NULL
    for (dataset in unique(markers_both[markers_both$cluster == celltype, 'dataset'])) {

        ## print(dataset)
            
        ul[[dataset]] <- unique(c(ul[[celltype]], markers_both[ grepl('repeats', markers_both$flavour) &
                                                             markers_both$cluster == celltype &
                                                             markers_both$dataset == dataset, 'gene']))

         
    }
    
   
    print(upset(fromList(ul), nsets  = length(ul)))
    grid.text(paste(celltype) ,x = 0.65, y=0.95, gp=gpar(fontsize = 10))
    cat('\n\n')
}
```


# Markers IV By celltype, method influence  {.tabset .tabset-fade .tabset-pills}



```{r, fig.width = 8, fig.height=6, results = 'asis'}

for (celltype in unique(markers$cluster)) {
    cat('## ', celltype, ' {.tabset .tabset-fade .tabset-pills} \n\n')
    

    for (dataset in unique(markers[markers$cluster == celltype, 'dataset'])) {
        cat('### ', dataset, ' \n\n')
    
        ul <- list()
        for (flavour in unique(markers[markers$cluster == celltype &
                                     markers$dataset == dataset, 'flavour'])) {
            
            ul[[paste(dataset, flavour)]] <-  markers[markers$flavour == flavour &
                                      markers$cluster == celltype &
                                      markers$dataset == dataset, 'gene']

        }

        print(upset(fromList(ul), nsets  = length(ul)))
        grid.text(paste(celltype, dataset) ,x = 0.65, y=0.95, gp=gpar(fontsize = 10))
        cat('\n\n')

    }
    cat('\n\n')
}


```


<!-- # Markers V  -->

<!-- ```{r, fig.width =12, fig.height=5, eval = FALSE} -->
<!-- ft <- as.data.frame(table(markers[markers$cluster == 'CD14+_monocyte', 'gene'])) -->

<!-- ft <- ft[order(ft$Freq, decreasing = TRUE),] -->
<!-- head(ft) -->
<!-- selected <- as.character(head(ft, 20)$Var1) -->


<!-- ## get cpms from the seurat object, only 5 cells (first 5 indexed) -->
<!-- foo <- list() -->
<!-- for (item in names(d)) { -->
<!--     ## foo <- list() -->
<!--     for (flavour in grep('repeats', names(d[[item]]), value = TRUE)) { -->
<!--         tmp <- as.data.frame(FetchData(d[[item]][[flavour]], -->
<!--                                        vars = selected, -->
<!--                                        slot = 'data'))[1:5,] -->
<!--         tmp$cell <- rownames(tmp) -->

        
<!--         tmp <- merge(d[[item]][[flavour]]@meta.data[, c('cell', 'scina')], -->
<!--                      tmp, by = 'cell', all.x = TRUE) -->
<!--         rownames(tmp) <- tmp$cell -->
<!--         tmp <- melt(tmp, id.vars = c('cell', 'scina')) -->
<!--         tmp$flavour <- flavour -->
<!--         tmp$dataset <-  item -->
<!--         foo[[paste(item, flavour)]] <- tmp -->
        
        
<!--     } -->
<!--     ## foo <- do.call(rbind.data.frame, foo) -->
<!-- } -->

<!-- foo <- do.call(rbind.data.frame, foo) -->
<!-- head(foo) -->


<!-- print(ggplot(foo[foo$scina %in% c('CD14+_monocyte', 'CD16+_monoctye', 'CD4+_T_cell'),], -->
<!--              aes(x = variable, y = value, fill = flavour)) +  -->
<!--       geom_boxplot() + -->
<!--       scale_y_continuous(trans='log10') + -->
<!--       scale_fill_discrete(drop=FALSE) + -->
<!--       facet_grid(dataset ~ scina) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->



<!-- ``` -->

<!-- # Cluster techniques and flavours and celltypes? -->


# Markers cluster dendrogram

With manhattan distances over binary data (presence/ausence of the positive marker). Equivalent to Hamming distance.

For genes

```{r}
detected <- unique(markers[grepl('genes', markers$flavour), 'gene'])

length(detected)

fc <- list()
annot <- list()
for (flavour in unique(grep('genes', markers$flavour, value = TRUE))) {
    for (dataset in unique (markers[markers$flavour == flavour, 'dataset'])) {
        tmp  <- markers[markers$flavour == flavour & markers$dataset == dataset, 'gene']

        fc[[paste(flavour, dataset)]] <- as.numeric(detected %in% tmp)
        annot[[paste(flavour, dataset)]] <- c(flavour, dataset,
                                              sum(fc[[paste(flavour, dataset)]]))
    }
}

fc <- do.call(cbind.data.frame, fc)

annot <- do.call(rbind.data.frame, annot)
colnames(annot) <- c('flavour', 'dataset', 'num_markers')
annot$num_markers <- as.numeric(annot$num_markers)
rownames(annot) <- paste(annot$flavour, annot$dataset)
rownames(fc) <- detected
## fc <- listfeature = detected)

## plot(hclust(dist(t(fc), method = 'manhattan')))

```


```{r}

rownames(annot) <- paste(annot$flavour, annot$dataset)


ggtree(hclust(dist(t(fc), method = 'manhattan')),
       branch.length = "none", size = 0.1) +
    geom_th_heatmap(th_data = annot['dataset'], name = "hm2", gap = 0.2, 
                    cluster_column = FALSE,
                    size_coltree = 0.5,
                    rel_width = 0.1) +
    scale_fill_brewer(palette = "Set2") +
    labs(fill = "Dataset") +
    new_scale_fill() +
    geom_th_heatmap(th_data = annot['flavour'], name = "hm1", gap = 0.2, 
                    cluster_column = FALSE,
                    size_coltree = 0.5,
                    rel_width = 0.1) +
    scale_fill_brewer(palette = "Set1") +
    labs(fill = "Flavour")+
    new_scale_fill() +
    geom_th_heatmap(th_data = annot['num_markers'], name = "hm3", gap = 0.2,
                    rel_width = 0.1) +
    labs(fill = "Num. markers")+
    scale_fill_viridis_c() 

```

For repeats

```{r}
detected <- unique(markers[grepl('repeats', markers$flavour), 'gene'])

length(detected)

fc <- list()
annot <- list()
for (flavour in unique(grep('repeats', markers$flavour, value = TRUE))) {
    for (dataset in unique (markers[markers$flavour == flavour, 'dataset'])) {
        tmp  <- markers[markers$flavour == flavour & markers$dataset == dataset, 'gene']

        fc[[paste(flavour, dataset)]] <- as.numeric(detected %in% tmp)
        
        annot[[paste(flavour, dataset)]] <- c(flavour, dataset,
                                              sum(fc[[paste(flavour, dataset)]]))
    }
}

fc <- do.call(cbind.data.frame, fc)

annot <- do.call(rbind.data.frame, annot)
colnames(annot) <- c('flavour', 'dataset', 'num_markers')
rownames(fc) <- detected
## fc <- listfeature = detected)

## plot(hclust(dist(t(fc), method = 'manhattan')))
```

@todo add heatmap labels (columns)

```{r}
## head(fc)
head(annot)

rownames(annot) <- paste(annot$flavour, annot$dataset)
annot$num_markers <- as.numeric(annot$num_markers)

annot$analyte <- 'NA'
annot[grepl('genes', annot$flavour), 'analyte'] <- 'genes'
annot[grepl('repeats', annot$flavour), 'analyte'] <- 'repeats'
annot[grepl('both', annot$flavour), 'analyte'] <- 'both'

annot$analyte <- factor(annot$analyte, levels = c('genes', 'both', 'repeats'))

annot$multimapper <- 'NA'
annot[grepl('multi', annot$flavour), 'multimapper'] <- 'multi'
annot[grepl('alevin', annot$flavour), 'multimapper'] <- 'deduplicated'
annot[grepl('genes|cellranger|unique', annot$flavour), 'multimapper'] <- 'unique'

## table(annot$analyte, annot$multimapper, useNA = 'always')

annot$overlap <- 'overlap_allowed'
annot[grepl('repeats_alevin', annot$flavour), 'overlap'] <- 'alevin_decoy'
annot[grepl('genes|both', annot$flavour), 'overlap'] <- 'genes'
annot[grepl('only', annot$flavour), 'overlap'] <- 'no_overlap'

annot$mapper <- 'NA'
annot[grepl('cellranger|star|featurecounts|only', annot$flavour), 'mapper'] <- 'star'
annot[grepl('bowtie', annot$flavour), 'mapper'] <- 'bowtie'
annot[grepl('alevin', annot$flavour), 'mapper'] <- 'alevin'
table(annot$mapper)

annot$counter <- 'featurecounts'
annot[grepl('alevin', annot$flavour), 'counter'] <- 'alevin'

```

```{r}


dim(as.matrix(dist(t(fc), method = 'manhattan')))
```

```{r}
## ?geom_th_heatmap

dim(fc)

fc[1:3,1:3]
str(annot)

ggtree(hclust(dist(t(fc), method = 'manhattan'), method = 'ward.D2'),
       branch.length = "none", size = 0.1) +

    geom_th_heatmap(th_data = as.matrix(dist(t(fc), method = 'manhattan')),
                    name = "hm0", gap = 0.2, 
                    cluster_column = TRUE,
                    hclust_method = 'ward.D2',
                    dist_method = 'manhattan',
                    show_coltree = FALSE,
                    rel_width = 2) +
    labs(fill = "Hamming distance") +
    scale_fill_viridis_c(direction = -1) +
    geom_th_text(name = "hm0", side = "top", nudge_y = 0.5,
                 angle = 90, size = 3) +

    new_scale_fill() +
    geom_th_heatmap(th_data = annot['dataset'], name = "hm1", gap = 0.2, 
                    cluster_column = FALSE,
                    size_coltree = 0.5,
                    rel_width = 0.1) +
    scale_fill_brewer(palette = "Set2") +
    labs(fill = "Dataset") +
    geom_th_text(name = "hm1", side = "top", nudge_y = 0.5,
                  angle = 90, size = 3) +
    
    new_scale_fill() +
    geom_th_heatmap(th_data = annot['mapper'], name = "hm2", gap = 0.2, 
                    cluster_column = FALSE,
                    size_coltree = 0.5,
                    rel_width = 0.1) +
    scale_fill_brewer(palette = "Set1") +
    labs(fill = "Mapper")+
    geom_th_text(name = "hm2", side = "top", nudge_y = 0.5,
                 angle = 90, size = 3) +


    new_scale_fill() +
    geom_th_heatmap(th_data = annot['counter'], name = "hm3", gap = 0.2, 
                    cluster_column = FALSE,
                    size_coltree = 0.5,
                    rel_width = 0.1) +
    scale_fill_brewer(palette = "Set1") +
    labs(fill = "Counter") +
     geom_th_text(name = "hm3", side = "top", nudge_y = 0.5,
                  angle = 90, size = 3) +
    
    new_scale_fill() +
    geom_th_heatmap(th_data = annot['overlap'], name = "hm4", gap = 0.2, 
                    cluster_column = FALSE,
                    size_coltree = 0.5,
                    rel_width = 0.1) +
    scale_fill_brewer(palette = "Set1") +
    labs(fill = "Overlap") +
    geom_th_text(name = "hm4", side = "top", nudge_y = 0.5,
                 angle = 90, size = 3) +
    new_scale_fill() +
    
    geom_th_heatmap(th_data = annot['multimapper'], name = "hm5", gap = 0.2, 
                    cluster_column = FALSE,
                    size_coltree = 0.5,
                    rel_width = 0.1) +
    scale_fill_brewer(palette = "Set1") +
    labs(fill = "Multimappers")+
    geom_th_text(name = "hm5", side = "top", nudge_y = 0.5,
                 angle = 90, size = 3) +

    
    new_scale_fill() +
    geom_th_heatmap(th_data = annot['num_markers'], name = "hm6", gap = 0.2,
                    rel_width = 0.1) +
    labs(fill = "Num. markers")+
    scale_fill_viridis_c() +
    geom_th_text(name = "hm6", side = "top", nudge_y = 0.5,
                 angle = 90, size = 3) +
    geom_th_title(name = "hm6", side = "top",
                  label = "test", nudge_y = -0.8) +
    
    theme(plot.margin=unit(c(2, 1, 1.5, 1.2),"cm"))

#     guides(fill = guide_legend(order = 1, ncol = 1),
        ## fill2 = guide_legend(ncol = 1))  

    ## guides(fill=guide_legend(ncol = 4)) +

```

# Bulk  {.tabset .tabset-fade .tabset-pills}

```{r}
opts_chunk$set(fig.width = 5,
               fig.height = 5,
               dev = 'png',
               cache = FALSE,
               echo = TRUE,
               include = TRUE,
               cache.lazy = FALSE,
               warning = TRUE,
               error = TRUE,
               message = FALSE)
```

```{r}

bulk <- list()
for (bulk_id in grep('bulk', params, value = TRUE)) {
    short_id <- sapply(strsplit(dirname(bulk_id), .Platform$file.sep),
                       function(x) return(x[6]))
    bulk[[short_id]] <- readRDS(bulk_id)
}


## str(bulk)
```

```{r}

monocyte_markers <- unique(markers_both[grepl('repeats', markers_both$flavour) & markers_both$cluster == 'CD14+_monocyte',])

plot(monocyte_markers$pct.1, monocyte_markers$pct.2, cex = monocyte_markers$avg_logFC,
     col = ifelse(monocyte_markers$avg_logFC < 1, 'red', 'black'))

monocyte_markers <- unique(markers_both[grepl('repeats', markers_both$flavour) & markers_both$cluster == 'CD14+_monocyte' & markers_both$avg_logFC >= 1,])

plot(monocyte_markers$pct.1, monocyte_markers$pct.2, cex = monocyte_markers$avg_logFC,
     col = ifelse(monocyte_markers$avg_logFC < 1, 'red', 'black'))

other_markers <- unique(markers_both[grepl('repeats', markers_both$flavour) & !markers_both$cluster == 'CD14+_monocyte' & markers_both$avg_logFC >= 1,])

```

Or maybe just go for the count tables and plot log2cpms for some candidates?


```{r, eval = FALSE}
## lapply(bulk, function(x) head(x[[1]], 10))

vs_bulk <- list()

for (gse in names(bulk)) {
    for (flavour in names(bulk[[gse]])) {
       
        common <- intersect(unique(monocyte_markers$gene),
                            rownames(bulk[[gse]][[flavour]][[1]]))

        vs_bulk[[paste(gse, flavour)]] <- c(
            length(common),
            
            sum(bulk[[gse]][[flavour]][[1]][, 'logFC'] >= 0 &
                bulk[[gse]][[flavour]][[1]][, 'FDR'] < 0.05),
            sum(bulk[[gse]][[flavour]][[1]][, 'logFC'] < 0 &
                bulk[[gse]][[flavour]][[1]][, 'FDR'] < 0.05),
            
            sum(bulk[[gse]][[flavour]][[1]][common, 'logFC'] >= 0 &
                bulk[[gse]][[flavour]][[1]][common, 'FDR'] < 0.05),
            sum(bulk[[gse]][[flavour]][[1]][common, 'logFC'] < 0 &
                bulk[[gse]][[flavour]][[1]][common, 'FDR'] < 0.05))
            
                                            
        print(sprintf('%s %s: %s markers', gse, flavour, length(common)))
        
        print(max(bulk[[gse]][[flavour]][[1]]$FDR))
        print(summary((bulk[[gse]][[flavour]][[1]][common, 'logFC'])))
        print(table(bulk[[gse]][[flavour]][[1]][common, 'logFC'] >= 0 &
                    bulk[[gse]][[flavour]][[1]][common, 'FDR'] < 0.05))
    }

}
```





But these are edgeRs of something against something else... maybe we could just have CPMs represented? Did we get a list of markers anyway?



```{r}

avgs <- list()
for (gse in names(bulk)) {
    avgs[[gse]] <- list()
    for (flavour in grep('repeats', names(bulk[[gse]]), value = TRUE)) {

        ## on averaging only markers
        ## curr <- intersect(monocyte_markers$gene, rownames(cpm(bulk[[gse]][[flavour]])))

        ## rather, average everything
        curr <- rownames(cpm(bulk[[gse]][[flavour]]))
        tmp <- list()
        for (group in colnames(bulk[[gse]][[flavour]]$design)) {
            ids <- row.names(bulk[[gse]][[flavour]]$design[bulk[[gse]][[flavour]]$design[,group] == 1,])
            

            
           tmp[[paste(gse, flavour, group)]] <-  aveLogCPM(bulk[[gse]][[flavour]]$counts[curr,ids])
        }

        avgs[[gse]][[flavour]] <- do.call(cbind.data.frame, tmp)
        rownames(avgs[[gse]][[flavour]]) <- curr
    }
}

## str(avgs)
```

## Monocyte markers bulk vs scRNA-seq 

```{r}
## this is CPM representation of a seurat object

## RelativeCounts(GetAssayData(d[[1]][[1]], 'counts'), scale.factor = 1e6, verbose = TRUE)

```

```{r, results = 'asis', fig.width = 6, fig.height = 6}
 ## bottom, left, top, and right. 
par(mfrow = c(1,2), pty = 's', mar=c(6.1, 6.1, 4.1, 2.1))
for (flavour in unique(sapply(avgs, names))) {

    for (item in names(d)) {

        cat('### ', item, ' ', flavour, ' \n\n')
        common <- intersect(rownames(avgs[[1]][[flavour]][1]),
                            rownames(GetAssayData(d[[item]][[flavour]])))

        common_markers <- common %in% monocyte_markers$gene
        ## plot(AverageExpression(d[[flavour]][[item]]),
        ##      avgs[[1]][[flavour]][1])
        ## }

        tryCatch({
        ## better normalize this
            plot(x = log1p(AverageExpression(
                     object = d[[item]][[flavour]])$RNA[common,'CD14+_monocyte']),
             y = avgs[[1]][[flavour]][1][common,],
             col = ifelse(common_markers, 'red', ac(col = 'black', alpha = 0.5)),
             pch = 19,
             xlab = sprintf('%s %s\n(log ncounts in monocytes)', item, flavour),
             ylab = sprintf('%s %s\n(log CPMs in monocytes)', names(avgs)[1], flavour))
             

        plot(log1p(AverageExpression(object = d[[item]][[flavour]])$RNA[common,'CD14+_monocyte']),
             avgs[[2]][[flavour]][1][common,],
             col = ifelse(common_markers, 'red', ac(col = 'black', alpha = 0.5)),
             pch = 19,
             xlab = sprintf('%s %s\n(log ncounts in monocytes)', item, flavour),
             ylab = sprintf('%s %s\n(log CPMs in monocytes)', names(avgs)[2], flavour))
        }, error = function(x) print(x))

        cat('\n\n')
    }
}

```

## Monocyte markers in sc monocytes vs bulk  nonmonocytes



```{r, fig.width = 10, fig.height = 10}
## bottom, left, top, and right. 
par(mfrow = c(1,2), pty = 's', mar=c(6.1, 6.1, 4.1, 2.1))

for (flavour in unique(sapply(avgs, names))) {

    for (item in names(d)) {

        cat('### ', item, ' ', flavour, ' \n\n')
        common <- intersect(rownames(avgs[[1]][[flavour]][1]),
                            rownames(GetAssayData(d[[item]][[flavour]])))

        common_markers <- common %in% monocyte_markers$gene
        ## plot(AverageExpression(d[[flavour]][[item]]),
        ##      avgs[[1]][[flavour]][1])
        ## }

        tryCatch({
        ## better normalize this
            plot(x = log1p(AverageExpression(
                     object = d[[item]][[flavour]])$RNA[common,'CD14+_monocyte']),
             y = avgs[[1]][[flavour]][2][common,],
             col = ifelse(common_markers, 'red', ac(col = 'black', alpha = 0.5)),
             pch = 19,
             xlab = sprintf('%s %s\n(log ncounts in monocytes)', item, flavour),
             ylab = sprintf('%s %s\n(log CPMs in non monocytes)', names(avgs)[1], flavour))
             

        plot(log1p(AverageExpression(object = d[[item]][[flavour]])$RNA[common,'CD14+_monocyte']),
             avgs[[2]][[flavour]][2][common,],
             col = ifelse(common_markers, 'red', ac(col = 'black', alpha = 0.5)),
             pch = 19,
             xlab = sprintf('%s %s\n(log ncounts in monocytes)', item, flavour),
             ylab = sprintf('%s %s\n(log CPMs in non monocytes)', names(avgs)[2], flavour))
        }, error = function(x) print(x))

        cat('\n\n')
    }
}

```

# Paired monocyte vs T cell plots

Also, do the top repeat markers show a difference in the bulk datasets? Maybe use log2FC? For that we would need to have logFC that are monocytes vs T cell sonly, and parse the bulk `de` files and not the expression ones.


```{r}
## repeats detected in at least 3 datasets with at least 3 methods

tmp <- apply(as.data.frame.matrix(table(monocyte_markers$gene, monocyte_markers$dataset)) > 0,
      1,
      sum)

## length(tmp[tmp >= 3])



top_monocyte_markers <- names(tmp[tmp >= 3])
## top_monocyte_markers <- head(monocyte_markers[order(monocyte_markers$avg_logFC, decreasing = TRUE), 'gene'], 10)
## length(unique(top_monocyte_markers))

```

```{r, fig.width = 6, fig.height = 4}
top <- list()
for (marker in top_monocyte_markers) {
    for (gse in names(avgs)) {
        for (flavour in names(avgs[[gse]])) {
            for (ssample in colnames(avgs[[gse]][[flavour]])) {
                top[[paste(marker, flavour, ssample, gse)]] <- c(marker, gse, flavour, ssample,
                                               avgs[[gse]][[flavour]][marker, ssample])
            }
        }
    }
}

top <- do.call(rbind.data.frame, top)
## head(top)
colnames(top) <- c('marker', 'gse', 'flavour', 'id', 'avg_cpm')

## table(top$id)
top$celltype <- ifelse(!grepl('T_cells|tcell', top$id), 'monocyte', 't_cell')
top$paired <- paste(top$gse, top$flavour, top$marker)
## table(top$paired)
top$avg_cpm <- as.numeric(top$avg_cpm)
                          
top %>%
  ggplot(aes(y = avg_cpm, x  = celltype)) +
    geom_point(aes(color = marker)) +
    facet_grid(gse ~ flavour) +
    geom_line(aes(group = paired)) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

```

```{r, fig.width = 6, fig.height = 4}
top %>%
  ggplot(aes(y = avg_cpm, x  = celltype)) +
    geom_point(aes(color = gse)) +
    facet_grid(marker ~ flavour) +
    geom_line(aes(group = paired)) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```



these top markers, how do they behave in our single cell data?
https://pubmed.ncbi.nlm.nih.gov/11601494/


This is very unefficient, todo fix

```{r, fig.width = 8, fig.height =6}
## top_monocyte_markers
avges <- list()
for (marker in top_monocyte_markers){
    for (run in names(d)) {
        for (flavour in grep('repeats', names(d[[run]]), value = TRUE)) {
            tryCatch({
                avges[[paste(run, flavour, marker, 'mono')]] <- c(
                    log1p(AverageExpression(
                        object = d[[run]][[flavour]])$RNA[marker,]$'CD14+_monocyte'),
                    marker, run, flavour, 'CD14+_monocyte')
                
                avges[[paste(run, flavour, marker, 't')]] <- c(
                    log1p(AverageExpression(
                        object = d[[run]][[flavour]])$RNA[marker,]$'Cytotoxic_T_cell'),
                    marker, run, flavour, 'Cytotoxic_T_cell')
                }, error = function(x) print(x))
        }
    }
}
## names(avges)

avges <- do.call(rbind.data.frame, avges)
colnames(avges) <- c('log1p_ave_expr', 'marker', 'flavour', 'run', 'cell')
avges$log1p_ave_expr <- as.numeric(as.character(avges$log1p_ave_expr))

avges$paired <- paste(avges$run, avgs$flavour, avges$marker)

avges %>%
  ggplot(aes(y = log1p_ave_expr, x  = cell)) +
    geom_point(aes(color = run)) +
    facet_grid(marker ~ flavour) +
    geom_line(aes(group = paired)) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

```

```{r, fig.width = 8, fig.height = 10}

monocyte_markers %>%
    filter(gene %in% top_monocyte_markers) %>%
    ggplot(aes(y = avg_logFC, x  = dataset)) +
    geom_point(aes(color = flavour)) +
    facet_grid(gene ~ flavour) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```


Add bulk data, would need logFCs too


```{r, fig.width = 7, fig.height = 7}
monocyte_markers %>%
    filter(gene %in% top_monocyte_markers) %>%
    ggplot(aes(y = avg_logFC, x  = flavour)) +
    geom_point(aes(color = dataset)) +
    facet_grid(gene ~ dataset) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

# Repeat markers in dimreds  {.tabset .tabset-fade .tabset-pills} 


```{r, fig.width = 10, fig.height = 10, results = 'asis'}
for (flavour in grep('repeats|both', unique(unlist(sapply(d, names))), value = TRUE)){
    cat(sprintf('## %s  {.tabset .tabset-fade .tabset-pills} \n\n', flavour))
    p <- list()

    for (dimred in c('pca', 'tsne', 'umap')) {
        cat(sprintf('### %s  {.tabset .tabset-fade .tabset-pills} \n\n', dimred))
        for (feature in top_monocyte_markers) {
            cat(sprintf('#### %s \n\n', feature))
            for (dataset in names(d)) {
                
                p[[dataset]] <- tryCatch({
                    FeaturePlot(d[[dataset]][[flavour]], reduction = dimred,
                                features = feature) +
                        theme(aspect.ratio = 1,
                              legend.position="none", plot.title = element_text(size=14),
                              axis.text=element_text(size=12),
                              axis.title=element_text(size=12)) +
                        ggtitle(sprintf('%s\n%s', dataset, flavour))
                }, error = function(x) return(ggplot() + theme_void())
                )
                
            }
            do.call("grid.arrange", c(p, ncol = 3, nrow = 3, top = feature))
            cat('\n\n')
        }
        cat(sprintf('\n\n'))
    }
    
    cat(sprintf('\n\n'))
}

```


<!-- ```{r, results = 'asis'} -->
<!-- for (flavour in unique(sapply(avgs, names))) { -->
<!--     cat('## ', flavour, '\n\n') -->
<!--     par(mfrow = c(1,2))  -->
<!--     boxplot(avgs[[1]][[flavour]]) -->
<!--     boxplot(avgs[[2]][[flavour]]) -->
<!--     cat('\n\n') -->

<!--     ## stripchart(avgs[[1]][[flavour]], vertical = TRUE,data=x, method="jitter",   -->
<!-- ## add=TRUE, pch=19) -->
<!-- } -->
<!-- ``` -->

<!-- here -->

<!-- ```{r} -->

<!-- res <- list() -->
<!-- for (item in names(bulk)) { -->
<!--     res[[item]] <- lapply(conts, function(x) { -->
<!--         qlf <- glmQLFTest(fit[[item]], contrast = x) -->
<!--         tt <- topTags(qlf, n = Inf, sort.by = "PValue")$table -->
<!--     }) -->
<!-- } -->


<!-- ``` -->



<!-- ```{r} -->
<!-- names(bulk) -->

<!-- ## samplemonocyte-sampletcell -->
<!-- bulk_yardstick <- bulk[[1]][['repeats_star_multi']][[1]] -->
<!-- selected <- rownames(bulk_yardstick[bulk_yardstick$FDR < 0.05 & abs(bulk_yardstick$logFC) >= 2,]) -->
<!-- selected <- head(selected, 10) -->
<!-- selected -->

<!-- ``` -->

<!-- One  -->


<!-- ```{r, fig.width =12, fig.height=5} -->

<!-- ## get cpms from the seurat object, only 5 cells (first 5 indexed) -->
<!-- foo <- list() -->
<!-- for (item in names(d)) { -->
<!--     ## foo <- list() -->
<!--     for (flavour in grep('repeats', names(d[[item]]), value = TRUE)) { -->
<!--         tmp <- as.data.frame(FetchData(d[[item]][[flavour]], -->
<!--                                        vars = selected, -->
<!--                                        slot = 'data'))[1:5,] -->
<!--         tmp$cell <- rownames(tmp) -->

        
<!--         tmp <- merge(d[[item]][[flavour]]@meta.data[, c('cell', 'scina')], -->
<!--                      tmp, by = 'cell', all.x = TRUE) -->
<!--         rownames(tmp) <- tmp$cell -->

<!--         tmp <- melt(tmp, id.vars = c('cell', 'scina')) -->
<!--         tmp <- tmp[grepl('T_cell|monocyte', tmp$scina),] -->
<!--         tmp$flavour <- flavour -->
<!--         tmp$dataset <-  item -->
<!--         foo[[paste(item, flavour)]] <- tmp -->
        
        
<!--     } -->
<!--     ## foo <- do.call(rbind.data.frame, foo) -->
<!-- } -->

<!-- foo <- do.call(rbind.data.frame, foo) -->
<!-- head(foo) -->
<!-- foo$scina <- as.character(foo$scina) -->
<!-- table(foo$scina) -->

<!-- print(ggplot(foo, -->
<!--              aes(x = variable, y = value, fill = flavour)) +  -->
<!--       geom_boxplot() + -->
<!--       scale_fill_discrete(drop=FALSE) + -->
<!--       facet_grid(dataset ~ scina) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->

<!-- ``` -->

# Timestamp

```{r sessionInfo, cache = FALSE}
date()
sessionInfo()
devtools::session_info()

```

```{r}
knitr::knit_exit()
```





<!-- # Most variable repeats cellranger {.tabset} -->

<!-- Similarly for repeats expression (avg +- sd of repeats CPM by gene expression cluster?) -->


<!-- ```{r, fig.height = 5, fig.width = 15, results = 'asis'} -->
<!-- res <- 'RNA_snn_res.0.2' -->

<!-- ## ## selected <- c('AluY', 'AluSx', 'MER49', 'MIR', 'L2a', 'L2b', 'L1PA11') -->
<!-- ## selected <- c(names(head(sort(apply(GetAssayData(d[[run]]$repeats_cellranger),1, coef_var), decreasing = TRUE), -->
<!-- ##                          5)), c('AluY', 'AluSx', 'L1PA11', 'MIR', 'L2a')) -->

<!-- ## maybe with variable features? -->

<!-- for (run in names(d)[sapply(d, function(x) return('repeats_cellranger' %in% names(x)))]) { -->
<!--     cat("## ", run, "\n\n") -->
<!--     repeats_cellranger <- as.data.frame(FetchData( -->
<!--         d[[run]]$repeats_cellranger, -->
<!--         vars = head(VariableFeatures(d[[run]]$repeats_cellranger), 10) , slot = 'data')) -->
        
<!--     repeats_cellranger$cell <- rownames(repeats_cellranger) -->
        
<!--     d[[run]]$genes_cellranger@meta.data$cell <- rownames(d[[run]]$genes_cellranger@meta.data) -->
<!--     repeats_cellranger <- merge(d[[run]]$genes_cellranger@meta.data[, c(res, 'cell') ], -->
<!--                      repeats_cellranger, by = 'cell', all.x = TRUE) -->
    
<!--     rownames(repeats_cellranger) <- repeats_cellranger$cell -->
    
<!--     repeats_cellranger <- melt(repeats_cellranger, id.vars = c('cell', res)) -->
    
<!--     colnames(repeats_cellranger) <- c('cell', res, 'repfamily', 'CPM') -->

<!--     print(ggplot(repeats_cellranger, -->
<!--            aes(x = get(res), y = CPM)) +  -->
<!--           geom_boxplot(outlier.size=1) + -->
<!--           scale_y_continuous(trans='log10') + -->
<!--         scale_fill_discrete(drop=FALSE) + -->
<!--         facet_wrap(~ factor(repfamily), ncol = 10) + -->
<!--         theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--         ylab('repeat expression [CPM]') + -->
<!--         xlab('gene-based clustering') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_cellranger', res))) -->
<!--     cat('\n\n') -->
<!-- } -->

<!-- ``` -->

<!-- # Most variable repeats cellranger by identity {.tabset} -->


<!-- ```{r, fig.height = 5, fig.width = 15, results = 'asis'} -->
<!-- scina <- 'scina' -->
<!-- ## agg <- list() -->

<!-- for (run in names(d)[sapply(d, function(x) return('repeats_cellranger' %in% names(x)))]) { -->
<!--     cat("## ", run, "\n\n") -->
    
<!--     repeats_cellranger <- as.data.frame(FetchData( -->
<!--         d[[run]]$repeats_cellranger, -->
<!--         vars = head(VariableFeatures(d[[run]]$repeats_cellranger), 10) , slot = 'data')) -->


    
<!--     repeats_cellranger$cell <- rownames(repeats_cellranger) -->

<!--     d[[run]]$genes_cellranger@meta.data$cell <- rownames(d[[run]]$genes_cellranger@meta.data) -->
<!--     repeats_cellranger <- merge(d[[run]]$genes_cellranger@meta.data[, c(scina, 'cell') ], -->
<!--                      repeats_cellranger, by = 'cell', all.x = TRUE) -->
    
<!--     rownames(repeats_cellranger) <- repeats_cellranger$cell -->
    
<!--     repeats_cellranger <- melt(repeats_cellranger, id.vars = c('cell', scina)) -->
    
<!--     colnames(repeats_cellranger) <- c('cell', scina, 'repfamily', 'CPM') -->

<!--     print(ggplot(repeats_cellranger, -->
<!--            aes(x = scina, y = CPM)) +  -->
<!--           geom_boxplot(outlier.size=1) + -->
<!--           scale_y_continuous(trans='log10') + -->
<!--         scale_fill_discrete(drop=FALSE) + -->
<!--         facet_wrap(~ factor(repfamily), ncol = 10) + -->
<!--         theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--         ylab('repeat expresssion [CPM]') + -->
<!--         xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_cellranger', scina))) -->
<!--     cat('\n\n') -->

<!-- } -->

<!-- ``` -->

<!-- # similarly for repeats alevin -->



<!-- ```{r, fig.height = 5, fig.width = 15, results = 'asis'} -->
<!-- scina <- 'scina' -->
<!-- ## agg <- list() -->

<!-- for (run in names(d)[sapply(d, function(x) return('repeats_alevin' %in% names(x)))]) { -->
<!--     cat("## ", run, "\n\n") -->
    
<!--     repeats_alevin <- as.data.frame(FetchData( -->
<!--         d[[run]]$repeats_alevin, -->
<!--         vars = head(VariableFeatures(d[[run]]$repeats_alevin), 10) , slot = 'data')) -->


    
<!--     repeats_alevin$cell <- rownames(repeats_alevin) -->

<!--     d[[run]]$genes_alevin@meta.data$cell <- rownames(d[[run]]$genes_alevin@meta.data) -->
<!--     repeats_alevin <- merge(d[[run]]$genes_alevin@meta.data[, c(scina, 'cell') ], -->
<!--                      repeats_alevin, by = 'cell', all.x = TRUE) -->
    
<!--     rownames(repeats_alevin) <- repeats_alevin$cell -->
    
<!--     repeats_alevin <- melt(repeats_alevin, id.vars = c('cell', scina)) -->
    
<!--     colnames(repeats_alevin) <- c('cell', scina, 'repfamily', 'CPM') -->

<!--     print(ggplot(repeats_alevin, -->
<!--            aes(x = scina, y = CPM)) +  -->
<!--           geom_boxplot(outlier.size=1) + -->
<!--           scale_y_continuous(trans='log10') + -->
<!--         scale_fill_discrete(drop=FALSE) + -->
<!--         facet_wrap(~ factor(repfamily), ncol = 10) + -->
<!--         theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--         ylab('repeat expresssion [CPM]') + -->
<!--         xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_alevin', scina))) -->
<!--     cat('\n\n') -->

<!-- } -->

<!-- ``` -->

<!-- # All integrated (repeats, scina) -->

<!-- Using alevin -->

<!-- ```{r fig.width = 14, fig.height = 14} -->

<!-- ## first aggregate all repeats_alevin stuff -->
<!-- ## and get the ones with highest variability -->
<!-- selected <- unique(unlist(lapply(d[sapply(d, function(x) return('repeats_alevin' %in% names(x)))], -->
<!--               function(x) return(head(VariableFeatures(x$repeats_alevin), 3))))) -->

<!-- length(selected) -->
<!-- ## aggregate(agg['CPM'], by = list(agg$repfamily, agg$run),  FUN = function(x) c(mean = mean(x, na.rm = TRUE), n = length(x), sd = sd(x, na.rm = TRUE) )) -->

<!-- ## now get these data for selected features, aggregate and plot -->

<!-- tmp <- list() -->
<!-- for (run in names(d)[sapply(d, function(x) return('repeats_alevin' %in% names(x)))]) { -->
<!--     foo <- data.frame(run = run, -->
<!--                              FetchData(d[[run]]$repeats_alevin, vars = selected, slot = 'data')) -->

<!--     ## add annotations -->
<!--     bar <- data.frame(cell = rownames(foo), -->
<!--                       scina = d[[run]]$repeats_alevin@meta.data[rownames(foo), 'scina']) -->

<!--     stopifnot(rownames(foo) == bar$cell) -->
    
    
<!--     tmp[[run]] <- data.frame(foo, scina = bar$scina) -->
<!--     rm(foo, bar) -->
<!-- } -->

<!-- tmp <- do.call(rbind.data.frame, tmp) -->
<!-- str(tmp) -->

<!-- ## melt -->
<!-- repeats_alevin <- melt(tmp, id.vars = c('run', 'scina')) -->
<!-- head(repeats_alevin)     -->
<!-- colnames(repeats_alevin) <- c('run', 'scina', 'repfamily', 'CPM') -->

<!-- ``` -->

<!-- ```{r, fig.width = 12, fig.height = 12} -->
<!-- print(ggplot(repeats_alevin, -->
<!--              aes(x = scina, y = CPM, fill = run)) +  -->
<!--       geom_boxplot(outlier.size=1) + -->
<!--       scale_y_continuous(trans='log10') + -->
<!--       scale_fill_discrete(drop=FALSE) + -->
<!--       facet_wrap(~ factor(repfamily), ncol = 4) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--       ylab('repeat expresssion [CPM]') + -->
<!--       xlab('scina identity')) -->
<!-- ``` -->


<!-- ```{r, fig.width = 8, fig.height = 14} -->
<!-- print(ggplot(repeats_alevin[repeats_alevin$scina %in% c('CD14+_monocyte', 'CD16+_monocyte', -->
<!--                                                   'CD4+_T_cell', -->
<!--                                                   'B_cell', -->
<!--                                                   'Cytotoxic_T_cell'),], -->
<!--              aes(x = run, y = CPM, fill = run)) +  -->
<!--       geom_boxplot(outlier.size=1) + -->
<!--       ## scale_y_continuous(trans='log10') + -->
<!--       scale_fill_discrete(drop=FALSE) + -->
<!--       facet_grid( factor(scina) ~ factor(repfamily)) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--       ylab('repeat expresssion [CPM]') + -->
<!--       xlab('run') + ggtitle(sprintf('%s\n%s', 'repeats_alevin', scina))) -->

<!-- ``` -->

<!-- # Cluster by average expression (repeats, SCINA) -->

<!-- Avoiding non-determined cell types -->

<!-- ```{r, results = 'asis', fig.height = 10, fig.width = 10} -->
<!-- repeats_alevin <- repeats_alevin[repeats_alevin$scina != 'unknown',] -->
<!-- foo <- reshape2::dcast(repeats_alevin, run + scina ~ repfamily, mean, value.var = 'CPM') -->


<!-- rownames(foo) <- paste(foo$scina, foo$run, sep = '_') -->

<!-- annot <- foo[,c('run', 'scina')] -->
<!-- rownames(annot) <- paste(annot$scina, annot$run, sep = '_') -->
<!-- foo <- foo[,-(1:2)] -->

<!-- pheatmap(foo, annotation_row = annot, show_rownames = FALSE) -->

<!-- pheatmap(t(foo), annotation_col = annot, show_colnames = FALSE) -->


<!-- pheatmap(foo[order(rownames(foo)),], annotation_row = annot, -->
<!--          show_rownames = FALSE, cluster_rows = FALSE, -->
<!--          cluster_cols = FALSE) -->


<!-- ``` -->



<!-- <\!-- here start -\-> -->


<!-- # Most variable repeats star {.tabset} -->

<!-- Similarly for repeats expression (avg +- sd of repeats CPM by gene expression cluster?) -->

<!-- What's the point here? -->


<!-- ```{r, fig.height = 5, fig.width = 15, results = 'asis'} -->
<!-- res <- 'RNA_snn_res.0.2' -->

<!-- ## ## selected <- c('AluY', 'AluSx', 'MER49', 'MIR', 'L2a', 'L2b', 'L1PA11') -->
<!-- ## selected <- c(names(head(sort(apply(GetAssayData(d[[run]]$repeats_star),1, coef_var), decreasing = TRUE), -->
<!-- ##                          5)), c('AluY', 'AluSx', 'L1PA11', 'MIR', 'L2a')) -->

<!-- ## maybe with variable features? -->

<!-- for (run in names(d)[sapply(d, function(x) return('repeats_star' %in% names(x)))]) { -->
<!--     cat("## ", run, "\n\n") -->
<!--     repeats_star <- as.data.frame(FetchData( -->
<!--         d[[run]]$repeats_star, -->
<!--         vars = head(VariableFeatures(d[[run]]$repeats_star), 10) , slot = 'data')) -->
        
<!--     repeats_star$cell <- rownames(repeats_star) -->
        
<!--     d[[run]]$genes@meta.data$cell <- rownames(d[[run]]$genes@meta.data) -->
<!--     repeats_star <- merge(d[[run]]$genes@meta.data[, c(res, 'cell') ], -->
<!--                      repeats_star, by = 'cell', all.x = TRUE) -->
    
<!--     rownames(repeats_star) <- repeats_star$cell -->
    
<!--     repeats_star <- melt(repeats_star, id.vars = c('cell', res)) -->
    
<!--     colnames(repeats_star) <- c('cell', res, 'repfamily', 'CPM') -->

<!--     print(ggplot(repeats_star, -->
<!--            aes(x = get(res), y = CPM)) +  -->
<!--           geom_boxplot(outlier.size=1) + -->
<!--           scale_y_continuous(trans='log10') + -->
<!--         scale_fill_discrete(drop=FALSE) + -->
<!--         facet_wrap(~ factor(repfamily), ncol = 10) + -->
<!--         theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--         ylab('repeat expression [CPM]') + -->
<!--         xlab('gene-based clustering') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_star', res))) -->
<!--     cat('\n\n') -->
<!-- } -->

<!-- ``` -->

<!-- # Most variable repeats star by identity {.tabset} -->

<!-- Repeats star unique -->

<!-- ```{r, fig.height = 5, fig.width = 15, results = 'asis'} -->
<!-- scina <- 'scina' -->
<!-- ## agg <- list() -->

<!-- for (run in names(d)[sapply(d, function(x) return('repeats_star_unique' %in% names(x)))]) { -->
<!--     cat("## ", run, "\n\n") -->

<!--     repeats_star_unique <- as.data.frame(FetchData( -->
<!--         d[[run]]$repeats_star_unique, -->
<!--         vars = head(VariableFeatures(d[[run]]$repeats_star_unique), 10) , slot = 'data')) -->


    
<!--     repeats_star_unique$cell <- rownames(repeats_star_unique) -->

<!--     ## d[[run]]$genes_alevin@meta.data$cell <- rownames(d[[run]]$genes_alevin@meta.data) -->
<!--     repeats_star_unique <- merge(d[[run]]$repeats_star_unique@meta.data[, c(scina, 'cell') ], -->
<!--                      repeats_star_unique, by = 'cell', all.x = TRUE) -->
    
<!--     rownames(repeats_star_unique) <- repeats_star_unique$cell -->
    
<!--     repeats_star_unique <- melt(repeats_star_unique, id.vars = c('cell', scina)) -->
    
<!--     colnames(repeats_star_unique) <- c('cell', scina, 'repfamily', 'CPM') -->

<!--     print(ggplot(repeats_star_unique, -->
<!--            aes(x = scina, y = CPM)) +  -->
<!--           geom_boxplot(outlier.size=1) + -->
<!--           scale_y_continuous(trans='log10') + -->
<!--         scale_fill_discrete(drop=FALSE) + -->
<!--         facet_wrap(~ factor(repfamily), ncol = 10) + -->
<!--         theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--         ylab('repeat expresssion [CPM]') + -->
<!--         xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_star_unique', scina))) -->
<!--     cat('\n\n') -->

<!-- } -->

<!-- ``` -->

<!-- # All integrated (repeats, scina) -->

<!-- ```{r fig.width = 14, fig.height = 14} -->

<!-- ## first aggregate all repeats_star stuff -->
<!-- ## get the ones with the highest var -->
<!-- selected <- unique(unlist( -->
<!--     lapply(d[sapply(d, -->
<!--                     function(x) return('repeats_star_unique' %in% names(x)))], -->
<!--            function(x) return(head(VariableFeatures(x$repeats_star_unique), 3))))) -->


<!-- ## aggregate(agg['CPM'], by = list(agg$repfamily, agg$run),  FUN = function(x) c(mean = mean(x, na.rm = TRUE), n = length(x), sd = sd(x, na.rm = TRUE) )) -->

<!-- ## now get these data for selected features, aggregate and plot -->
<!-- ## this crashes because of the repeat counting/aggregation, that edited repeat names -->
<!-- ## @todo fix (doesn't find `SAR`) -->
<!-- tmp <- list() -->
<!-- for (run in  names(d)[sapply(d, function(x) return('repeats_star_unique' %in% names(x)))]) { -->
<!--     foo <- data.frame(run = run, -->
<!--                       FetchData(d[[run]]$repeats_star_unique, vars = selected, -->
<!--                                 slot = 'data')) -->

<!--     ## add annotations -->
<!--     bar <- data.frame(cell = rownames(foo), -->
<!--                       scina = d[[run]]$repeats_star_unique@meta.data[rownames(foo), 'scina']) -->

<!--     stopifnot(rownames(foo) == bar$cell) -->
    
    
<!--     tmp[[run]] <- data.frame(foo, scina = bar$scina) -->
<!--     rm(foo, bar) -->
<!-- } -->

<!-- str(tmp) -->
<!-- tmp <- do.call(rbind.data.frame, tmp) -->
<!-- str(tmp) -->

<!-- ## melt -->
<!-- repeats_star_unique <- melt(tmp, id.vars = c('run', 'scina')) -->

<!-- colnames(repeats_star_unique) <- c('run', 'scina', 'repfamily', 'CPM') -->


<!-- ``` -->

<!-- ```{r, fig.width = 12, fig.height = 12} -->
<!-- print(ggplot(repeats_star_unique, -->
<!--              aes(x = scina, y = CPM, fill = run)) +  -->
<!--       geom_boxplot(outlier.size=1) + -->
<!--       scale_y_continuous(trans='log10') + -->
<!--       scale_fill_discrete(drop=FALSE) + -->
<!--       facet_wrap(~ factor(repfamily), ncol = 4) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--       ylab('repeat expresssion [CPM]') + -->
<!--       xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_star_unique', scina))) -->
<!-- ``` -->


<!-- ```{r, fig.width = 8, fig.height = 14} -->
<!-- print(ggplot(repeats_star, -->
<!--              aes(x = repfamily, y = CPM, fill = run)) +  -->
<!--       geom_boxplot(outlier.size=1) + -->
<!--       scale_y_continuous(trans='log10') + -->
<!--       scale_fill_discrete(drop=FALSE) + -->
<!--       facet_grid(factor(scina)~ factor(run)) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--       ylab('repeat expresssion [CPM]') + -->
<!--       xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_star', scina))) -->

<!-- ``` -->

<!-- # Cluster by average expression (repeats, SCINA) -->

<!-- Avoiding non-determined cell types -->

<!-- ```{r, results = 'asis', fig.height = 10, fig.width = 10} -->
<!-- repeats_star <- repeats_star[repeats_star$scina != 'unknown',] -->
<!-- foo <- reshape2::dcast(repeats_star, run + scina ~ repfamily, mean, value.var = 'CPM') -->


<!-- rownames(foo) <- paste(foo$run, foo$scina, sep = '_') -->

<!-- annot <- foo[,c('run', 'scina')] -->
<!-- rownames(annot) <- paste(annot$run, annot$scina, sep = '_') -->
<!-- foo <- foo[,-(1:2)] -->

<!-- pheatmap(foo, annotation_row = annot, show_rownames = FALSE) -->

<!-- pheatmap(t(foo), annotation_col = annot, show_colnames = FALSE) -->
<!-- ``` -->

<!-- # Pairwise correlation per identity? -->

<!-- ```{r} -->

<!-- library(MatrixCorrelation) -->

<!-- X1 <- scale( matrix( rnorm(100*300), 100,300), scale = FALSE) -->
<!-- usv <- svd(X1) -->
<!-- X2 <- usv$u[,-3] %*% diag(usv$d[-3]) %*% t(usv$v[,-3]) -->
<!-- smi <- SMI(X1,X2,5,5) -->
<!-- plot(smi, B = 1000) # default B = 10000 -->
<!-- print(smi) -->
<!-- summary(smi) -->
<!-- is.signif(smi, B = 1000) # default B = 10000 -->


<!-- for (run in  names(d)) { -->

<!--     repeats <- grep('repeat', names(d[[run]])) -->
    
<!--     smi <- SMI(d[[run]][[repeats[1]]], d[[run]][[repeats[2]]], 5, 5) -->

<!--     foo <- data.frame(run = run, -->
<!--                       FetchData(d[[run]]$repeats_star_unique, vars = selected, -->
<!--                                 slot = 'data')) -->

<!--     ## add annotations -->
<!--     bar <- data.frame(cell = rownames(foo), -->
<!--                       scina = d[[run]]$repeats_star_unique@meta.data[rownames(foo), 'scina']) -->

<!--     stopifnot(rownames(foo) == bar$cell) -->
    
    
<!--     tmp[[run]] <- data.frame(foo, scina = bar$scina) -->
<!--     rm(foo, bar) -->
<!-- } -->

<!-- str(tmp) -->
<!-- tmp <- do.call(rbind.data.frame, tmp) -->
<!-- str(tmp) -->



<!-- ``` -->

<!-- <\!-- here stop -\-> -->

<!-- # All integrated (repeats, scina) -->

<!-- ```{r fig.width = 14, fig.height = 14} -->

<!-- ## first aggregate all repeats_featurecounts stuff -->
<!-- sds <- lapply(d[sapply(d, function(x) return('repeats_featurecounts' %in% names(x)))], -->
<!--               function(x) return(apply(as.data.frame(GetAssayData(x$repeats_featurecounts, -->
<!--                                                         slot = 'data')), 1, sd))) -->

<!-- ## get the ones with the highest average sd -->

<!-- mean_sd <- apply(do.call(cbind.data.frame, sds), 1, mean) -->
<!-- selected <- names(head(sort(mean_sd, decreasing = TRUE), 20)) -->
<!-- ## selected -->


<!-- ## aggregate(agg['CPM'], by = list(agg$repfamily, agg$run),  FUN = function(x) c(mean = mean(x, na.rm = TRUE), n = length(x), sd = sd(x, na.rm = TRUE) )) -->

<!-- ## now get these data for selected features, aggregate and plot -->

<!-- tmp <- list() -->
<!-- for (run in names(d)) { -->
<!--     foo <- data.frame(run = run, -->
<!--                              FetchData(d[[run]]$repeats_featurecounts, vars = selected, slot = 'data')) -->

<!--     ## add annotations -->
<!--     bar <- data.frame(cell = rownames(foo), -->
<!--                       scina = d[[run]]$repeats_featurecounts@meta.data[rownames(foo), 'scina']) -->

<!--     stopifnot(rownames(foo) == bar$cell) -->
    
    
<!--     tmp[[run]] <- data.frame(foo, scina = bar$scina) -->
<!--     rm(foo, bar) -->
<!-- } -->

<!-- tmp <- do.call(rbind.data.frame, tmp) -->
<!-- str(tmp) -->

<!-- ## melt -->
<!-- repeats_featurecounts <- melt(tmp, id.vars = c('run', 'scina')) -->
<!-- head(repeats_featurecounts)     -->
<!-- colnames(repeats_featurecounts) <- c('run', 'scina', 'repfamily', 'CPM') -->

<!-- ``` -->

<!-- ```{r, fig.width = 12, fig.height = 12} -->
<!-- print(ggplot(repeats_featurecounts, -->
<!--              aes(x = scina, y = CPM, fill = run)) +  -->
<!--       geom_boxplot(outlier.size=1) + -->
<!--       scale_y_continuous(trans='log10') + -->
<!--       scale_fill_discrete(drop=FALSE) + -->
<!--       facet_wrap(~ factor(repfamily), ncol = 4) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--       ylab('repeat expresssion [CPM]') + -->
<!--       xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_featurecounts', scina))) -->
<!-- ``` -->


<!-- ```{r, fig.width = 8, fig.height = 14} -->
<!-- print(ggplot(repeats_featurecounts, -->
<!--              aes(x = repfamily, y = CPM, fill = run)) +  -->
<!--       geom_boxplot(outlier.size=1) + -->
<!--       scale_y_continuous(trans='log10') + -->
<!--       scale_fill_discrete(drop=FALSE) + -->
<!--       facet_grid(factor(scina)~ factor(run)) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--       ylab('repeat expresssion [CPM]') + -->
<!--       xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_featurecounts', scina))) -->

<!-- ``` -->

<!-- # Cluster by average expression (repeats, SCINA) -->

<!-- Avoiding non-determined cell types -->

<!-- ```{r, results = 'asis', fig.height = 10, fig.width = 10} -->
<!-- repeats_featurecounts <- repeats_featurecounts[repeats_featurecounts$scina != 'unknown',] -->
<!-- foo <- reshape2::dcast(repeats_featurecounts, run + scina ~ repfamily, mean, value.var = 'CPM') -->


<!-- rownames(foo) <- paste(foo$run, foo$scina, sep = '_') -->

<!-- annot <- foo[,c('run', 'scina')] -->
<!-- rownames(annot) <- paste(annot$run, annot$scina, sep = '_') -->
<!-- foo <- foo[,-(1:2)] -->

<!-- pheatmap(foo, annotation_row = annot, show_rownames = FALSE) -->

<!-- pheatmap(t(foo), annotation_col = annot, show_colnames = FALSE) -->
<!-- ``` -->




<!-- # Pheatmaps -->

<!-- Note the cell sampling: same cells in all plots; `annotation_col` comes from genes' clustering. -->

<!-- ```{r, results = 'asis', dpi=50, fig.width = 15, fig.height = 15} -->
<!-- NCELLS <- 1000 -->
<!-- NFEATURES <- 500 -->

<!-- ## stop() -->
<!-- for (run in names(d)) { -->
<!--     cat("## ", run, "{.tabset}\n\n") -->
    
<!--     set.seed(654) -->
<!--     idx_cell <- sample(colnames(GetAssayData(d[[run]][[1]])), NCELLS, replace = FALSE) -->
    
<!--     for (item in names(d[[run]])) { -->
<!--         cat("### ", item, "{.tabset}\n\n") -->

<!--         set.seed(464) -->
<!--         idx_feat <- sample(VariableFeatures(d[[run]][[item]]), NFEATURES, replace = FALSE) -->
        
<!--         print(pheatmap(GetAssayData( -->
<!--             object = subset(d[[run]][[item]], -->
<!--                             features = idx_feat, -->
<!--                             cells = intersect(idx_cell, colnames(GetAssayData(d[[run]][[item]])))), -->
<!--             slot = 'scale.data'), -->
<!--             main = sprintf('%s %s', run, item), -->
<!--             color = viridis(100), -->
<!--             cluster_rows = TRUE, -->
<!--             cluster_cols = TRUE, -->
<!--             show_rownames = FALSE, -->
<!--             show_colnames = FALSE, -->
<!--             clustering_distance_cols = "euclidean", -->
<!--             clustering_method = "ward.D2", -->
<!--             clustering_distance_rows = "euclidean", -->
<!--             fontsize_row = 12, -->
<!--             annotation_col = as.data.frame(d[[run]]$genes_cellranger@meta.data[idx_cell, c(res), -->
<!--                                                                                drop = FALSE]), -->
<!--             scale = "none")) -->
<!--         cat('\n\n') -->
<!--     } -->
<!--     cat('\n\n') -->
<!-- } -->
<!-- ``` -->




<!-- # ARIS -->

<!-- ## Data load -->

<!-- ```{r} -->

<!-- aris_df <- NULL -->

<!-- for (rds in params) { -->
<!--     ## the params are RDS objects from Seurat -->
<!--     ## aris[[sapply(strsplit(dirname(rds), .Platform$file.sep), -->
<!--     ##       function(x) return(x[6]))]] <- list() -->
<!--     fns <- list.files(dirname(rds), pattern = 'aris.*rds') -->
<!--     for (fn in fns) { -->
<!--         curr_nest <- readRDS(file.path(dirname(rds), fn)) -->
<!--         ## curr <- do.call(function(x) rbind.data.frame(x, setNames = c('cat1' , 'cat2', 'ari')), -->
<!--         ##                 curr_nest) -->
<!--         ## aris_df <- NULL -->
<!--         for (item in names(curr_nest)) { -->
<!--             curr <- curr_nest[[item]] -->
<!--             colnames(curr) <- c('set1', 'set2', 'ari') -->
<!--             curr$origin <- item -->

<!--             parsed <- strcapture( -->
<!--                 pattern = "aris_regress_nCount_(.*)_nFeature_(.*).rds", -->
<!--                 x = fn, -->
<!--                 proto = list(ncount = logical(), nfeature = logical())) -->
            
<!--             curr$ncount_scaling <- parsed$ncount -->
<!--             curr$nfeature_scaling <- parsed$nfeature -->
            
<!--             curr$sample <- sapply(strsplit(dirname(rds), .Platform$file.sep), -->
<!--                                   function(x) return(x[6])) -->
<!--             aris_df <- rbind(aris_df, curr) -->

<!--         } -->

<!--         ## aris[[sapply(strsplit(dirname(rds), .Platform$file.sep), -->
<!--         ##              function(x) return(x[6]))]][[basename(fn)]] <- curr_nest -->
<!--     } -->
<!-- } -->

<!-- summary(aris_df) -->
<!-- rm(parsed, curr) -->
<!-- ``` -->

<!-- Extract resolutions and n for clustering -->

<!-- ```{r notscina, eval = TRUE} -->
<!-- parsed <- strcapture( -->
<!--     pattern = "(.*)_RNA_snn_res.(.*) \\(n=(.*)\\)", -->
<!--     x = aris_df$set1, -->
<!--     proto = list(id1 = character(), -->
<!--                  res1 = numeric(), -->
<!--                  size1 = numeric())) -->

<!-- aris_df <- cbind(aris_df, parsed) -->
<!-- rm(parsed) -->

<!-- parsed <- strcapture( -->
<!--     pattern = "(.*)_RNA_snn_res.(.*) \\(n=(.*)\\)", -->
<!--     x = aris_df$set2, -->
<!--     proto = list(id2 = character(), -->
<!--                  res2 = numeric(), -->
<!--                  size2 = numeric())) -->

<!-- aris_df <- cbind(aris_df, parsed) -->
<!-- rm(parsed) -->

<!-- ## head(aris_df) -->
<!-- ``` -->

<!-- Extract resolutions and n for scina -->

<!-- ```{r eval = TRUE} -->
<!-- parsed <- strcapture( -->
<!--     pattern = "scina_scina \\(n=(.*)\\)", -->
<!--     x = aris_df$set2, -->
<!--     proto = list(size_scina = numeric())) -->

<!-- aris_df <- cbind(aris_df, parsed) -->
<!-- rm(parsed) -->


<!-- ## head(aris_df) -->
<!-- ``` -->



<!-- ## ARIs comparison {.tabset} -->

<!-- Plot ARIs decay with resolution, scaling etc -->

<!-- ```{r, fig.width=12, fig.height=8, results = 'asis'} -->

<!-- aris_df$regressed_out <- 'none' -->

<!-- aris_df$regressed_out[aris_df$ncount_scaling & aris_df$nfeature_scaling] <- 'both' -->
<!-- aris_df$regressed_out[aris_df$ncount_scaling & !aris_df$nfeature_scaling] <- 'genes_ncount' -->
<!-- aris_df$regressed_out[!aris_df$ncount_scaling & aris_df$nfeature_scaling] <- 'genes_nfeature' -->

<!-- aris_df$regressed_out <- factor(aris_df$regressed_out, -->
<!--                                    levels = c('none', 'genes_ncount', 'genes_nfeature', 'both')) -->

<!-- ``` -->

<!-- ```{r, fig.width=12, fig.height=8, results = 'asis'} -->

<!-- aris_df$label1 <- sprintf('res%s_(n=%s)', aris_df$res1, aris_df$size1) -->
<!-- aris_df$label1[grepl('NA', aris_df$label1)] <- NA -->

<!-- aris_df$label2 <- sprintf('res%s_(n=%s)', aris_df$res2, aris_df$size2) -->
<!-- aris_df$label2[grepl('NA', aris_df$label2)] <- NA -->


<!-- ## cat('### genes vs repeats cr\n\n') -->

<!-- ## print(ggplot(aris_df[aris_df$id1 == 'genes' & aris_df$id2 == 'repeats_cr',], -->
<!-- ##              aes(x = res1, y = ari, col = as.factor(regressed_out))) + -->
<!-- ##       geom_point() + -->
<!-- ##       labs(title = "genes cellranger vs repeats cellranger (from repeats bam)", -->
<!-- ##            subtitle = "by repeats resolution", y = "ARI", x = "genes resolution", -->
<!-- ##            color = 'regressed out') + -->
<!-- ##       facet_grid(sample~paste0('repeats res.\n', res2)) + -->
<!-- ##       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->

<!-- ## cat('\n\n') -->

<!-- cat('### genes vs repeats fc\n\n') -->

<!-- print(ggplot(aris_df[aris_df$id1 == 'genes' & aris_df$id2 == 'repeats_fc',], -->
<!--              aes(x = res1, y = ari, col = as.factor(regressed_out))) + -->
<!--       geom_point() + -->
<!--       labs(title = "genes cellranger vs repeats featurecounts (from genes bam)", -->
<!--            subtitle = "by repeats resolution", y = "ARI", x = "genes resolution", -->
<!--            color = 'regressed out') + -->
<!--       facet_grid(sample~paste0('repeats res.\n', res2)) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->

<!-- cat('\n\n') -->

<!-- cat('### repeats fc vs repeats cr\n\n') -->

<!-- print(ggplot(aris_df[aris_df$id1 == 'repeats_fc' & aris_df$id2 == 'repeats_cr',], -->
<!--              aes(x = res1, y = ari, col = as.factor(regressed_out))) + -->
<!--       geom_point() + -->
<!--       labs(title = "repeats featurecounts (from genes bam) vs repeats cellranger (repeats bam)", -->
<!--            subtitle = "by repeats cellranger resolution", y = "ARI", x = "repeats fc resolution", -->
<!--            color = 'regressed out') + -->
<!--       facet_grid(sample~paste0('repeats cr res.\n', res2)) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->

<!-- cat('\n\n') -->
<!-- ``` -->


<!-- Maybe add the number of clusters? as a size param? -->


<!-- Rather, plot both size and res -->

<!-- ```{r, eval = TRUE, fig.height = 5, fig.with = 15} -->
<!-- ## print(ggplot(aris_df[aris_df$id1 == 'repeats_fc' & aris_df$id2 == 'repeats_cr',], -->
<!-- ##              aes(x = res1, y = ari, col = as.factor(regressed_out))) + -->
<!-- ##       geom_point() + -->
<!-- ##       labs(title = "repeats featurecounts (from genes bam) vs repeats cellranger (repeats bam)", -->
<!-- ##            subtitle = "by repeats cellranger resolution", y = "ARI", x = "repeats fc resolution", -->
<!-- ##            color = 'regressed out') + -->
<!-- ##       facet_grid(sample~paste0('repeats cr res.\n', res2)) + -->
<!-- ##       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->

<!-- print(ggplot(aris_df, -->
<!--              aes(x = res1, y = size1, col = as.factor(regressed_out))) + -->
<!--       geom_point() + -->
<!--       facet_grid(sample~id1) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->

<!-- print(ggplot(aris_df, -->
<!--              aes(x = res2, y = size2, col = as.factor(regressed_out))) + -->
<!--       geom_point() + -->
<!--       facet_grid(sample~id2) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->
<!-- ``` -->

<!-- # ARIs flavours vs SCINA -->


<!-- ```{r, fig.height = 5, fig.with = 15} -->

<!-- print(ggplot(aris_df[grep('scina', aris_df$origin),], -->
<!--              aes(x = res1, y = ari, col = as.factor(regressed_out) )) + -->
<!--       geom_point() + -->
<!--       facet_grid(sample~id1) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->

<!-- ``` -->


<!-- ```{r} -->

<!-- print(ggplot(aris_df[grep('scina', aris_df$origin),], -->
<!--              aes(x = res1, y = size1, col = as.factor(regressed_out) )) + -->
<!--       geom_point() + -->
<!--       facet_grid(sample~id1) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->

<!-- ``` -->


<!-- ```{r} -->
<!-- print(ggplot(aris_df[grep('scina', aris_df$origin),], -->
<!--              aes(x = res1, y = ari, col = as.factor(regressed_out) )) + -->
<!--       geom_point() + -->
<!--       facet_grid(sample~id1) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->

<!-- ``` -->

<!-- # All variable features and all repeats methods at once, using scina {.tabset} -->

<!-- ## With repeats including control not overlapping genes -->

<!-- ```{r} -->

<!-- NCELLS = 200 ## cells per run:sample -->
<!-- NVARIABLE = 20 ## variable repeats features (per run:sample) -->

<!-- cr <- list() -->
<!-- for (flavour in c('repeats_cellranger', 'repeats_star', 'repeats_featurecounts', -->
<!--                   'control_repeats_only')) { -->
<!--     for (run in names(d)[sapply(d, function(x) return(flavour %in% names(x)))]) { -->

<!--         cr[[paste(flavour, run)]] <- data.frame( -->
<!--             rep = head(VariableFeatures(d[[run]][[flavour]]), NVARIABLE), -->
<!--             flavour = flavour, -->
<!--             run = run, -->
<!--             stringsAsFactors = FALSE) -->
<!--     }     -->
<!-- } -->

<!-- cr <- do.call(rbind.data.frame, cr) -->
<!-- table(cr$flavour, cr$run) -->

<!-- variable <- list() -->
<!-- for (flavour in c('repeats_cellranger', 'repeats_star', 'repeats_featurecounts', 'control_repeats_only')) { -->
<!--     for (run in names(d)[sapply(d, function(x) return(flavour %in% names(x)))]) { -->
<!--         idx <- sample(colnames(d[[run]][[flavour]]), size = NCELLS, replace = FALSE) -->
<!--         variable[[paste(run, flavour, sep = '.')]] <- GetAssayData(d[[run]][[flavour]], -->
<!--                             slot = 'data')[unique(cr$rep),idx] -->
<!--     } -->
<!-- } -->

<!-- foo <- do.call(cbind.data.frame, variable) -->

<!-- colannot <- data.frame(id = colnames(foo), -->
<!--                        sample = sapply(strsplit(colnames(foo), '.', fixed = TRUE), -->
<!--                                         function(x) return(x[1])), -->
<!--                        flavour = sapply(strsplit(colnames(foo), '.', fixed = TRUE), -->
<!--                                        function(x) return(x[2])), -->
<!--                        cell = sapply(strsplit(colnames(foo), '.', fixed = TRUE), -->
<!--                                      function(x) return(x[3])), -->
<!--                        scina = NA, -->
<!--                        stringsAsFactors = FALSE) -->


<!-- head(colannot) -->

<!-- for ( i in 1:nrow(colannot)) { -->
<!--     colannot[i, 'scina'] <- as.character(d[[colannot[i, 'sample']]][[colannot[i, 'flavour']]]@meta.data[colannot[i, 'cell'], 'scina']) -->

    
<!-- } -->

<!-- rownames(colannot) <- colannot$id -->

<!-- table(colannot$scina) -->


<!-- ## remove any repeat whose annotation by scina is 'unknown' -->
<!-- colannot <- colannot[colannot$scina != 'unknown',] -->
<!-- foo <- foo[,colannot$id] -->
<!-- dim(foo) -->

<!-- ## remove any repeat with 0 sd after sampling -->
<!-- sum(is.infinite(unlist(foo))) -->
<!-- foo <- foo[apply(foo, 1, sd) > 0,] -->

<!-- ``` -->

<!-- ```{r, fig.height = 20, fig.width = 20} -->


<!-- newCols <- colorRampPalette(grDevices::rainbow(length(unique(colannot$scina)))) -->
<!-- mycolors <- newCols(length(unique(colannot$scina))) -->
<!-- names(mycolors) <- unique(colannot$scina) -->
<!-- mycolors <- list(scina = mycolors) -->

<!-- print(pheatmap(foo, -->
<!--                color = viridis(100), -->
<!--             cluster_rows = TRUE, -->
<!--             cluster_cols = TRUE, -->
<!--             show_rownames = TRUE, -->
<!--             show_colnames = FALSE, -->
<!--             clustering_distance_cols = "euclidean", -->
<!--             clustering_method = "ward.D2", -->
<!--             clustering_distance_rows = "euclidean", -->
<!--             fontsize_row = 12, -->
<!--             annotation_col = colannot[,c('flavour', 'sample', 'scina')], -->
<!--             annotation_colors = mycolors, -->
<!--             scale = "none")) -->



<!-- ``` -->

<!-- ## Without non-overlapping genes repeats control (biased by transcriptome) -->

<!-- ```{r} -->

<!-- NCELLS = 200 ## cells per run:sample -->
<!-- NVARIABLE = 20 ## variable repeats features (per run:sample) -->

<!-- cr <- list() -->
<!-- for (flavour in c('repeats_cellranger', 'repeats_star', 'repeats_featurecounts')) { -->
<!--     for (run in names(d)[sapply(d, function(x) return(flavour %in% names(x)))]) { -->

<!--         cr[[paste(flavour, run)]] <- data.frame( -->
<!--             rep = head(VariableFeatures(d[[run]][[flavour]]), NVARIABLE), -->
<!--             flavour = flavour, -->
<!--             run = run, -->
<!--             stringsAsFactors = FALSE) -->
<!--     }     -->
<!-- } -->

<!-- cr <- do.call(rbind.data.frame, cr) -->
<!-- table(cr$flavour, cr$run) -->

<!-- variable <- list() -->
<!-- for (flavour in c('repeats_cellranger', 'repeats_star', 'repeats_featurecounts')) { -->
<!--     for (run in names(d)[sapply(d, function(x) return(flavour %in% names(x)))]) { -->
<!--         idx <- sample(colnames(d[[run]][[flavour]]), size = NCELLS, replace = FALSE) -->
<!--         variable[[paste(run, flavour, sep = '.')]] <- GetAssayData(d[[run]][[flavour]], -->
<!--                             slot = 'data')[unique(cr$rep),idx] -->
<!--     } -->
<!-- } -->

<!-- foo <- do.call(cbind.data.frame, variable) -->

<!-- colannot <- data.frame(id = colnames(foo), -->
<!--                        sample = sapply(strsplit(colnames(foo), '.', fixed = TRUE), -->
<!--                                         function(x) return(x[1])), -->
<!--                        flavour = sapply(strsplit(colnames(foo), '.', fixed = TRUE), -->
<!--                                        function(x) return(x[2])), -->
<!--                        cell = sapply(strsplit(colnames(foo), '.', fixed = TRUE), -->
<!--                                      function(x) return(x[3])), -->
<!--                        scina = NA, -->
<!--                        stringsAsFactors = FALSE) -->


<!-- head(colannot) -->

<!-- for ( i in 1:nrow(colannot)) { -->
<!--     colannot[i, 'scina'] <- as.character(d[[colannot[i, 'sample']]][[colannot[i, 'flavour']]]@meta.data[colannot[i, 'cell'], 'scina']) -->

    
<!-- } -->

<!-- rownames(colannot) <- colannot$id -->

<!-- table(colannot$scina) -->


<!-- ## remove any repeat whose annotation by scina is 'unknown' -->
<!-- colannot <- colannot[colannot$scina != 'unknown',] -->
<!-- foo <- foo[,colannot$id] -->
<!-- dim(foo) -->

<!-- ## remove any repeat with 0 sd after sampling -->
<!-- sum(is.infinite(unlist(foo))) -->
<!-- foo <- foo[apply(foo, 1, sd) > 0,] -->

<!-- ``` -->

<!-- ```{r, fig.height = 20, fig.width = 20} -->


<!-- newCols <- colorRampPalette(grDevices::rainbow(length(unique(colannot$scina)))) -->
<!-- mycolors <- newCols(length(unique(colannot$scina))) -->
<!-- names(mycolors) <- unique(colannot$scina) -->
<!-- mycolors <- list(scina = mycolors) -->

<!-- print(pheatmap(foo, -->
<!--                color = viridis(100), -->
<!--             cluster_rows = TRUE, -->
<!--             cluster_cols = TRUE, -->
<!--             show_rownames = TRUE, -->
<!--             show_colnames = FALSE, -->
<!--             clustering_distance_cols = "euclidean", -->
<!--             clustering_method = "ward.D2", -->
<!--             clustering_distance_rows = "euclidean", -->
<!--             fontsize_row = 12, -->
<!--             annotation_col = colannot[,c('flavour', 'sample', 'scina')], -->
<!--             annotation_colors = mycolors, -->
<!--             scale = "none")) -->



<!-- ``` -->

