---
title: "Across samples report - Kumar"
author: "Izaskun Mallona at Robinsonlab"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    code_download: true
    number_sections: true
    df_print: kable
    theme: lumen
    always_allow_html: yes
params:
  results_path: ""
  markers_rds: ""
  seurat_rds: ""
---


Rendered for samples:

```{r}
## print(params)
```
# Config


```{r}
suppressPackageStartupMessages({
library(Matrix)

library(knitr)
library(Seurat)
library(dplyr)
library(scater)
library(scran)
library(data.table)
## library(SingleCellExperiment)
## library(scran)
## library(org.Mm.eg.db)
## suppressPackageStartupMessages(library(dplyr))
## suppressPackageStartupMessages(library(DT))
library(pheatmap)
## library(Cairo)
## library(RColorBrewer)
## library(corrplot)
library(ggplot2)
library(reshape2)
## library(xtable)
## library(mclust)
library(viridis)
library(UpSetR)
library(ggplot2)
library(grid)
library(GEOquery)
})
```

```{r}
BASE <- file.path('/home', 'imallona', 'repeats_sc')
```

```{r}
opts_chunk$set(fig.width = 5,
               fig.height = 5,
               dev = 'png',
               cache = FALSE,
               echo = TRUE,
               include = TRUE,
               cache.lazy = FALSE,
               warning = TRUE,
               error = TRUE,
               message = FALSE)



options(bitmapType='cairo')
## getOption('bitmapType')
```

```{r}
theme_set(theme_classic()) 

## theme_update(panel.background = element_blank(),
##              ## panel.border = element_blank(),
##              panel.grid.major = element_blank(),
##              panel.grid.minor = element_blank(),
##              axis.line = element_line(colour = "black"))

theme_update(panel.spacing = unit(.05, "lines"),
             panel.border = element_rect(color = "gray20", fill = NA, size = 0.1), 
             strip.background = element_rect(color = "gray20", size = 0.1),
             text = element_text(size = 15))

```


```{r functions}
## add alpha
ac <- function(col, alpha=1){
  apply(sapply(col, col2rgb)/255, 2, 
                     function(x) 
                       rgb(x[1], x[2], x[3], alpha=alpha))  
}

## load_standard_cellranger_genes <- function(run_name) {
##     ## barcodes_fn <- params$genes_cellranger ## should end with `barcodes.tsv.gz`
##     barcodes_fn <- file.path(BASE, 'runs', run_name, 'cellranger_standard', 'outs',
##                              'filtered_feature_bc_matrix', 'barcodes.tsv.gz')
    
##     path <- dirname(barcodes_fn)

##     genes <- CreateSeuratObject(counts = Read10X(data.dir = path),
##                                 project = 'cellranger_standard',
##                                 min.cells = 0,
##                                 min.features = 0)
##     return(genes)
## }

## load_cellranger_repeats <-  function(run_name) {
##     barcodes_fn <- file.path(BASE, 'runs', run_name, 'cellranger_repeats', 'outs',
##                              'filtered_feature_bc_matrix', 'barcodes.tsv.gz')

##     path <- dirname(barcodes_fn)

##     repeats <- CreateSeuratObject(counts = Read10X(data.dir = path),
##                                 project = 'cellranger_repeats',
##                                 min.cells = 0,
##                                 min.features = 0)
##     return(repeats)
## }

load_featurecounts_repeats <-  function(repeats_fn, id) {
    ## repeats <- file.path(BASE, "runs", run_name, 'count_repeats_on_cellranger_standard',
    ##                             paste0(run_name, "_repeats.counts.gz"))

    tmp <- data.table::fread(repeats_fn, sep = '\t', header = TRUE,
                             drop = c('Chr', 'Start', 'End', 'Strand', 'Length'))
    rownames(tmp) <- tmp$Geneid ## here

    colnames(tmp) <- gsub(".*:([A-Z]+)-1.bam","\\1", colnames(tmp))

    tmp <- as.data.frame(tmp)
    rownames(tmp) <- tmp$Geneid
    tmp <- tmp[,-1]

    featurecounts <- CreateSeuratObject(
        tmp,
        project = id,
        assay = "RNA",
        min.cells = 0,
        min.features = 0,
        names.field = 1,
        names.delim = "_",
        meta.data = NULL)

    return(featurecounts)
}

coef_var <- function(x){
        (sd(x)/mean(x))*100
}


```


# Load metadata

<!-- ```{r} -->
<!-- gse <- getGEO('GSE121861') -->
<!-- names(gse) -->
<!-- as.data.frame(gse[[1]]@phenoData) -->
<!-- str(pData(phenoData(gse[[1]]))) -->

<!-- str(show(pData(phenoData(gse[[1]])))) -->
<!-- ``` -->

```{r}
annot <- read.table(text = 'Run,Assay Type,AvgSpotLen,Bases,BioProject,BioSample,Bytes,Center Name,Consent,DATASTORE filetype,DATASTORE provider,DATASTORE region,Experiment,GEO_Accession (exp),Instrument,LibraryLayout,LibrarySelection,LibrarySource,model,Mouse_ID,Organism,Platform,protocol,ReleaseDate,Sample Name,seeding_density,source_name,SRA Study,STRAIN
SRR8165390,RNA-Seq,131,2922728162,PRJNA498671,SAMN10321158,2198839780,GEO,public,"sra,fastq","ncbi,gs,s3","gs.US,s3.us-east-1,ncbi.public",SRX4940653,GSM3448086,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,CT26,1,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448086,1x106,tumor,SRP166967,BALB/c
SRR8165391,RNA-Seq,131,2968585367,PRJNA498671,SAMN10321158,2225896912,GEO,public,"fastq,sra","ncbi,gs,s3","ncbi.public,gs.US,s3.us-east-1",SRX4940653,GSM3448086,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,CT26,1,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448086,1x106,tumor,SRP166967,BALB/c
SRR8165392,RNA-Seq,131,2465791254,PRJNA498671,SAMN10321157,1866182909,GEO,public,"fastq,sra","gs,ncbi,s3","gs.US,s3.us-east-1,ncbi.public",SRX4940654,GSM3448087,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,CT26,2,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448087,1x106,tumor,SRP166967,BALB/c
SRR8165393,RNA-Seq,131,2505045666,PRJNA498671,SAMN10321157,1889946041,GEO,public,"sra,fastq","s3,ncbi,gs","gs.US,ncbi.public,s3.us-east-1",SRX4940654,GSM3448087,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,CT26,2,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448087,1x106,tumor,SRP166967,BALB/c
SRR8165394,RNA-Seq,131,3303995933,PRJNA498671,SAMN10321156,2497179903,GEO,public,"fastq,sra","gs,s3,ncbi","ncbi.public,gs.US,s3.us-east-1",SRX4940655,GSM3448088,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,MC-38,3,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448088,2x105,tumor,SRP166967,C57B6/J
SRR8165396,RNA-Seq,131,2635313769,PRJNA498671,SAMN10321155,1991407014,GEO,public,"fastq,sra","s3,gs,ncbi","ncbi.public,gs.US,s3.us-east-1",SRX4940656,GSM3448089,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,MC-38,4,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448089,2x105,tumor,SRP166967,C57B6/J
SRR8165397,RNA-Seq,131,2675908049,PRJNA498671,SAMN10321155,2015167140,GEO,public,"fastq,sra","gs,ncbi,s3","ncbi.public,s3.us-east-1,gs.US",SRX4940656,GSM3448089,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,MC-38,4,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448089,2x105,tumor,SRP166967,C57B6/J
SRR8165399,RNA-Seq,131,2726481254,PRJNA498671,SAMN10321154,2049992900,GEO,public,"sra,fastq","gs,s3,ncbi","gs.US,s3.us-east-1,ncbi.public",SRX4940657,GSM3448090,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,CT26,1,Mus musculus,ILLUMINA,overnight,2018-11-06T00:00:00Z,GSM3448090,1x106,tumor,SRP166967,BALB/c
SRR8165400,RNA-Seq,131,2707421802,PRJNA498671,SAMN10321153,2042859442,GEO,public,"sra,fastq","ncbi,gs,s3","s3.us-east-1,gs.US,ncbi.public",SRX4940658,GSM3448091,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,CT26,2,Mus musculus,ILLUMINA,overnight,2018-11-06T00:00:00Z,GSM3448091,1x106,tumor,SRP166967,BALB/c
SRR8165401,RNA-Seq,131,2750252252,PRJNA498671,SAMN10321153,2068312862,GEO,public,"fastq,sra","s3,ncbi,gs","ncbi.public,s3.us-east-1,gs.US",SRX4940658,GSM3448091,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,CT26,2,Mus musculus,ILLUMINA,overnight,2018-11-06T00:00:00Z,GSM3448091,1x106,tumor,SRP166967,BALB/c
SRR8165402,RNA-Seq,131,2905315642,PRJNA498671,SAMN10321152,2191405713,GEO,public,"sra,fastq","s3,ncbi,gs","s3.us-east-1,ncbi.public,gs.US",SRX4940659,GSM3448092,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,MC-38,3,Mus musculus,ILLUMINA,overnight,2018-11-06T00:00:00Z,GSM3448092,2x105,tumor,SRP166967,C57B6/J
SRR8165404,RNA-Seq,131,2953280375,PRJNA498671,SAMN10321152,2220004587,GEO,public,"fastq,sra","gs,ncbi,s3","s3.us-east-1,ncbi.public,gs.US",SRX4940659,GSM3448092,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,MC-38,3,Mus musculus,ILLUMINA,overnight,2018-11-06T00:00:00Z,GSM3448092,2x105,tumor,SRP166967,C57B6/J
SRR8165405,RNA-Seq,131,4017681313,PRJNA498671,SAMN10321151,3029510810,GEO,public,"fastq,sra","ncbi,s3,gs","ncbi.public,s3.us-east-1,gs.US",SRX4940660,GSM3448093,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,MC-38,4,Mus musculus,ILLUMINA,overnight,2018-11-06T00:00:00Z,GSM3448093,2x105,tumor,SRP166967,C57B6/J
SRR8165408,RNA-Seq,131,552843580,PRJNA498671,SAMN10321150,402330998,GEO,public,"fastq,sra","ncbi,s3,gs","ncbi.public,gs.US,s3.us-east-1",SRX4940661,GSM3448094,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,LL2,5,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448094,2x105,tumor,SRP166967,C57B6/J
SRR8165409,RNA-Seq,131,573055439,PRJNA498671,SAMN10321149,418320616,GEO,public,"sra,fastq","gs,s3,ncbi","s3.us-east-1,ncbi.public,gs.US",SRX4940662,GSM3448095,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,LL2,6,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448095,2x105,tumor,SRP166967,C57B6/J
SRR8165410,RNA-Seq,131,574865466,PRJNA498671,SAMN10321149,419081975,GEO,public,"sra,fastq","ncbi,s3,gs","s3.us-east-1,ncbi.public,gs.US",SRX4940662,GSM3448095,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,LL2,6,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448095,2x105,tumor,SRP166967,C57B6/J
SRR8165411,RNA-Seq,131,776307834,PRJNA498671,SAMN10321148,564720270,GEO,public,"sra,fastq","s3,ncbi,gs","gs.US,ncbi.public,s3.us-east-1",SRX4940663,GSM3448096,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,B16-F10,7,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448096,2x105,tumor,SRP166967,C57B6/J
SRR8165412,RNA-Seq,131,778869670,PRJNA498671,SAMN10321148,566039220,GEO,public,"fastq,sra","gs,s3,ncbi","ncbi.public,s3.us-east-1,gs.US",SRX4940663,GSM3448096,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,B16-F10,7,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448096,2x105,tumor,SRP166967,C57B6/J
SRR8165413,RNA-Seq,131,880131622,PRJNA498671,SAMN10321147,639293962,GEO,public,"fastq,sra","gs,s3,ncbi","gs.US,ncbi.public,s3.us-east-1",SRX4940664,GSM3448097,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,B16-F10,8,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448097,2x105,tumor,SRP166967,C57B6/J
SRR8165414,RNA-Seq,131,883392867,PRJNA498671,SAMN10321147,640947670,GEO,public,"sra,fastq","gs,ncbi,s3","s3.us-east-1,gs.US,ncbi.public",SRX4940664,GSM3448097,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,B16-F10,8,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448097,2x105,tumor,SRP166967,C57B6/J
SRR8165416,RNA-Seq,131,845564390,PRJNA498671,SAMN10321146,611252607,GEO,public,"fastq,sra","s3,gs,ncbi","gs.US,ncbi.public,s3.us-east-1",SRX4940665,GSM3448098,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,EMT6,9,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448098,2x105,tumor,SRP166967,BALB/c
SRR8165418,RNA-Seq,131,825939280,PRJNA498671,SAMN10321145,599023791,GEO,public,"fastq,sra","ncbi,gs,s3","gs.US,ncbi.public,s3.us-east-1",SRX4940666,GSM3448099,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,EMT6,10,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448099,2x105,tumor,SRP166967,BALB/c
SRR8165421,RNA-Seq,131,914832474,PRJNA498671,SAMN10321142,663621522,GEO,public,"fastq,sra","gs,s3,ncbi","s3.us-east-1,gs.US,ncbi.public",SRX4940668,GSM3448101,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,SA1N,12,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448101,1x106,tumor,SRP166967,A/J
SRR8165424,RNA-Seq,131,345450537,PRJNA498671,SAMN10321141,246088421,GEO,public,"fastq,sra","s3,gs,ncbi","ncbi.public,s3.us-east-1,gs.US",SRX4940669,GSM3448102,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,LL2,5,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448102,2x105,tumor,SRP166967,C57B6/J
SRR8165425,RNA-Seq,131,491688195,PRJNA498671,SAMN10321140,351152739,GEO,public,"sra,fastq","s3,ncbi,gs","s3.us-east-1,gs.US,ncbi.public",SRX4940670,GSM3448103,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,LL2,6,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448103,2x105,tumor,SRP166967,C57B6/J
SRR8165426,RNA-Seq,131,494032833,PRJNA498671,SAMN10321140,352371301,GEO,public,"fastq,sra","s3,ncbi,gs","ncbi.public,s3.us-east-1,gs.US",SRX4940670,GSM3448103,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,LL2,6,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448103,2x105,tumor,SRP166967,C57B6/J
SRR8165427,RNA-Seq,131,506652587,PRJNA498671,SAMN10321139,363785962,GEO,public,"fastq,sra","gs,s3,ncbi","s3.us-east-1,gs.US,ncbi.public",SRX4940671,GSM3448104,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,B16-F10,7,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448104,2x105,tumor,SRP166967,C57B6/J
SRR8165428,RNA-Seq,131,507896432,PRJNA498671,SAMN10321139,364289314,GEO,public,"fastq,sra","gs,s3,ncbi","s3.us-east-1,gs.US,ncbi.public",SRX4940671,GSM3448104,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,B16-F10,7,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448104,2x105,tumor,SRP166967,C57B6/J
SRR8165429,RNA-Seq,131,922392877,PRJNA498671,SAMN10321138,659497193,GEO,public,"sra,fastq","s3,gs,ncbi","ncbi.public,gs.US,s3.us-east-1",SRX4940672,GSM3448105,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,B16-F10,8,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448105,2x105,tumor,SRP166967,C57B6/J
SRR8165430,RNA-Seq,131,925464434,PRJNA498671,SAMN10321138,660846472,GEO,public,"fastq,sra","s3,gs,ncbi","s3.us-east-1,gs.US,ncbi.public",SRX4940672,GSM3448105,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,B16-F10,8,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448105,2x105,tumor,SRP166967,C57B6/J
SRR8165432,RNA-Seq,131,634645613,PRJNA498671,SAMN10321137,453881596,GEO,public,"fastq,sra","gs,s3,ncbi","s3.us-east-1,ncbi.public,gs.US",SRX4940673,GSM3448106,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,EMT6,9,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448106,2x105,tumor,SRP166967,BALB/c
SRR8165433,RNA-Seq,131,671663986,PRJNA498671,SAMN10321136,480362140,GEO,public,"sra,fastq","gs,s3,ncbi","ncbi.public,s3.us-east-1,gs.US",SRX4940674,GSM3448107,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,EMT6,10,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448107,2x105,tumor,SRP166967,BALB/c
SRR8165434,RNA-Seq,131,674518869,PRJNA498671,SAMN10321136,481822827,GEO,public,"fastq,sra","gs,ncbi,s3","ncbi.public,s3.us-east-1,gs.US",SRX4940674,GSM3448107,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,EMT6,10,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448107,2x105,tumor,SRP166967,BALB/c
SRR8165435,RNA-Seq,131,578334608,PRJNA498671,SAMN10321135,414115339,GEO,public,"sra,fastq","gs,s3,ncbi","s3.us-east-1,gs.US,ncbi.public",SRX4940675,GSM3448108,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,SA1N,11,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448108,1x106,tumor,SRP166967,A/J
SRR8165436,RNA-Seq,131,581138532,PRJNA498671,SAMN10321135,415666483,GEO,public,"fastq,sra","ncbi,gs,s3","gs.US,s3.us-east-1,ncbi.public",SRX4940675,GSM3448108,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,SA1N,11,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448108,1x106,tumor,SRP166967,A/J
SRR8165437,RNA-Seq,131,602031460,PRJNA498671,SAMN10321134,432873839,GEO,public,"fastq,sra","ncbi,s3,gs","s3.us-east-1,gs.US,ncbi.public",SRX4940676,GSM3448109,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,SA1N,12,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448109,1x106,tumor,SRP166967,A/J
SRR8165438,RNA-Seq,131,603939606,PRJNA498671,SAMN10321134,433757290,GEO,public,"fastq,sra","gs,s3,ncbi","gs.US,ncbi.public,s3.us-east-1",SRX4940676,GSM3448109,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,SA1N,12,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448109,1x106,tumor,SRP166967,A/J
SRR8165398,RNA-Seq,131,2683696654,PRJNA498671,SAMN10321154,2024775876,GEO,public,"fastq,sra","ncbi,gs,s3","s3.us-east-1,ncbi.public,gs.US",SRX4940657,GSM3448090,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,CT26,1,Mus musculus,ILLUMINA,overnight,2018-11-06T00:00:00Z,GSM3448090,1x106,tumor,SRP166967,BALB/c
SRR8165406,RNA-Seq,131,4081297533,PRJNA498671,SAMN10321151,3066746894,GEO,public,"fastq,sra","gs,ncbi,s3","s3.us-east-1,gs.US,ncbi.public",SRX4940660,GSM3448093,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,MC-38,4,Mus musculus,ILLUMINA,overnight,2018-11-06T00:00:00Z,GSM3448093,2x105,tumor,SRP166967,C57B6/J
SRR8165415,RNA-Seq,131,842460476,PRJNA498671,SAMN10321146,609712707,GEO,public,"sra,fastq","ncbi,gs,s3","ncbi.public,s3.us-east-1,gs.US",SRX4940665,GSM3448098,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,EMT6,9,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448098,2x105,tumor,SRP166967,BALB/c
SRR8165417,RNA-Seq,131,822278354,PRJNA498671,SAMN10321145,597111751,GEO,public,"sra,fastq","ncbi,gs,s3","gs.US,ncbi.public,s3.us-east-1",SRX4940666,GSM3448099,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,EMT6,10,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448099,2x105,tumor,SRP166967,BALB/c
SRR8165419,RNA-Seq,131,900657357,PRJNA498671,SAMN10321144,650669858,GEO,public,"fastq,sra","ncbi,gs,s3","ncbi.public,gs.US,s3.us-east-1",SRX4940667,GSM3448100,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,SA1N,11,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448100,1x106,tumor,SRP166967,A/J
SRR8165420,RNA-Seq,131,903347573,PRJNA498671,SAMN10321144,651803028,GEO,public,"fastq,sra","ncbi,s3,gs","gs.US,s3.us-east-1,ncbi.public",SRX4940667,GSM3448100,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,SA1N,11,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448100,1x106,tumor,SRP166967,A/J
SRR8165422,RNA-Seq,131,916869917,PRJNA498671,SAMN10321142,664303890,GEO,public,"sra,fastq","s3,ncbi,gs","gs.US,ncbi.public,s3.us-east-1",SRX4940668,GSM3448101,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,SA1N,12,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448101,1x106,tumor,SRP166967,A/J
SRR8165423,RNA-Seq,131,343589420,PRJNA498671,SAMN10321141,245119618,GEO,public,"sra,fastq","gs,ncbi,s3","s3.us-east-1,gs.US,ncbi.public",SRX4940669,GSM3448102,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,LL2,5,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448102,2x105,tumor,SRP166967,C57B6/J
SRR8165431,RNA-Seq,131,632519090,PRJNA498671,SAMN10321137,452918714,GEO,public,"fastq,sra","ncbi,gs,s3","ncbi.public,gs.US,s3.us-east-1",SRX4940673,GSM3448106,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,EMT6,9,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448106,2x105,tumor,SRP166967,BALB/c
SRR8165395,RNA-Seq,131,3354326788,PRJNA498671,SAMN10321156,2527235059,GEO,public,"fastq,sra","gs,ncbi,s3","gs.US,ncbi.public,s3.us-east-1",SRX4940655,GSM3448088,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,MC-38,3,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448088,2x105,tumor,SRP166967,C57B6/J
SRR8165407,RNA-Seq,131,551165994,PRJNA498671,SAMN10321150,401583816,GEO,public,"fastq,sra","gs,ncbi,s3","gs.US,ncbi.public,s3.us-east-1",SRX4940661,GSM3448094,Illumina HiSeq 2500,PAIRED,cDNA,TRANSCRIPTOMIC,LL2,5,Mus musculus,ILLUMINA,fresh,2018-11-06T00:00:00Z,GSM3448094,2x105,tumor,SRP166967,C57B6/J', sep =',', quote = '"', header = TRUE)

str(annot)
```

# Data load Seurat objects

```{r justfordebugging, eval = FALSE}
## params <- list(genes_cellranger = "/home/imallona/repeats_sc/runs/frozen_pbmc_donor_a/cellranger_standard/outs/filtered_feature_bc_matrix/barcodes.tsv.gz",
##                repeats_cellranger= "/home/imallona/repeats_sc/runs/frozen_pbmc_donor_a/cellranger_repeats/outs/filtered_feature_bc_matrix/barcodes.tsv.gz",
##                repeats_featurecounts= "/home/imallona/repeats_sc/runs/frozen_pbmc_donor_a/count_repeats_on_cellranger_standard/frozen_pbmc_donor_a_repeats.counts.gz")

## SRR10974769

## params <- list(run1 = '/home/imallona/repeats_sc/runs/SRR10974769/SRR10974769_pmbc_cellranger_regress_nCount_FALSE_nFeature_FALSE.rds',
##                run2 = '/home/imallona/repeats_sc/runs/5k_pbmc_v3/5k_pbmc_v3_pmbc_cellranger_regress_nCount_FALSE_nFeature_FALSE.rds',
##                run3 = '/home/imallona/repeats_sc/runs/frozen_pbmc_donor_a/frozen_pbmc_donor_a/summarize_cellranger_run_regress_nCount_FALSE_nFeature_FALSE.rds',
##                run4 = '/home/imallona/repeats_sc/runs/5k_pbmc_v3/5k_pbmc_v3_pmbc_cellranger_regress_nCount_FALSE_nFeature_FALSE.rds')

params <- list(results_path = '/home/imallona/repeats_sc/runs/GSE121861_kumar_mouse',
               regress_genes_nCount = FALSE,
               regress_genes_nFeature = FALSE)

```

```{r}
d <- list(genes_cellranger = list(),
          repeats_only_unique = list(),
          repeats_only_multi = list())
## fns <- list()
```

Focus in cellranger genes and repeats only cellranger

```{r, message = FALSE}

## cr cellranger
## fc featurecounts
for (srr in annot$Run) {

    d[['genes_cellranger']][[srr]] <- CreateSeuratObject(
        counts = Read10X(data.dir = file.path(params$results_path, srr, 'cellranger_standard',
                                              'outs', 'filtered_feature_bc_matrix')),
                         project = srr,
                         min.cells = 0,
        min.features = 0)

    d[['repeats_only_unique']][[srr]] <- load_featurecounts_repeats(
        repeats_fn = file.path(
            params$results_path, srr, 'count_repeats_on_cellranger_standard', 'unique_reads',
            'GSE121861_kumar_mouse_repeats.counts.gz'),
        srr)
                                                                 
    d[['repeats_only_multi']][[srr]] <- load_featurecounts_repeats(
        repeats_fn = file.path(
            params$results_path, srr, 'count_repeats_on_cellranger_standard', 'multimappers',
            'GSE121861_kumar_mouse_repeats.counts.gz'),
        id = srr)
}

for (flavour in names(d)) {
    for (srr in names(d[[flavour]])) {

        d[[flavour]][[srr]] <- RenameCells(
            d[[flavour]][[srr]],
            new.names = paste(srr, colnames(d[[flavour]][[srr]]), sep = '_'))
    }
}
        
```


# Basic Seurat processing

## QC before filtering {.tabset}

```{r}
for (item in names(d$genes_cellranger)) {
    d$genes_cellranger[[item]][['percent.mt']] <- PercentageFeatureSet(d$genes_cellranger[[item]],
                                                                       pattern = "^mt-")
}

```

Subset to cells present at the QC-ed d$genes_cellranger assay

```{r}
for (item in names(d$genes_cellranger)) {
    libsize.drop <- isOutlier(d$genes_cellranger[[item]]$nCount_RNA, nmads = 3, type = "both", log = TRUE)
    feature.drop <- isOutlier(d$genes_cellranger[[item]]$nFeature_RNA, nmads = 3, type = "both", log = TRUE)
    mito.drop <- isOutlier(d$genes_cellranger[[item]]$percent.mt, nmads = 3, type = "higher")

    selected_cells <- colnames(d$genes_cellranger)[!(libsize.drop | feature.drop | mito.drop)]

    d$genes_cellranger[[item]] <- subset(d$genes_cellranger[[item]], cells = selected_cells)
    d$repeats_only_unique[[item]] <- subset(d$repeats_only_unique[[item]], cells = selected_cells)
    d$repeats_only_multi[[item]] <- subset(d$repeats_only_multi[[item]], cells = selected_cells)
}
```


```{r, eval = FALSE}
sapply(d$genes_cellranger, function(x) head(colnames(x)))
sapply(d$repeats_only_multi, function(x) head(colnames(x)))

```

Remove runs with low number of cells

```{r}
## table(sapply(d$genes_cellranger, function(x) ncol(x) >100))

idx <- names(d$genes_cellranger)[sapply(d$genes_cellranger, ncol) >= 100]

for (flavour in names(d))
    d[[flavour]] <- d[[flavour]][idx]
```

Integrate, add SRR identifiers. Num variable features: 10 pct if genes, 25 pct if repeats

Skip integration and merge

```{r, message = FALSE, eval = FALSE}
for (flavour in names(d)) {
    tryCatch({
        for (i in 1:length(d[[flavour]])) {
            d[[flavour]][[i]] <- NormalizeData(d[[flavour]][[i]], verbose = FALSE)
            
            if (grepl('repeats', flavour)) {
                d[[flavour]][[i]] <- FindVariableFeatures(
                    d[[flavour]][[i]], selection.method = "vst",
                    nfeatures = round(nrow(d[[flavour]][[i]])/4))
            } else {
                d[[flavour]][[i]] <- FindVariableFeatures(
                    d[[flavour]][[i]], selection.method = "vst",
                    nfeatures = round(nrow(d[[flavour]][[i]])/10))
            }

        }
                
        anchors <- FindIntegrationAnchors(object.list = d[[flavour]], dims = 1:30,
                                          k.anchor = 5,
                                          k.filter = 100,
                                          k.score = 30)
        
        d[[flavour]] <- IntegrateData(anchorset = anchors, dims = 1:30)
    }, error = function(x) print(x))
    
}

```


```{r}
for (flavour in names(d)) {
    d[[flavour]] <- merge(d[[flavour]][[1]], d[[flavour]][2:length(d[[flavour]])])
}

d

head(d[[1]]@meta.data)
```

Does it have the srr? Merge metadata, just in case, by SRR

```{r}
for  (flavour in names(d)) {
    d[[flavour]]@meta.data$cell <- rownames(d[[flavour]]@meta.data)
    d[[flavour]]@meta.data <- merge(d[[flavour]]@meta.data,
                                    annot[c('Run', 'model', 'source_name')], by.x  = 'orig.ident',
                                    by.y = 'Run', all.x = TRUE)

    rownames(d[[flavour]]@meta.data) <- d[[flavour]]@meta.data$cell
}

```


## Dimred PCA {.tabset}

```{r}
for (flavour in names(d)) {
    d[[flavour]] <- NormalizeData(d[[flavour]], normalization.method = "LogNormalize",
                                  scale.factor = 10000)
    
    if (grepl('repeats', flavour)) {
        d[[flavour]] <- FindVariableFeatures(
            d[[flavour]], selection.method = "vst",
            nfeatures = round(nrow(d[[flavour]])/4))
    } else {
        d[[flavour]] <- FindVariableFeatures(
            d[[flavour]], selection.method = "vst",
            nfeatures = round(nrow(d[[flavour]])/10))
    }
}

```

```{r, message = FALSE, results ='asis'}
NUM_PCS <- 1:30
for (item in names(d)) {
    cat("### ", item, "\n\n")

    to_regress <- NULL       
    
    ## if (item == 'genes_cellranger') {
        d[[item]] <- ScaleData(d[[item]], vars.to.regress = to_regress)
    ## } else {
    ##     ## in this case, the variables to regress come from the genes!
    ##     ## e.g. is the genes library size the one to normalize the repets'
    ##     tmp <- d$genes_cellranger@meta.data[,c("cell", "nCount_RNA", "nFeature_RNA")]
    ##     ## tmp$cell <- rownames(tmp)
    ##     ## d[[item]]@meta.data$cell <- rownames(d[[item]]@meta.data)
    ##     d[[item]]@meta.data <- merge(d[[item]]@meta.data, tmp, by = 'cell',
    ##                                  suffixes = c('', '_genes'))
    ##     rownames(d[[item]]@meta.data) <- d[[item]]@meta.data$cell


    ##     d[[item]] <- ScaleData(d[[item]],
    ##                            vars.to.regress = if(is.null(to_regress)) NULL else paste0(to_regress, '_genes'))
        
    ##     rm(tmp) 
    ## }
    
    d[[item]] <- RunPCA(d[[item]], features = VariableFeatures(object = d[[item]]))

    Idents(d[[item]]) <- "model"
    
    DimPlot(d[[item]], reduction = "pca") + ggtitle(item)
    
    print(ElbowPlot(d[[item]]) + ggtitle(item))


    cat('\n\n')
}


```

```{r, message = FALSE, results = 'asis'}
for (item in names(d)) {

    d[[item]] <- FindNeighbors(d[[item]], dims = NUM_PCS)
    for (res in seq(0.2, 2, by = 0.6)) {
        d[[item]] <- FindClusters(d[[item]], resolution = res, verbose = FALSE)

    }

    d[[item]] <- RunUMAP(d[[item]], dims = NUM_PCS)
        
    d[[item]] <- RunTSNE(d[[item]], dims = NUM_PCS, check_duplicates = FALSE )
}


```


## Dimred UMAP and tSNE {.tabset}

```{r results = 'asis', cache = FALSE, warning = FALSE}
for (item in names(d)) {
    cat("### ", item, " {.tabset} \n\n")
    
    p1 <- DimPlot(d[[item]], reduction = "umap", group.by = 'model')
    p2 <- DimPlot(d[[item]], reduction = "tsne", group.by = 'model')

    cat('#### umap \n\n')
    print(p1 + ggtitle(item) + theme(legend.position="right", aspect.ratio = 1))
    cat('\n\n')
    
    cat('#### tsne \n\n')

    print(p2 + ggtitle(item) + theme(legend.position="right", aspect.ratio = 1))

    cat('\n\n')
    rm(p1, p2)
    cat('\n\n')

}
```

# Differential expression {.tabset}


```{r, results = 'asis'}

markers <- list()

for (item in names(d)) {
    Idents(d[[item]]) <- "model"
    d[[item]]@active.assay <- 'RNA'
    
    cat('##' , item, '\n\n')

    markers[[item]] <- FindAllMarkers(object = d[[item]])

    markers[[item]] <- markers[[item]][markers[[item]]$p_val_adj < 0.05,]
    markers[[item]] <- markers[[item]][order(markers[[item]]$avg_logFC, decreasing = TRUE),]
    markers[[item]]$flavour <- item

    print(nrow(markers[[item]]))
    cat('\n\n')
}


## markers <- do.call(rbind.data.frame, markers)
```

# Export objects

```{r}
saveRDS(object = markers, file = params$markers_rds)
saveRDS(object = d, file = params$seurat_rds)
```


# Timestamp

```{r, cache = FALSE}
date()
sessionInfo()
devtools::session_info()

```



```{r}
knitr::knit_exit()
```


```{r}

setdiff(names(d[[1]]), names(d[[2]]))
stopifnot(names(d[[1]]) == names(d[[2]]))
```

# Seurat integration? Worth it?

Merge the two datasets, and plot them without integration

```{r}
combined <- list()

for (flavour in names(d[[1]])) {
    combined[[flavour]] <- merge(d[[1]][[flavour]], y = d[[2]][[flavour]],
                              add.cell.ids = c(names(d)[1],
                                               names(d)[2]),
                              project = flavour)

    combined[[flavour]] <- NormalizeData(combined[[flavour]],
                                      normalization.method = "LogNormalize", 
                                      scale.factor = 10000)

    combined[[flavour]]$srr <- sapply(strsplit(colnames(combined[[flavour]]), '_'),
                                      function(x) return(x[1]))

}

for (item in names(combined)) {
    ## if repeats, we take half of them
    ## else if genes, then the top 10 pct
    if (grepl('repeats', item)) {
        combined[[item]] <- FindVariableFeatures(combined[[item]], selection.method = "vst",
                                          nfeatures = round(nrow(combined[[item]])/4))
    } else {
        combined[[item]] <- FindVariableFeatures(combined[[item]], selection.method = "vst",
                                          nfeatures = round(nrow(combined[[item]])/10))
    }

                                      
    NUM_PCS <- 1:20
    cat("### ", item, "\n\n")

    to_regress <- NULL
    if (as.logical(params$regress_genes_nCount)) {
        to_regress <- c(to_regress, 'nCount_RNA')
    }
    if (as.logical(params$regress_genes_nFeature)) {
        to_regress <- c(to_regress, 'nFeature_RNA')
        
    }
    
    if (item == 'genes_cellranger') {
        combined[[item]] <- ScaleData(combined[[item]], vars.to.regress = to_regress)
    } else {
        ## in this case, the variables to regress come from the genes!
        ## e.g. is the genes library size the one to normalize the repets'
        tmp <- combined$genes_cellranger@meta.data
        tmp$cell <- rownames(tmp)
        combined[[item]]@meta.data$cell <- rownames(combined[[item]]@meta.data)
        combined[[item]]@meta.data <- merge(combined[[item]]@meta.data, tmp, by = 'cell',
                                     suffixes = c('', '_genes'))
        rownames(combined[[item]]@meta.data) <- combined[[item]]@meta.data$cell


        combined[[item]] <- ScaleData(combined[[item]],
                               vars.to.regress = if(is.null(to_regress)) NULL else paste0(to_regress, '_genes'))
        
        rm(tmp) 
    }
    
    combined[[item]] <- RunPCA(combined[[item]], features = VariableFeatures(object = combined[[item]]))
    

    print(ElbowPlot(combined[[item]]) + ggtitle(item))

    print(DimPlot(combined[[item]], reduction = "pca", group.by = 'srr') + ggtitle (item))
    
    cat('\n\n')
}


```


##  Cluster


```{r, message = FALSE, results = 'asis'}
for (item in names(combined)) {

    combined[[item]] <- FindNeighbors(combined[[item]], dims = NUM_PCS)
    for (res in seq(0.2, 2, by = 1)) {
        combined[[item]] <- FindClusters(combined[[item]], resolution = res, verbose = FALSE)

    }

    combined[[item]] <- RunUMAP(combined[[item]], dims = NUM_PCS)
        
    combined[[item]] <- RunTSNE(combined[[item]], dims = NUM_PCS, check_duplicates = FALSE )
}


```


## Dimred UMAP and tSNE {.tabset}

```{r fig.width = 8, fig.height = 5, results = 'asis', cache = FALSE, warning = FALSE}
for (item in names(combined)) {
    cat("### ", item, "\n\n")
    
    p1 <- DimPlot(combined[[item]], reduction = "umap", group.by = 'srr')
    p2 <- DimPlot(combined[[item]], reduction = "tsne", group.by = 'srr')

    print(p1 + ggtitle(item))
    print(p2 + ggtitle(item))
    
    rm(p1, p2)
    cat('\n\n')

}
```

# Differential expression WT vs DKO {.tabset}

```{r}

markers <- list()

for (item in names(combined)) {
    cat('## , item\n')
    print(item)
    Idents(combined[[item]]) <- 'srr'
    markers[[item]] <- FindAllMarkers(object = combined[[item]])

    markers[[item]] <- markers[[item]][markers[[item]]$p_val_adj < 0.05,]
    markers[[item]] <- markers[[item]][order(markers[[item]]$avg_logFC, decreasing = TRUE),]
    markers[[item]]$flavour <- item
    print(nrow(markers[[item]]))
    print(head(markers[[item]]))

    cat('\n\n')
}


markers <- do.call(rbind.data.frame, markers)
```

# Upset plots

## Within SRR

```{r, fig.height = 6, fig.width = 12}
markers <- as.data.frame(markers)


for (srr in unique(markers$cluster)) {
    cat('## ', srr, '\n\n')
    ul <- list()
    for (flavour in unique(markers[markers$cluster == srr, 'flavour'])) {

        for (flavour in grep('repeats', unique(markers[ markers$cluster == srr &
                                                        markers$flavour == flavour, 'flavour']),
                             value = TRUE)) {
                    
            ul[[paste(srr, flavour, flavour, sep = "_")]] <- markers[
                markers$cluster == srr &
                markers$flavour == flavour &
                markers$flavour == flavour, 'gene']
        }
    }
    print(upset(fromList(ul), nsets = 10,  text.scale = 0.8))
    grid.text(paste(srr) ,x = 0.65, y=0.95, gp=gpar(fontsize = 10))
    
    ## rm(ul)
    cat('\n\n')

}

```




<!-- # Most variable repeats cellranger {.tabset} -->

<!-- Similarly for repeats expression (avg +- sd of repeats CPM by gene expression cluster?) -->


<!-- ```{r, fig.height = 5, fig.width = 15, results = 'asis'} -->
<!-- res <- 'RNA_snn_res.0.2' -->

<!-- ## ## selected <- c('AluY', 'AluSx', 'MER49', 'MIR', 'L2a', 'L2b', 'L1PA11') -->
<!-- ## selected <- c(names(head(sort(apply(GetAssayData(d[[run]]$repeats_cellranger),1, coef_var), decreasing = TRUE), -->
<!-- ##                          5)), c('AluY', 'AluSx', 'L1PA11', 'MIR', 'L2a')) -->

<!-- ## maybe with variable features? -->

<!-- for (run in names(combined)[sapply(combined, function(x) return('repeats_cellranger' %in% names(x)))]) { -->
<!--     cat("## ", run, "\n\n") -->
<!--     repeats_cellranger <- as.data.frame(FetchData( -->
<!--         combined[[run]]$repeats_cellranger, -->
<!--         vars = head(VariableFeatures(combined[[run]]$repeats_cellranger), 10) , slot = 'data')) -->
        
<!--     repeats_cellranger$cell <- rownames(repeats_cellranger) -->
        
<!--     combined[[run]]$genes_cellranger@meta.data$cell <- rownames(combined[[run]]$genes_cellranger@meta.data) -->
<!--     repeats_cellranger <- merge(combined[[run]]$genes_cellranger@meta.data[, c(res, 'cell') ], -->
<!--                      repeats_cellranger, by = 'cell', all.x = TRUE) -->
    
<!--     rownames(repeats_cellranger) <- repeats_cellranger$cell -->
    
<!--     repeats_cellranger <- melt(repeats_cellranger, id.vars = c('cell', res)) -->
    
<!--     colnames(repeats_cellranger) <- c('cell', res, 'repfamily', 'CPM') -->

<!--     print(ggplot(repeats_cellranger, -->
<!--            aes(x = get(res), y = CPM)) +  -->
<!--           geom_boxplot(outlier.size=1) + -->
<!--           scale_y_continuous(trans='log10') + -->
<!--         scale_fill_discrete(drop=FALSE) + -->
<!--         facet_wrap(~ factor(repfamily), ncol = 10) + -->
<!--         theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--         ylab('repeat expression [CPM]') + -->
<!--         xlab('gene-based clustering') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_cellranger', res))) -->
<!--     cat('\n\n') -->
<!-- } -->

<!-- ``` -->

<!-- # Most variable repeats cellranger by identity {.tabset} -->


<!-- ```{r, fig.height = 5, fig.width = 15, results = 'asis'} -->
<!-- scina <- 'scina' -->
<!-- ## agg <- list() -->

<!-- for (run in names(d)[sapply(d, function(x) return('repeats_cellranger' %in% names(x)))]) { -->
<!--     cat("## ", run, "\n\n") -->
    
<!--     repeats_cellranger <- as.data.frame(FetchData( -->
<!--         combined[[run]]$repeats_cellranger, -->
<!--         vars = head(VariableFeatures(combined[[run]]$repeats_cellranger), 10) , slot = 'data')) -->


    
<!--     repeats_cellranger$cell <- rownames(repeats_cellranger) -->

<!--     combined[[run]]$genes@meta.data$cell <- rownames(combined[[run]]$genes@meta.data) -->
<!--     repeats_cellranger <- merge(combined[[run]]$genes@meta.data[, c(scina, 'cell') ], -->
<!--                      repeats_cellranger, by = 'cell', all.x = TRUE) -->
    
<!--     rownames(repeats_cellranger) <- repeats_cellranger$cell -->
    
<!--     repeats_cellranger <- melt(repeats_cellranger, id.vars = c('cell', scina)) -->
    
<!--     colnames(repeats_cellranger) <- c('cell', scina, 'repfamily', 'CPM') -->

<!--     print(ggplot(repeats_cellranger, -->
<!--            aes(x = scina, y = CPM)) +  -->
<!--           geom_boxplot(outlier.size=1) + -->
<!--           scale_y_continuous(trans='log10') + -->
<!--         scale_fill_discrete(drop=FALSE) + -->
<!--         facet_wrap(~ factor(repfamily), ncol = 10) + -->
<!--         theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--         ylab('repeat expresssion [CPM]') + -->
<!--         xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_cellranger', scina))) -->
<!--     cat('\n\n') -->

<!-- } -->

<!-- ``` -->

<!-- # All integrated (repeats, scina) -->

<!-- ```{r fig.width = 14, fig.height = 14} -->

<!-- ## first aggregate all repeats_cellranger stuff -->
<!-- sds <- lapply(d[sapply(d, function(x) return('repeats_cellranger' %in% names(x)))], -->
<!--               function(x) return(apply(as.data.frame(GetAssayData(x$repeats_cellranger, -->
<!--                                                         slot = 'data')), 1, sd))) -->

<!-- ## get the ones with the highest average sd -->

<!-- mean_sd <- apply(do.call(cbind.data.frame, sds), 1, mean) -->
<!-- selected <- names(head(sort(mean_sd, decreasing = TRUE), 10)) -->
<!-- ## selected -->


<!-- ## aggregate(agg['CPM'], by = list(agg$repfamily, agg$run),  FUN = function(x) c(mean = mean(x, na.rm = TRUE), n = length(x), sd = sd(x, na.rm = TRUE) )) -->

<!-- ## now get these data for selected features, aggregate and plot -->

<!-- tmp <- list() -->
<!-- for (run in names(d)) { -->
<!--     foo <- data.frame(run = run, -->
<!--                              FetchData(combined[[run]]$repeats_cellranger, vars = selected, slot = 'data')) -->

<!--     ## add annotations -->
<!--     bar <- data.frame(cell = rownames(foo), -->
<!--                       scina = combined[[run]]$repeats_cellranger@meta.data[rownames(foo), 'scina']) -->

<!--     stopifnot(rownames(foo) == bar$cell) -->
    
    
<!--     tmp[[run]] <- data.frame(foo, scina = bar$scina) -->
<!--     rm(foo, bar) -->
<!-- } -->

<!-- tmp <- do.call(rbind.data.frame, tmp) -->
<!-- str(tmp) -->

<!-- ## melt -->
<!-- repeats_cellranger <- melt(tmp, id.vars = c('run', 'scina')) -->
<!-- head(repeats_cellranger)     -->
<!-- colnames(repeats_cellranger) <- c('run', 'scina', 'repfamily', 'CPM') -->

<!-- ``` -->

<!-- ```{r, fig.width = 12, fig.height = 12} -->
<!-- print(ggplot(repeats_cellranger, -->
<!--              aes(x = scina, y = CPM, fill = run)) +  -->
<!--       geom_boxplot(outlier.size=1) + -->
<!--       scale_y_continuous(trans='log10') + -->
<!--       scale_fill_discrete(drop=FALSE) + -->
<!--       facet_wrap(~ factor(repfamily), ncol = 4) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--       ylab('repeat expresssion [CPM]') + -->
<!--       xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_cellranger', scina))) -->
<!-- ``` -->


<!-- ```{r, fig.width = 8, fig.height = 14} -->
<!-- print(ggplot(repeats_cellranger, -->
<!--              aes(x = repfamily, y = CPM, fill = run)) +  -->
<!--       geom_boxplot(outlier.size=1) + -->
<!--       scale_y_continuous(trans='log10') + -->
<!--       scale_fill_discrete(drop=FALSE) + -->
<!--       facet_grid(factor(scina)~ factor(run)) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--       ylab('repeat expresssion [CPM]') + -->
<!--       xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_cellranger', scina))) -->




<!-- ``` -->

<!-- # Cluster by average expression (repeats, SCINA) -->

<!-- Avoiding non-determined cell types -->

<!-- ```{r, results = 'asis', fig.height = 10, fig.width = 10} -->
<!-- repeats_cellranger <- repeats_cellranger[repeats_cellranger$scina != 'unknown',] -->
<!-- foo <- reshape2::dcast(repeats_cellranger, run + scina ~ repfamily, mean, value.var = 'CPM') -->


<!-- rownames(foo) <- paste(foo$run, foo$scina, sep = '_') -->

<!-- annot <- foo[,c('run', 'scina')] -->
<!-- rownames(annot) <- paste(annot$run, annot$scina, sep = '_') -->
<!-- foo <- foo[,-(1:2)] -->

<!-- pheatmap(foo, annotation_row = annot, show_rownames = FALSE) -->

<!-- pheatmap(t(foo), annotation_col = annot, show_colnames = FALSE) -->


<!-- ``` -->



<!-- <\!-- here start -\-> -->


<!-- # Most variable repeats star {.tabset} -->

<!-- Similarly for repeats expression (avg +- sd of repeats CPM by gene expression cluster?) -->

<!-- What's the point here? -->


<!-- ```{r, fig.height = 5, fig.width = 15, results = 'asis'} -->
<!-- res <- 'RNA_snn_res.0.2' -->

<!-- ## ## selected <- c('AluY', 'AluSx', 'MER49', 'MIR', 'L2a', 'L2b', 'L1PA11') -->
<!-- ## selected <- c(names(head(sort(apply(GetAssayData(combined[[run]]$repeats_star),1, coef_var), decreasing = TRUE), -->
<!-- ##                          5)), c('AluY', 'AluSx', 'L1PA11', 'MIR', 'L2a')) -->

<!-- ## maybe with variable features? -->

<!-- for (run in names(d)[sapply(d, function(x) return('repeats_star' %in% names(x)))]) { -->
<!--     cat("## ", run, "\n\n") -->
<!--     repeats_star <- as.data.frame(FetchData( -->
<!--         combined[[run]]$repeats_star, -->
<!--         vars = head(VariableFeatures(combined[[run]]$repeats_star), 10) , slot = 'data')) -->
        
<!--     repeats_star$cell <- rownames(repeats_star) -->
        
<!--     combined[[run]]$genes@meta.data$cell <- rownames(combined[[run]]$genes@meta.data) -->
<!--     repeats_star <- merge(combined[[run]]$genes@meta.data[, c(res, 'cell') ], -->
<!--                      repeats_star, by = 'cell', all.x = TRUE) -->
    
<!--     rownames(repeats_star) <- repeats_star$cell -->
    
<!--     repeats_star <- melt(repeats_star, id.vars = c('cell', res)) -->
    
<!--     colnames(repeats_star) <- c('cell', res, 'repfamily', 'CPM') -->

<!--     print(ggplot(repeats_star, -->
<!--            aes(x = get(res), y = CPM)) +  -->
<!--           geom_boxplot(outlier.size=1) + -->
<!--           scale_y_continuous(trans='log10') + -->
<!--         scale_fill_discrete(drop=FALSE) + -->
<!--         facet_wrap(~ factor(repfamily), ncol = 10) + -->
<!--         theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--         ylab('repeat expression [CPM]') + -->
<!--         xlab('gene-based clustering') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_star', res))) -->
<!--     cat('\n\n') -->
<!-- } -->

<!-- ``` -->

<!-- # Most variable repeats star by identity {.tabset} -->


<!-- ```{r, fig.height = 5, fig.width = 15, results = 'asis'} -->
<!-- scina <- 'scina' -->
<!-- ## agg <- list() -->

<!-- for (run in names(d)[sapply(d, function(x) return('repeats_star' %in% names(x)))]) { -->
<!--     cat("## ", run, "\n\n") -->
    
<!--     repeats_star <- as.data.frame(FetchData( -->
<!--         combined[[run]]$repeats_star, -->
<!--         vars = head(VariableFeatures(combined[[run]]$repeats_star), 10) , slot = 'data')) -->


    
<!--     repeats_star$cell <- rownames(repeats_star) -->

<!--     combined[[run]]$genes@meta.data$cell <- rownames(combined[[run]]$genes@meta.data) -->
<!--     repeats_star <- merge(combined[[run]]$genes@meta.data[, c(scina, 'cell') ], -->
<!--                      repeats_star, by = 'cell', all.x = TRUE) -->
    
<!--     rownames(repeats_star) <- repeats_star$cell -->
    
<!--     repeats_star <- melt(repeats_star, id.vars = c('cell', scina)) -->
    
<!--     colnames(repeats_star) <- c('cell', scina, 'repfamily', 'CPM') -->

<!--     print(ggplot(repeats_star, -->
<!--            aes(x = scina, y = CPM)) +  -->
<!--           geom_boxplot(outlier.size=1) + -->
<!--           scale_y_continuous(trans='log10') + -->
<!--         scale_fill_discrete(drop=FALSE) + -->
<!--         facet_wrap(~ factor(repfamily), ncol = 10) + -->
<!--         theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--         ylab('repeat expresssion [CPM]') + -->
<!--         xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_star', scina))) -->
<!--     cat('\n\n') -->

<!-- } -->

<!-- ``` -->

<!-- # All integrated (repeats, scina) -->

<!-- ```{r fig.width = 14, fig.height = 14} -->

<!-- ## first aggregate all repeats_star stuff -->
<!-- sds <- lapply(d[sapply(d, function(x) return('repeats_star' %in% names(x)))], -->
<!--               function(x) return(apply(as.data.frame(GetAssayData(x$repeats_star, -->
<!--                                                         slot = 'data')), 1, sd))) -->

<!-- ## get the ones with the highest average sd -->

<!-- mean_sd <- apply(do.call(cbind.data.frame, sds), 1, mean) -->
<!-- selected <- names(head(sort(mean_sd, decreasing = TRUE), 20)) -->
<!-- ## selected -->


<!-- ## aggregate(agg['CPM'], by = list(agg$repfamily, agg$run),  FUN = function(x) c(mean = mean(x, na.rm = TRUE), n = length(x), sd = sd(x, na.rm = TRUE) )) -->

<!-- ## now get these data for selected features, aggregate and plot -->

<!-- tmp <- list() -->
<!-- for (run in names(d)) { -->
<!--     foo <- data.frame(run = run, -->
<!--                              FetchData(combined[[run]]$repeats_star, vars = selected, slot = 'data')) -->

<!--     ## add annotations -->
<!--     bar <- data.frame(cell = rownames(foo), -->
<!--                       scina = combined[[run]]$repeats_star@meta.data[rownames(foo), 'scina']) -->

<!--     stopifnot(rownames(foo) == bar$cell) -->
    
    
<!--     tmp[[run]] <- data.frame(foo, scina = bar$scina) -->
<!--     rm(foo, bar) -->
<!-- } -->

<!-- tmp <- do.call(rbind.data.frame, tmp) -->
<!-- str(tmp) -->

<!-- ## melt -->
<!-- repeats_star <- melt(tmp, id.vars = c('run', 'scina')) -->
<!-- head(repeats_star)     -->
<!-- colnames(repeats_star) <- c('run', 'scina', 'repfamily', 'CPM') -->

<!-- ``` -->

<!-- ```{r, fig.width = 12, fig.height = 12} -->
<!-- print(ggplot(repeats_star, -->
<!--              aes(x = scina, y = CPM, fill = run)) +  -->
<!--       geom_boxplot(outlier.size=1) + -->
<!--       scale_y_continuous(trans='log10') + -->
<!--       scale_fill_discrete(drop=FALSE) + -->
<!--       facet_wrap(~ factor(repfamily), ncol = 4) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--       ylab('repeat expresssion [CPM]') + -->
<!--       xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_star', scina))) -->
<!-- ``` -->


<!-- ```{r, fig.width = 8, fig.height = 14} -->
<!-- print(ggplot(repeats_star, -->
<!--              aes(x = repfamily, y = CPM, fill = run)) +  -->
<!--       geom_boxplot(outlier.size=1) + -->
<!--       scale_y_continuous(trans='log10') + -->
<!--       scale_fill_discrete(drop=FALSE) + -->
<!--       facet_grid(factor(scina)~ factor(run)) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--       ylab('repeat expresssion [CPM]') + -->
<!--       xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_star', scina))) -->

<!-- ``` -->

<!-- # Cluster by average expression (repeats, SCINA) -->

<!-- Avoiding non-determined cell types -->

<!-- ```{r, results = 'asis', fig.height = 10, fig.width = 10} -->
<!-- repeats_star <- repeats_star[repeats_star$scina != 'unknown',] -->
<!-- foo <- reshape2::dcast(repeats_star, run + scina ~ repfamily, mean, value.var = 'CPM') -->


<!-- rownames(foo) <- paste(foo$run, foo$scina, sep = '_') -->

<!-- annot <- foo[,c('run', 'scina')] -->
<!-- rownames(annot) <- paste(annot$run, annot$scina, sep = '_') -->
<!-- foo <- foo[,-(1:2)] -->

<!-- pheatmap(foo, annotation_row = annot, show_rownames = FALSE) -->

<!-- pheatmap(t(foo), annotation_col = annot, show_colnames = FALSE) -->
<!-- ``` -->


<!-- <\!-- here stop -\-> -->



<!-- # Most variable repeats featurecounts {.tabset} -->

<!-- Similarly for repeats expression (avg +- sd of repeats CPM by gene expression cluster?) -->

<!-- What's the point here? -->


<!-- ```{r, fig.height = 5, fig.width = 15, results = 'asis'} -->
<!-- res <- 'RNA_snn_res.0.2' -->

<!-- ## ## selected <- c('AluY', 'AluSx', 'MER49', 'MIR', 'L2a', 'L2b', 'L1PA11') -->
<!-- ## selected <- c(names(head(sort(apply(GetAssayData(combined[[run]]$repeats_featurecounts),1, coef_var), decreasing = TRUE), -->
<!-- ##                          5)), c('AluY', 'AluSx', 'L1PA11', 'MIR', 'L2a')) -->

<!-- ## maybe with variable features? -->

<!-- for (run in names(d)[sapply(d, function(x) return('repeats_featurecounts' %in% names(x)))]) { -->
<!--     cat("## ", run, "\n\n") -->
<!--     repeats_featurecounts <- as.data.frame(FetchData( -->
<!--         combined[[run]]$repeats_featurecounts, -->
<!--         vars = head(VariableFeatures(combined[[run]]$repeats_featurecounts), 10) , slot = 'data')) -->
        
<!--     repeats_featurecounts$cell <- rownames(repeats_featurecounts) -->
        
<!--     combined[[run]]$genes@meta.data$cell <- rownames(combined[[run]]$genes@meta.data) -->
<!--     repeats_featurecounts <- merge(combined[[run]]$genes@meta.data[, c(res, 'cell') ], -->
<!--                      repeats_featurecounts, by = 'cell', all.x = TRUE) -->
    
<!--     rownames(repeats_featurecounts) <- repeats_featurecounts$cell -->
    
<!--     repeats_featurecounts <- melt(repeats_featurecounts, id.vars = c('cell', res)) -->
    
<!--     colnames(repeats_featurecounts) <- c('cell', res, 'repfamily', 'CPM') -->

<!--     print(ggplot(repeats_featurecounts, -->
<!--            aes(x = get(res), y = CPM)) +  -->
<!--           geom_boxplot(outlier.size=1) + -->
<!--           scale_y_continuous(trans='log10') + -->
<!--         scale_fill_discrete(drop=FALSE) + -->
<!--         facet_wrap(~ factor(repfamily), ncol = 10) + -->
<!--         theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--         ylab('repeat expression [CPM]') + -->
<!--         xlab('gene-based clustering') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_featurecounts', res))) -->
<!--     cat('\n\n') -->
<!-- } -->

<!-- ``` -->

<!-- # Most variable repeats featurecounts by identity {.tabset} -->


<!-- ```{r, fig.height = 5, fig.width = 15, results = 'asis'} -->
<!-- scina <- 'scina' -->
<!-- ## agg <- list() -->

<!-- for (run in names(d)[sapply(d, function(x) return('repeats_featurecounts' %in% names(x)))]) { -->
<!--     cat("## ", run, "\n\n") -->
    
<!--     repeats_featurecounts <- as.data.frame(FetchData( -->
<!--         combined[[run]]$repeats_featurecounts, -->
<!--         vars = head(VariableFeatures(combined[[run]]$repeats_featurecounts), 10) , slot = 'data')) -->


    
<!--     repeats_featurecounts$cell <- rownames(repeats_featurecounts) -->

<!--     combined[[run]]$genes@meta.data$cell <- rownames(combined[[run]]$genes@meta.data) -->
<!--     repeats_featurecounts <- merge(combined[[run]]$genes@meta.data[, c(scina, 'cell') ], -->
<!--                      repeats_featurecounts, by = 'cell', all.x = TRUE) -->
    
<!--     rownames(repeats_featurecounts) <- repeats_featurecounts$cell -->
    
<!--     repeats_featurecounts <- melt(repeats_featurecounts, id.vars = c('cell', scina)) -->
    
<!--     colnames(repeats_featurecounts) <- c('cell', scina, 'repfamily', 'CPM') -->

<!--     print(ggplot(repeats_featurecounts, -->
<!--            aes(x = scina, y = CPM)) +  -->
<!--           geom_boxplot(outlier.size=1) + -->
<!--           scale_y_continuous(trans='log10') + -->
<!--         scale_fill_discrete(drop=FALSE) + -->
<!--         facet_wrap(~ factor(repfamily), ncol = 10) + -->
<!--         theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--         ylab('repeat expresssion [CPM]') + -->
<!--         xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_featurecounts', scina))) -->
<!--     cat('\n\n') -->

<!-- } -->

<!-- ``` -->

<!-- # All integrated (repeats, scina) -->

<!-- ```{r fig.width = 14, fig.height = 14} -->

<!-- ## first aggregate all repeats_featurecounts stuff -->
<!-- sds <- lapply(d[sapply(d, function(x) return('repeats_featurecounts' %in% names(x)))], -->
<!--               function(x) return(apply(as.data.frame(GetAssayData(x$repeats_featurecounts, -->
<!--                                                         slot = 'data')), 1, sd))) -->

<!-- ## get the ones with the highest average sd -->

<!-- mean_sd <- apply(do.call(cbind.data.frame, sds), 1, mean) -->
<!-- selected <- names(head(sort(mean_sd, decreasing = TRUE), 20)) -->
<!-- ## selected -->


<!-- ## aggregate(agg['CPM'], by = list(agg$repfamily, agg$run),  FUN = function(x) c(mean = mean(x, na.rm = TRUE), n = length(x), sd = sd(x, na.rm = TRUE) )) -->

<!-- ## now get these data for selected features, aggregate and plot -->

<!-- tmp <- list() -->
<!-- for (run in names(d)) { -->
<!--     foo <- data.frame(run = run, -->
<!--                              FetchData(combined[[run]]$repeats_featurecounts, vars = selected, slot = 'data')) -->

<!--     ## add annotations -->
<!--     bar <- data.frame(cell = rownames(foo), -->
<!--                       scina = combined[[run]]$repeats_featurecounts@meta.data[rownames(foo), 'scina']) -->

<!--     stopifnot(rownames(foo) == bar$cell) -->
    
    
<!--     tmp[[run]] <- data.frame(foo, scina = bar$scina) -->
<!--     rm(foo, bar) -->
<!-- } -->

<!-- tmp <- do.call(rbind.data.frame, tmp) -->
<!-- str(tmp) -->

<!-- ## melt -->
<!-- repeats_featurecounts <- melt(tmp, id.vars = c('run', 'scina')) -->
<!-- head(repeats_featurecounts)     -->
<!-- colnames(repeats_featurecounts) <- c('run', 'scina', 'repfamily', 'CPM') -->

<!-- ``` -->

<!-- ```{r, fig.width = 12, fig.height = 12} -->
<!-- print(ggplot(repeats_featurecounts, -->
<!--              aes(x = scina, y = CPM, fill = run)) +  -->
<!--       geom_boxplot(outlier.size=1) + -->
<!--       scale_y_continuous(trans='log10') + -->
<!--       scale_fill_discrete(drop=FALSE) + -->
<!--       facet_wrap(~ factor(repfamily), ncol = 4) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--       ylab('repeat expresssion [CPM]') + -->
<!--       xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_featurecounts', scina))) -->
<!-- ``` -->


<!-- ```{r, fig.width = 8, fig.height = 14} -->
<!-- print(ggplot(repeats_featurecounts, -->
<!--              aes(x = repfamily, y = CPM, fill = run)) +  -->
<!--       geom_boxplot(outlier.size=1) + -->
<!--       scale_y_continuous(trans='log10') + -->
<!--       scale_fill_discrete(drop=FALSE) + -->
<!--       facet_grid(factor(scina)~ factor(run)) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--       ylab('repeat expresssion [CPM]') + -->
<!--       xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats_featurecounts', scina))) -->

<!-- ``` -->

<!-- # Cluster by average expression (repeats, SCINA) -->

<!-- Avoiding non-determined cell types -->

<!-- ```{r, results = 'asis', fig.height = 10, fig.width = 10} -->
<!-- repeats_featurecounts <- repeats_featurecounts[repeats_featurecounts$scina != 'unknown',] -->
<!-- foo <- reshape2::dcast(repeats_featurecounts, run + scina ~ repfamily, mean, value.var = 'CPM') -->


<!-- rownames(foo) <- paste(foo$run, foo$scina, sep = '_') -->

<!-- annot <- foo[,c('run', 'scina')] -->
<!-- rownames(annot) <- paste(annot$run, annot$scina, sep = '_') -->
<!-- foo <- foo[,-(1:2)] -->

<!-- pheatmap(foo, annotation_row = annot, show_rownames = FALSE) -->

<!-- pheatmap(t(foo), annotation_col = annot, show_colnames = FALSE) -->
<!-- ``` -->




<!-- # Pheatmaps -->

<!-- Note the cell sampling: same cells in all plots; `annotation_col` comes from genes' clustering. -->

<!-- ```{r, results = 'asis', dpi=50, fig.width = 15, fig.height = 15} -->
<!-- NCELLS <- 1000 -->
<!-- NFEATURES <- 500 -->

<!-- ## stop() -->
<!-- for (run in names(d)) { -->
<!--     cat("## ", run, "{.tabset}\n\n") -->
    
<!--     set.seed(654) -->
<!--     idx_cell <- sample(colnames(GetAssayData(combined[[run]][[1]])), NCELLS, replace = FALSE) -->
    
<!--     for (item in names(combined[[run]])) { -->
<!--         cat("### ", item, "{.tabset}\n\n") -->

<!--         set.seed(464) -->
<!--         idx_feat <- sample(VariableFeatures(combined[[run]][[item]]), NFEATURES, replace = FALSE) -->
        
<!--         print(pheatmap(GetAssayData( -->
<!--             object = subset(combined[[run]][[item]], -->
<!--                             features = idx_feat, -->
<!--                             cells = intersect(idx_cell, colnames(GetAssayData(combined[[run]][[item]])))), -->
<!--             slot = 'scale.data'), -->
<!--             main = sprintf('%s %s', run, item), -->
<!--             color = viridis(100), -->
<!--             cluster_rows = TRUE, -->
<!--             cluster_cols = TRUE, -->
<!--             show_rownames = FALSE, -->
<!--             show_colnames = FALSE, -->
<!--             clustering_distance_cols = "euclidean", -->
<!--             clustering_method = "ward.D2", -->
<!--             clustering_distance_rows = "euclidean", -->
<!--             fontsize_row = 12, -->
<!--             annotation_col = as.data.frame(combined[[run]]$genes_cellranger@meta.data[idx_cell, c(res), -->
<!--                                                                                drop = FALSE]), -->
<!--             scale = "none")) -->
<!--         cat('\n\n') -->
<!--     } -->
<!--     cat('\n\n') -->
<!-- } -->
<!-- ``` -->

<!-- ```{r} -->
<!-- knitr::exit() -->
<!-- ``` -->



<!-- # ARIS -->

<!-- ## Data load -->

<!-- ```{r} -->

<!-- aris_df <- NULL -->

<!-- for (rds in params) { -->
<!--     ## the params are RDS objects from Seurat -->
<!--     ## aris[[sapply(strsplit(dirname(rds), .Platform$file.sep), -->
<!--     ##       function(x) return(x[6]))]] <- list() -->
<!--     fns <- list.files(dirname(rds), pattern = 'aris.*rds') -->
<!--     for (fn in fns) { -->
<!--         curr_nest <- readRDS(file.path(dirname(rds), fn)) -->
<!--         ## curr <- do.call(function(x) rbind.data.frame(x, setNames = c('cat1' , 'cat2', 'ari')), -->
<!--         ##                 curr_nest) -->
<!--         ## aris_df <- NULL -->
<!--         for (item in names(curr_nest)) { -->
<!--             curr <- curr_nest[[item]] -->
<!--             colnames(curr) <- c('set1', 'set2', 'ari') -->
<!--             curr$origin <- item -->

<!--             parsed <- strcapture( -->
<!--                 pattern = "aris_regress_nCount_(.*)_nFeature_(.*).rds", -->
<!--                 x = fn, -->
<!--                 proto = list(ncount = logical(), nfeature = logical())) -->
            
<!--             curr$ncount_scaling <- parsed$ncount -->
<!--             curr$nfeature_scaling <- parsed$nfeature -->
            
<!--             curr$sample <- sapply(strsplit(dirname(rds), .Platform$file.sep), -->
<!--                                   function(x) return(x[6])) -->
<!--             aris_df <- rbind(aris_df, curr) -->

<!--         } -->

<!--         ## aris[[sapply(strsplit(dirname(rds), .Platform$file.sep), -->
<!--         ##              function(x) return(x[6]))]][[basename(fn)]] <- curr_nest -->
<!--     } -->
<!-- } -->

<!-- summary(aris_df) -->
<!-- rm(parsed, curr) -->
<!-- ``` -->

<!-- Extract resolutions and n for clustering -->

<!-- ```{r notscina, eval = TRUE} -->
<!-- parsed <- strcapture( -->
<!--     pattern = "(.*)_RNA_snn_res.(.*) \\(n=(.*)\\)", -->
<!--     x = aris_df$set1, -->
<!--     proto = list(id1 = character(), -->
<!--                  res1 = numeric(), -->
<!--                  size1 = numeric())) -->

<!-- aris_df <- cbind(aris_df, parsed) -->
<!-- rm(parsed) -->

<!-- parsed <- strcapture( -->
<!--     pattern = "(.*)_RNA_snn_res.(.*) \\(n=(.*)\\)", -->
<!--     x = aris_df$set2, -->
<!--     proto = list(id2 = character(), -->
<!--                  res2 = numeric(), -->
<!--                  size2 = numeric())) -->

<!-- aris_df <- cbind(aris_df, parsed) -->
<!-- rm(parsed) -->

<!-- ## head(aris_df) -->
<!-- ``` -->

<!-- Extract resolutions and n for scina -->

<!-- ```{r eval = TRUE} -->
<!-- parsed <- strcapture( -->
<!--     pattern = "scina_scina \\(n=(.*)\\)", -->
<!--     x = aris_df$set2, -->
<!--     proto = list(size_scina = numeric())) -->

<!-- aris_df <- cbind(aris_df, parsed) -->
<!-- rm(parsed) -->


<!-- ## head(aris_df) -->
<!-- ``` -->



<!-- ## ARIs comparison {.tabset} -->

<!-- Plot ARIs decay with resolution, scaling etc -->

<!-- ```{r, fig.width=12, fig.height=8, results = 'asis'} -->

<!-- aris_df$regressed_out <- 'none' -->

<!-- aris_df$regressed_out[aris_df$ncount_scaling & aris_df$nfeature_scaling] <- 'both' -->
<!-- aris_df$regressed_out[aris_df$ncount_scaling & !aris_df$nfeature_scaling] <- 'genes_ncount' -->
<!-- aris_df$regressed_out[!aris_df$ncount_scaling & aris_df$nfeature_scaling] <- 'genes_nfeature' -->

<!-- aris_df$regressed_out <- factor(aris_df$regressed_out, -->
<!--                                    levels = c('none', 'genes_ncount', 'genes_nfeature', 'both')) -->

<!-- ``` -->

<!-- ```{r, fig.width=12, fig.height=8, results = 'asis'} -->

<!-- aris_df$label1 <- sprintf('res%s_(n=%s)', aris_df$res1, aris_df$size1) -->
<!-- aris_df$label1[grepl('NA', aris_df$label1)] <- NA -->

<!-- aris_df$label2 <- sprintf('res%s_(n=%s)', aris_df$res2, aris_df$size2) -->
<!-- aris_df$label2[grepl('NA', aris_df$label2)] <- NA -->


<!-- ## cat('### genes vs repeats cr\n\n') -->

<!-- ## print(ggplot(aris_df[aris_df$id1 == 'genes' & aris_df$id2 == 'repeats_cr',], -->
<!-- ##              aes(x = res1, y = ari, col = as.factor(regressed_out))) + -->
<!-- ##       geom_point() + -->
<!-- ##       labs(title = "genes cellranger vs repeats cellranger (from repeats bam)", -->
<!-- ##            subtitle = "by repeats resolution", y = "ARI", x = "genes resolution", -->
<!-- ##            color = 'regressed out') + -->
<!-- ##       facet_grid(sample~paste0('repeats res.\n', res2)) + -->
<!-- ##       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->

<!-- ## cat('\n\n') -->

<!-- cat('### genes vs repeats fc\n\n') -->

<!-- print(ggplot(aris_df[aris_df$id1 == 'genes' & aris_df$id2 == 'repeats_fc',], -->
<!--              aes(x = res1, y = ari, col = as.factor(regressed_out))) + -->
<!--       geom_point() + -->
<!--       labs(title = "genes cellranger vs repeats featurecounts (from genes bam)", -->
<!--            subtitle = "by repeats resolution", y = "ARI", x = "genes resolution", -->
<!--            color = 'regressed out') + -->
<!--       facet_grid(sample~paste0('repeats res.\n', res2)) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->

<!-- cat('\n\n') -->

<!-- cat('### repeats fc vs repeats cr\n\n') -->

<!-- print(ggplot(aris_df[aris_df$id1 == 'repeats_fc' & aris_df$id2 == 'repeats_cr',], -->
<!--              aes(x = res1, y = ari, col = as.factor(regressed_out))) + -->
<!--       geom_point() + -->
<!--       labs(title = "repeats featurecounts (from genes bam) vs repeats cellranger (repeats bam)", -->
<!--            subtitle = "by repeats cellranger resolution", y = "ARI", x = "repeats fc resolution", -->
<!--            color = 'regressed out') + -->
<!--       facet_grid(sample~paste0('repeats cr res.\n', res2)) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->

<!-- cat('\n\n') -->
<!-- ``` -->


<!-- Maybe add the number of clusters? as a size param? -->


<!-- Rather, plot both size and res -->

<!-- ```{r, eval = TRUE, fig.height = 5, fig.with = 15} -->
<!-- ## print(ggplot(aris_df[aris_df$id1 == 'repeats_fc' & aris_df$id2 == 'repeats_cr',], -->
<!-- ##              aes(x = res1, y = ari, col = as.factor(regressed_out))) + -->
<!-- ##       geom_point() + -->
<!-- ##       labs(title = "repeats featurecounts (from genes bam) vs repeats cellranger (repeats bam)", -->
<!-- ##            subtitle = "by repeats cellranger resolution", y = "ARI", x = "repeats fc resolution", -->
<!-- ##            color = 'regressed out') + -->
<!-- ##       facet_grid(sample~paste0('repeats cr res.\n', res2)) + -->
<!-- ##       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->

<!-- print(ggplot(aris_df, -->
<!--              aes(x = res1, y = size1, col = as.factor(regressed_out))) + -->
<!--       geom_point() + -->
<!--       facet_grid(sample~id1) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->

<!-- print(ggplot(aris_df, -->
<!--              aes(x = res2, y = size2, col = as.factor(regressed_out))) + -->
<!--       geom_point() + -->
<!--       facet_grid(sample~id2) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->
<!-- ``` -->

<!-- # ARIs flavours vs SCINA -->


<!-- ```{r, fig.height = 5, fig.with = 15} -->

<!-- print(ggplot(aris_df[grep('scina', aris_df$origin),], -->
<!--              aes(x = res1, y = ari, col = as.factor(regressed_out) )) + -->
<!--       geom_point() + -->
<!--       facet_grid(sample~id1) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->

<!-- ``` -->


<!-- ```{r} -->

<!-- print(ggplot(aris_df[grep('scina', aris_df$origin),], -->
<!--              aes(x = res1, y = size1, col = as.factor(regressed_out) )) + -->
<!--       geom_point() + -->
<!--       facet_grid(sample~id1) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->

<!-- ``` -->


<!-- ```{r} -->
<!-- print(ggplot(aris_df[grep('scina', aris_df$origin),], -->
<!--              aes(x = res1, y = ari, col = as.factor(regressed_out) )) + -->
<!--       geom_point() + -->
<!--       facet_grid(sample~id1) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->

<!-- ``` -->

<!-- # All variable features and all repeats methods at once, using scina {.tabset} -->

<!-- ## With repeats including control not overlapping genes -->

<!-- ```{r} -->

<!-- NCELLS = 200 ## cells per run:sample -->
<!-- NVARIABLE = 20 ## variable repeats features (per run:sample) -->

<!-- cr <- list() -->
<!-- for (flavour in c('repeats_cellranger', 'repeats_star', 'repeats_featurecounts', -->
<!--                   'control_repeats_only')) { -->
<!--     for (run in names(d)[sapply(d, function(x) return(flavour %in% names(x)))]) { -->

<!--         cr[[paste(flavour, run)]] <- data.frame( -->
<!--             rep = head(VariableFeatures(combined[[run]][[flavour]]), NVARIABLE), -->
<!--             flavour = flavour, -->
<!--             run = run, -->
<!--             stringsAsFactors = FALSE) -->
<!--     }     -->
<!-- } -->

<!-- cr <- do.call(rbind.data.frame, cr) -->
<!-- table(cr$flavour, cr$run) -->

<!-- variable <- list() -->
<!-- for (flavour in c('repeats_cellranger', 'repeats_star', 'repeats_featurecounts', 'control_repeats_only')) { -->
<!--     for (run in names(d)[sapply(d, function(x) return(flavour %in% names(x)))]) { -->
<!--         idx <- sample(colnames(combined[[run]][[flavour]]), size = NCELLS, replace = FALSE) -->
<!--         variable[[paste(run, flavour, sep = '.')]] <- GetAssayData(combined[[run]][[flavour]], -->
<!--                             slot = 'data')[unique(cr$rep),idx] -->
<!--     } -->
<!-- } -->

<!-- foo <- do.call(cbind.data.frame, variable) -->

<!-- colannot <- data.frame(id = colnames(foo), -->
<!--                        sample = sapply(strsplit(colnames(foo), '.', fixed = TRUE), -->
<!--                                         function(x) return(x[1])), -->
<!--                        flavour = sapply(strsplit(colnames(foo), '.', fixed = TRUE), -->
<!--                                        function(x) return(x[2])), -->
<!--                        cell = sapply(strsplit(colnames(foo), '.', fixed = TRUE), -->
<!--                                      function(x) return(x[3])), -->
<!--                        scina = NA, -->
<!--                        stringsAsFactors = FALSE) -->


<!-- head(colannot) -->

<!-- for ( i in 1:nrow(colannot)) { -->
<!--     colannot[i, 'scina'] <- as.character(combined[[colannot[i, 'sample']]][[colannot[i, 'flavour']]]@meta.data[colannot[i, 'cell'], 'scina']) -->

    
<!-- } -->

<!-- rownames(colannot) <- colannot$id -->

<!-- table(colannot$scina) -->


<!-- ## remove any repeat whose annotation by scina is 'unknown' -->
<!-- colannot <- colannot[colannot$scina != 'unknown',] -->
<!-- foo <- foo[,colannot$id] -->
<!-- dim(foo) -->

<!-- ## remove any repeat with 0 sd after sampling -->
<!-- sum(is.infinite(unlist(foo))) -->
<!-- foo <- foo[apply(foo, 1, sd) > 0,] -->

<!-- ``` -->

<!-- ```{r, fig.height = 20, fig.width = 20} -->


<!-- newCols <- colorRampPalette(grDevices::rainbow(length(unique(colannot$scina)))) -->
<!-- mycolors <- newCols(length(unique(colannot$scina))) -->
<!-- names(mycolors) <- unique(colannot$scina) -->
<!-- mycolors <- list(scina = mycolors) -->

<!-- print(pheatmap(foo, -->
<!--                color = viridis(100), -->
<!--             cluster_rows = TRUE, -->
<!--             cluster_cols = TRUE, -->
<!--             show_rownames = TRUE, -->
<!--             show_colnames = FALSE, -->
<!--             clustering_distance_cols = "euclidean", -->
<!--             clustering_method = "ward.D2", -->
<!--             clustering_distance_rows = "euclidean", -->
<!--             fontsize_row = 12, -->
<!--             annotation_col = colannot[,c('flavour', 'sample', 'scina')], -->
<!--             annotation_colors = mycolors, -->
<!--             scale = "none")) -->



<!-- ``` -->

<!-- ## Without non-overlapping genes repeats control (biased by transcriptome) -->

<!-- ```{r} -->

<!-- NCELLS = 200 ## cells per run:sample -->
<!-- NVARIABLE = 20 ## variable repeats features (per run:sample) -->

<!-- cr <- list() -->
<!-- for (flavour in c('repeats_cellranger', 'repeats_star', 'repeats_featurecounts')) { -->
<!--     for (run in names(d)[sapply(d, function(x) return(flavour %in% names(x)))]) { -->

<!--         cr[[paste(flavour, run)]] <- data.frame( -->
<!--             rep = head(VariableFeatures(combined[[run]][[flavour]]), NVARIABLE), -->
<!--             flavour = flavour, -->
<!--             run = run, -->
<!--             stringsAsFactors = FALSE) -->
<!--     }     -->
<!-- } -->

<!-- cr <- do.call(rbind.data.frame, cr) -->
<!-- table(cr$flavour, cr$run) -->

<!-- variable <- list() -->
<!-- for (flavour in c('repeats_cellranger', 'repeats_star', 'repeats_featurecounts')) { -->
<!--     for (run in names(d)[sapply(d, function(x) return(flavour %in% names(x)))]) { -->
<!--         idx <- sample(colnames(combined[[run]][[flavour]]), size = NCELLS, replace = FALSE) -->
<!--         variable[[paste(run, flavour, sep = '.')]] <- GetAssayData(combined[[run]][[flavour]], -->
<!--                             slot = 'data')[unique(cr$rep),idx] -->
<!--     } -->
<!-- } -->

<!-- foo <- do.call(cbind.data.frame, variable) -->

<!-- colannot <- data.frame(id = colnames(foo), -->
<!--                        sample = sapply(strsplit(colnames(foo), '.', fixed = TRUE), -->
<!--                                         function(x) return(x[1])), -->
<!--                        flavour = sapply(strsplit(colnames(foo), '.', fixed = TRUE), -->
<!--                                        function(x) return(x[2])), -->
<!--                        cell = sapply(strsplit(colnames(foo), '.', fixed = TRUE), -->
<!--                                      function(x) return(x[3])), -->
<!--                        scina = NA, -->
<!--                        stringsAsFactors = FALSE) -->


<!-- head(colannot) -->

<!-- for ( i in 1:nrow(colannot)) { -->
<!--     colannot[i, 'scina'] <- as.character(combined[[colannot[i, 'sample']]][[colannot[i, 'flavour']]]@meta.data[colannot[i, 'cell'], 'scina']) -->

    
<!-- } -->

<!-- rownames(colannot) <- colannot$id -->

<!-- table(colannot$scina) -->


<!-- ## remove any repeat whose annotation by scina is 'unknown' -->
<!-- colannot <- colannot[colannot$scina != 'unknown',] -->
<!-- foo <- foo[,colannot$id] -->
<!-- dim(foo) -->

<!-- ## remove any repeat with 0 sd after sampling -->
<!-- sum(is.infinite(unlist(foo))) -->
<!-- foo <- foo[apply(foo, 1, sd) > 0,] -->

<!-- ``` -->

<!-- ```{r, fig.height = 20, fig.width = 20} -->


<!-- newCols <- colorRampPalette(grDevices::rainbow(length(unique(colannot$scina)))) -->
<!-- mycolors <- newCols(length(unique(colannot$scina))) -->
<!-- names(mycolors) <- unique(colannot$scina) -->
<!-- mycolors <- list(scina = mycolors) -->

<!-- print(pheatmap(foo, -->
<!--                color = viridis(100), -->
<!--             cluster_rows = TRUE, -->
<!--             cluster_cols = TRUE, -->
<!--             show_rownames = TRUE, -->
<!--             show_colnames = FALSE, -->
<!--             clustering_distance_cols = "euclidean", -->
<!--             clustering_method = "ward.D2", -->
<!--             clustering_distance_rows = "euclidean", -->
<!--             fontsize_row = 12, -->
<!--             annotation_col = colannot[,c('flavour', 'sample', 'scina')], -->
<!--             annotation_colors = mycolors, -->
<!--             scale = "none")) -->



<!-- ``` -->

<!-- # F1s -->

<!-- ```{r} -->
<!-- cobra <- list() -->


<!-- for (rds in params) { -->
<!--     fn <- list.files(dirname(rds), pattern = '.*cobra.*TRUE.*TRUE.*') -->
   
    
<!--     cobra[[sapply(strsplit(dirname(rds), .Platform$file.sep), -->
<!--                   function(x) return(x[6]))]] <- readRDS(file = file.path(dirname(rds), fn)) -->
<!-- } -->

<!-- cobra <- do.call(rbind.data.frame, cobra) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- ## some classes are equivalent -->
<!-- cobra$flavour <- as.character(cobra$flavour) -->
<!-- cobra$flavour[cobra$flavour == 'control_repeats_only_multi'] <- 'repeats_only_multi' -->
<!-- cobra$flavour[cobra$flavour == 'control_repeats_only_unique'] <- 'repeats_only_unique' -->

<!-- table(cobra$flavour) -->
<!-- ``` -->


<!-- ```{r} -->
<!-- pal <- read.table(text = 'identity,color -->
<!-- CD4+_T_cell,"#A6CEE3" -->
<!-- Cytotoxic_T_cell,"#1F78B4" -->
<!-- B_cell,"#B2DF8A" -->
<!-- Natural_killer_cell,"#33A02C" -->
<!-- CD14+_monocyte,"#FB9A99" -->
<!-- CD16+_monocyte,"#E31A1C" -->
<!-- Dendritic_cell,"#FDBF6F" -->
<!-- Plasmacytoid_dendritic_cell,"#FF7F00" -->
<!-- Plasma_cell,"#CAB2D6" -->
<!-- Platelet,"#6A3D9A" -->
<!-- unknown,"#C0C0C0"', header = TRUE, sep = ',', stringsAsFactors = FALSE) -->

<!-- rownames(pal) <- pal$identity -->
<!-- ``` -->


<!-- ```{r, fig.width = 17, fig.height = 6} -->

<!-- ggplot(data = cobra,       -->
<!--     aes(x = res, y =  F1, group=flavour, col = cell)) + -->
<!--     geom_point() + -->
<!--     facet_grid(id~flavour) +  -->
<!--     theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--     scale_colour_manual(values = setNames(pal$color, as.character(pal$identity)))  -->

<!-- ``` -->

<!-- Add whether the analyte is repeat, genes, or both -->


<!-- ```{r} -->
<!-- cobra$analyte <- NA -->
<!-- cobra[grepl('genes', cobra$flavour), 'analyte'] <- 'genes' -->
<!-- cobra[grepl('repeats', cobra$flavour), 'analyte'] <- 'repeats' -->
<!-- cobra[grepl('both', cobra$flavour), 'analyte'] <- 'both' -->


<!-- table(cobra$analyte, useNA = 'always') -->

<!-- ``` -->


<!-- Add whether multimappers are allowed -->


<!-- ```{r} -->
<!-- cobra$multimapper <- 'NA' -->

<!-- cobra[grepl('multi', cobra$flavour), 'multimapper'] <- 'multi' -->
<!-- cobra[grepl('unique', cobra$flavour), 'multimapper'] <- 'unique' -->

<!-- table(cobra$multimapper, useNA = 'always') -->
<!-- ``` -->

<!-- Add whether repeats are overlapping genes or not -->

<!-- ```{r} -->
<!-- cobra$overlap <- 'overlap_allowed' -->

<!-- cobra[grepl('genes|both', cobra$flavour), 'overlap'] <- 'genes' -->
<!-- cobra[grepl('only', cobra$flavour), 'overlap'] <- 'no_overlap' -->


<!-- table(cobra$overlap, useNA = 'always') -->
<!-- ``` -->



<!-- ```{r, fig.width = 13, fig.height = 6} -->

<!-- ggplot(data = cobra,       -->
<!--     aes(x = res, y =  F1, group=flavour, col = flavour)) + -->
<!--     geom_point() + -->
<!--     facet_grid(id~cell) +  -->
<!--     theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--     xlab('unsup. clust. resolution') + -->
<!--     ylab('F1 score') -->

<!-- ``` -->


<!-- ```{r, fig.width = 13, fig.height = 6} -->

<!-- ggplot(data = cobra,       -->
<!--     aes(x = res, y =  F1, group=flavour, col = analyte)) + -->
<!--     geom_point(aes(shape= analyte, col = analyte)) + -->
<!--     facet_grid(id~cell) +  -->
<!--     theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--     xlab('unsup. clust. resolution') + -->
<!--     ylab('F1 score') -->

<!-- ``` -->


<!-- ```{r, fig.width = 13, fig.height = 6} -->

<!-- ggplot(data = cobra,       -->
<!--        aes(x = res, y =  F1, group=flavour, col = multimapper)) + -->
<!--     geom_point(aes(shape= multimapper, col = multimapper)) + -->
<!--     facet_grid(id~cell) +  -->
<!--     theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--     xlab('unsup. clust. resolution') + -->
<!--     ylab('F1 score') -->

<!-- ``` -->


<!-- ```{r, fig.width = 12, fig.height = 6} -->

<!-- ggplot(data = cobra,       -->
<!--     aes(x = res, y =  F1, group=flavour, col = overlap)) + -->
<!--     geom_point(aes(shape= overlap, col = overlap)) + -->
<!--     facet_grid(id~cell) +  -->
<!--     theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--     xlab('unsup. clust. resolution') + -->
<!--     ylab('F1 score') -->

<!-- ``` -->

<!-- ## T cells vs Monocytes only -->

<!-- ```{r, eval = FALSE} -->
<!-- ggplot(data = cobra[cobra$cell %in% c('CD14+_monocyte', 'CD16+_monocyte', 'Cytotoxic_T_cell', -->
<!--                                       'CD4+_T_cell', 'Plasmacytoid_dendritic_cell') & -->
<!--                     cobra$flavour %in% c('genes_cellranger', 'repeats_cellranger', -->
<!--                                          '',],       -->
<!--     aes(x = res, y =  F1, group=flavour, col = cell)) + -->
<!--     geom_point() + -->
<!--     facet_grid(id~flavour) +  -->
<!--     theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--     scale_colour_manual(values = setNames(pal$color, as.character(pal$identity)))  -->
<!-- ``` -->

<!-- # Markers {.tabset} -->

<!-- For each sample, each flavour, and each celltype, get the significant markers. Then, look for overlaps. -->


<!-- ```{r} -->
<!-- markers <- list() -->

<!-- for (rds in params) {     -->
<!--     ## the params are RDS objects from Seurat -->
<!--     ## aris[[sapply(strsplit(dirname(rds), .Platform$file.sep), -->
<!--     ##       function(x) return(x[6]))]] <- list() -->
<!--     fns <- list.files(dirname(rds), pattern = 'markers.*TRUE.*TRUE.*rds') -->
<!--     for (fn in fns) { -->
<!--         curr_nest <- readRDS(file.path(dirname(rds), fn)) -->
<!--         for (item in names(curr_nest)) { -->
<!--             curr_nest[[item]]$flavour <- item -->
<!--             curr_nest[[item]]$dataset <- sapply(strsplit(dirname(rds), '/'), -->
<!--                                                 function(x) return(x[6])) -->
<!--         } -->

<!--         curr_nest <- lapply(curr_nest, function(x) x %>% -->
<!--                                                    group_by(cluster)) -->
        
<!--         markers[[fn]] <- do.call(rbind.data.frame, curr_nest) -->
<!--         markers[[fn]] <- markers[[fn]][markers[[fn]]$p_val_adj < 0.05,] -->
<!--     } -->
<!-- } -->
<!-- markers <- do.call(rbind.data.frame, markers) -->


<!-- colnames(markers) -->
<!-- table(markers$cluster) -->
<!-- table(markers$flavour) -->
<!-- table(markers$dataset) -->
<!-- ``` -->


<!-- ```{r, fig.width = 8, fig.height=7} -->

<!-- markers <- as.data.frame(markers) -->


<!-- for (celltype in unique(markers$cluster)) { -->
<!--     cat('## ', celltype, '\n\n') -->
<!--     ul <- list() -->
<!--     for (dataset in unique(markers[markers$cluster == celltype, 'dataset'])) { -->

<!--         for (flavour in grep('repeats', unique(markers[ markers$cluster == celltype & -->
<!--                                                         markers$dataset == dataset, 'flavour']), -->
<!--                              value = TRUE)) { -->
                    
<!--             ul[[paste(celltype, dataset, flavour, sep = "_")]] <- markers[ -->
<!--                 markers$cluster == celltype & -->
<!--                 markers$dataset == dataset & -->
<!--                 markers$flavour == flavour, 'gene'] -->
<!--         } -->
<!--     } -->
<!--     print(upset(fromList(ul), nsets = 10,  text.scale = 0.8)) -->
<!--     grid.text(paste(celltype) ,x = 0.65, y=0.95, gp=gpar(fontsize = 10)) -->
    
<!--     ## rm(ul) -->
<!--     cat('\n\n') -->

<!-- } -->


<!-- str(ul) -->


<!-- ``` -->

<!-- # Markers II {.tabset} -->

<!-- Per celltype, concordance -->

<!-- ```{r, fig.width =8, fig.height=5} -->


<!-- ul <- list() -->
<!-- for (celltype in unique(markers$cluster)) { -->
<!--     cat('## ', celltype, '\n\n') -->

   
<!--     for (flavour in unique(markers[markers$cluster == celltype, 'flavour'])) { -->
<!--         ul <- list() -->
<!--         for (dataset in unique(markers[markers$cluster == celltype & -->
<!--                                      markers$flavour == flavour, 'dataset'])) { -->

<!--             ul[[paste(dataset, flavour, sep = "_")]] <- markers[ -->
<!--                 markers$flavour == flavour  & -->
<!--                 markers$cluster == celltype & -->
<!--                 markers$dataset == dataset, 'gene'] -->
   
        
<!--         } -->
<!--         ## str(ul) -->
        
<!--         print(upset(fromList(ul), nsets = length(ul))) -->
<!--         grid.text(paste(celltype, flavour) ,x = 0.65, y=0.95, gp=gpar(fontsize = 10)) -->

<!--     cat('\n\n') -->

<!--     } -->
<!-- } -->


<!-- ``` -->

<!-- # Markers III {.tabset} -->

<!-- By celltype, no matter which method -->

<!-- ```{r, fig.width =6, fig.height=4} -->


<!-- ul <- list() -->
<!-- for (celltype in unique(markers$cluster)) { -->
<!--     cat('## ', celltype, '\n\n') -->
    
<!--     ul[[celltype]] <- NULL -->
<!--     for (dataset in unique(markers[markers$cluster == celltype, 'dataset'])) { -->

<!--         print(dataset) -->
            
<!--         ul[[dataset]] <- unique(c(ul[[celltype]], markers[ grepl('repeats', markers$flavour) & -->
<!--                                                              markers$cluster == celltype & -->
<!--                                                              markers$dataset == dataset, 'gene'])) -->

         
<!--     } -->
    
   
<!--     print(upset(fromList(ul), nsets  = length(ul))) -->
<!--     grid.text(paste(celltype) ,x = 0.65, y=0.95, gp=gpar(fontsize = 10)) -->
<!--     cat('\n\n') -->
<!-- } -->




<!-- ``` -->

<!-- # Markers IV {.tabset} -->

<!-- By celltype, method influence -->

<!-- ```{r, fig.width =6, fig.height=4} -->



<!-- for (celltype in 'CD14+_monocyte') { -->
<!--     cat('## ', celltype, '\n\n') -->
    

<!--     for (dataset in unique(markers[markers$cluster == celltype, 'dataset'])) { -->

<!--         ul <- list() -->
<!--         for (flavour in unique(markers[markers$cluster == celltype & -->
<!--                                      markers$dataset == dataset, 'flavour'])) { -->
            
<!--             ul[[paste(dataset, flavour)]] <-  markers[markers$flavour == flavour & -->
<!--                                       markers$cluster == celltype & -->
<!--                                       markers$dataset == dataset, 'gene'] -->

<!--         } -->

<!--         print(upset(fromList(ul), nsets  = length(ul))) -->
<!--         grid.text(paste(celltype, dataset) ,x = 0.65, y=0.95, gp=gpar(fontsize = 10)) -->
<!--         cat('\n\n') -->

<!--     } -->
<!-- } -->






<!-- ``` -->


<!-- # Markers V -->

<!-- ```{r, fig.width =12, fig.height=5} -->
<!-- ft <- as.data.frame(table(markers[markers$cluster == 'CD14+_monocyte', 'gene'])) -->

<!-- ft <- ft[order(ft$Freq, decreasing = TRUE),] -->
<!-- head(ft) -->
<!-- selected <- as.character(head(ft, 20)$Var1) -->


<!-- ## get cpms from the seurat object, only 5 cells (first 5 indexed) -->
<!-- foo <- list() -->
<!-- for (item in names(d)) { -->
<!--     ## foo <- list() -->
<!--     for (flavour in grep('repeats', names(combined[[item]]), value = TRUE)) { -->
<!--         tmp <- as.data.frame(FetchData(combined[[item]][[flavour]], -->
<!--                                        vars = selected, -->
<!--                                        slot = 'data'))[1:5,] -->
<!--         tmp$cell <- rownames(tmp) -->

        
<!--         tmp <- merge(combined[[item]][[flavour]]@meta.data[, c('cell', 'scina')], -->
<!--                      tmp, by = 'cell', all.x = TRUE) -->
<!--         rownames(tmp) <- tmp$cell -->
<!--         tmp <- melt(tmp, id.vars = c('cell', 'scina')) -->
<!--         tmp$flavour <- flavour -->
<!--         tmp$dataset <-  item -->
<!--         foo[[paste(item, flavour)]] <- tmp -->
        
        
<!--     } -->
<!--     ## foo <- do.call(rbind.data.frame, foo) -->
<!-- } -->

<!-- foo <- do.call(rbind.data.frame, foo) -->
<!-- head(foo) -->


<!-- print(ggplot(foo[foo$scina %in% c('CD14+_monocyte', 'CD16+_monoctye', 'CD4+_T_cell'),], -->
<!--              aes(x = variable, y = value, fill = flavour)) +  -->
<!--       geom_boxplot() + -->
<!--       scale_y_continuous(trans='log10') + -->
<!--       scale_fill_discrete(drop=FALSE) + -->
<!--       facet_grid(dataset ~ scina) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->


<!-- ``` -->

<!-- # Cluster techniques and flavours and celltypes? -->

<!-- # Add bulk -->

<!-- ```{r} -->
<!-- bulk <- readRDS('/home/imallona/repeats_sc/runs/bulk_GSE80095/bulk_GSE80095_de.rds') -->
<!-- str(bulk) -->
<!-- names(bulk) -->

<!-- ## samplemonocyte-sampletcell -->
<!-- bulk_yardstick <- bulk[['repeats_star_multi']][[1]] -->
<!-- selected <- rownames(bulk_yardstick[bulk_yardstick$FDR < 0.05 & abs(bulk_yardstick$logFC) >= 2,]) -->
<!-- selected <- head(selected, 10) -->


<!-- ``` -->


<!-- ```{r, fig.width =12, fig.height=5} -->

<!-- ## get cpms from the seurat object, only 5 cells (first 5 indexed) -->
<!-- foo <- list() -->
<!-- for (item in names(d)) { -->
<!--     ## foo <- list() -->
<!--     for (flavour in grep('repeats', names(combined[[item]]), value = TRUE)) { -->
<!--         tmp <- as.data.frame(FetchData(combined[[item]][[flavour]], -->
<!--                                        vars = selected, -->
<!--                                        slot = 'data'))[1:5,] -->
<!--         tmp$cell <- rownames(tmp) -->

        
<!--         tmp <- merge(combined[[item]][[flavour]]@meta.data[, c('cell', 'scina')], -->
<!--                      tmp, by = 'cell', all.x = TRUE) -->
<!--         rownames(tmp) <- tmp$cell -->

<!--         tmp <- melt(tmp, id.vars = c('cell', 'scina')) -->
<!--         tmp <- tmp[grepl('T_cell|monocyte', tmp$scina),] -->
<!--         tmp$flavour <- flavour -->
<!--         tmp$dataset <-  item -->
<!--         foo[[paste(item, flavour)]] <- tmp -->
        
        
<!--     } -->
<!--     ## foo <- do.call(rbind.data.frame, foo) -->
<!-- } -->

<!-- foo <- do.call(rbind.data.frame, foo) -->
<!-- head(foo) -->
<!-- foo$scina <- as.character(foo$scina) -->
<!-- table(foo$scina) -->

<!-- print(ggplot(foo, -->
<!--              aes(x = variable, y = value, fill = flavour)) +  -->
<!--       geom_boxplot() + -->
<!--       scale_fill_discrete(drop=FALSE) + -->
<!--       facet_grid(dataset ~ scina) + -->
<!--       theme(axis.text.x = element_text(angle = 90, hjust = 1))) -->

<!-- ``` -->

# Timestamp

```{r sessionInfo, cache = FALSE}
date()
sessionInfo()
devtools::session_info()

```

