---
title: "Across samples report"
author: "Izaskun Mallona at Robinsonlab"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    code_download: true
    number_sections: true
    df_print: kable
    theme: lumen
    always_allow_html: yes
params:
  run1: ""
  run2: ""
  run3: ""
---


Rendered for samples:

```{r}
print(params)
```
# Config

```{r}
suppressPackageStartupMessages({
library(Matrix)

library(knitr)
library(Seurat)
## library(dplyr)
## library(scater)
library(data.table)
## library(SingleCellExperiment)
## library(scran)
## library(org.Mm.eg.db)
## suppressPackageStartupMessages(library(dplyr))
## suppressPackageStartupMessages(library(DT))
library(pheatmap)
## library(Cairo)
## library(RColorBrewer)
## library(corrplot)
library(ggplot2)
library(reshape2)
## library(xtable)
## library(mclust)
library(viridis)
})
```

```{r}
BASE <- file.path('/home', 'imallona', 'repeats_sc')
```

```{r}
opts_chunk$set(fig.width = 5,
               fig.height = 5,
               dev = 'png',
               cache = FALSE,
               echo = TRUE,
               include = TRUE,
               cache.lazy = FALSE,
               warning = TRUE,
               error = TRUE,
               message = FALSE)


options(bitmapType='cairo')
## getOption('bitmapType')
```


```{r functions}
## add alpha
ac <- function(col, alpha=1){
  apply(sapply(col, col2rgb)/255, 2, 
                     function(x) 
                       rgb(x[1], x[2], x[3], alpha=alpha))  
}

load_standard_cellranger_genes <- function(run_name) {
    ## barcodes_fn <- params$genes_cellranger ## should end with `barcodes.tsv.gz`
    barcodes_fn <- file.path(BASE, 'runs', run_name, 'cellranger_standard', 'outs',
                             'filtered_feature_bc_matrix', 'barcodes.tsv.gz')
    
    path <- dirname(barcodes_fn)

    genes <- CreateSeuratObject(counts = Read10X(data.dir = path),
                                project = 'cellranger_standard',
                                min.cells = 0,
                                min.features = 0)
    return(genes)
}

load_cellranger_repeats <-  function(run_name) {
    barcodes_fn <- file.path(BASE, 'runs', run_name, 'cellranger_repeats', 'outs',
                             'filtered_feature_bc_matrix', 'barcodes.tsv.gz')

    path <- dirname(barcodes_fn)

    repeats <- CreateSeuratObject(counts = Read10X(data.dir = path),
                                project = 'cellranger_repeats',
                                min.cells = 0,
                                min.features = 0)
    return(repeats)
}

load_featurecounts_repeats <-  function(run_name) {

    repeats <- file.path(BASE, "runs", run_name, 'count_repeats_on_cellranger_standard',
                                paste0(run_name, "_repeats.counts.gz"))

    tmp <- data.table::fread(repeats, sep = '\t', header = TRUE,
                             drop = c('Chr', 'Start', 'End', 'Strand', 'Length'))
    rownames(tmp) <- tmp$Geneid ## here

    colnames(tmp) <- gsub(".*:([A-Z]+)-1.bam","\\1", colnames(tmp))

    tmp <- as.data.frame(tmp)
    rownames(tmp) <- tmp$Geneid
    tmp <- tmp[,-1]

    featurecounts <- CreateSeuratObject(
        tmp,
        project = "seurat",
        assay = "RNA",
        min.cells = 0,
        min.features = 0,
        names.field = 1,
        names.delim = "_",
        meta.data = NULL)

    return(featurecounts)
}

coef_var <- function(x){
        (sd(x)/mean(x))*100
}


```

# Data load Seurat objects

```{r justfordebugging, eval = FALSE}
## params <- list(genes_cellranger = "/home/imallona/repeats_sc/runs/frozen_pbmc_donor_a/cellranger_standard/outs/filtered_feature_bc_matrix/barcodes.tsv.gz",
##                repeats_cellranger= "/home/imallona/repeats_sc/runs/frozen_pbmc_donor_a/cellranger_repeats/outs/filtered_feature_bc_matrix/barcodes.tsv.gz",
##                repeats_featurecounts= "/home/imallona/repeats_sc/runs/frozen_pbmc_donor_a/count_repeats_on_cellranger_standard/frozen_pbmc_donor_a_repeats.counts.gz")

## SRR10974769

## params <- list(run1 = '/home/imallona/repeats_sc/runs/SRR10974769/SRR10974769_pmbc_cellranger_regress_nCount_FALSE_nFeature_FALSE.rds',
##                run2 = '/home/imallona/repeats_sc/runs/5k_pbmc_v3/5k_pbmc_v3_pmbc_cellranger_regress_nCount_FALSE_nFeature_FALSE.rds',
##                run3 = '/home/imallona/repeats_sc/runs/frozen_pbmc_donor_a/frozen_pbmc_donor_a/summarize_cellranger_run_regress_nCount_FALSE_nFeature_FALSE.rds',
##                run4 = '/home/imallona/repeats_sc/runs/5k_pbmc_v3/5k_pbmc_v3_pmbc_cellranger_regress_nCount_FALSE_nFeature_FALSE.rds')

params <- list( run1 = '/home/imallona/repeats_sc/runs/frozen_pbmc_donor_a/frozen_pbmc_donor_a_pmbc_cellranger_regress_nCount_FALSE_nFeature_FALSE.rds',
        run2 = '/home/imallona/repeats_sc/runs/5k_pbmc_v3/5k_pbmc_v3_pmbc_cellranger_regress_nCount_FALSE_nFeature_FALSE.rds',
        run3 = '/home/imallona/repeats_sc/runs/SRR10974769/SRR10974769_pmbc_cellranger_regress_nCount_FALSE_nFeature_FALSE.rds',
        run4 = '/home/imallona/repeats_sc/runs/pbmc_10k_v3/pbmc_10k_v3_pbmc_cellranger_regress_nCount_FALSE_nFeature_FALSE.rds')
```


```{r}
d <- list()
```

```{r}
## cr cellranger
## fc featurecounts
for (rds in params) {
    
    ## d[[run]] <- list()
    ## d[[run]]$genes_cr <- load_standard_cellranger_genes(run_name = run)
    ## d[[run]]$reps_cr <- load_cellranger_repeats(run_name = run)
    ## d[[run]]$reps_fc <- load_featurecounts_repeats(run_name = run)
    d[[sapply(strsplit(dirname(rds), .Platform$file.sep),
               function(x) return(x[6]))]] <- readRDS(file = rds)
}
```

<!-- # PBMC markers -->

<!-- As in https://satijalab.org/seurat/v3.1/pbmc3k_tutorial.html -->

<!-- ```{r} -->
<!-- ids <- read.table(text='marker,type -->
<!-- IL7R,naive_t_cd4 -->
<!-- CCR7,naivet_t_cd4 -->
<!-- IL7R,memory_t_cd4 -->
<!-- S100A4,memory_t_cd4 -->
<!-- CD14,mono -->
<!-- LYZ,mono -->
<!-- MS4A1,b -->
<!-- CD8A,t_cd8 -->
<!-- FCGR3A,mono_fcgr3a -->
<!-- MS4A7,mono_fcgr3a -->
<!-- GNLY,nk -->
<!-- NKG7,nk -->
<!-- FCER1A,dc -->
<!-- CST3,dc -->
<!-- PPBP,platelet', sep = ',', header = TRUE, stringsAsFactors = FALSE) -->
<!-- ``` -->



<!-- ```{r, fig.height = 3, fig.width = 6, dpi = 50, results="asis", message = FALSE}  -->
<!-- for (run in names(d)) { -->
<!--     cat("## ", run, " {.tabset}\n\n") -->
<!--     ## print(run) -->
<!--     ## for (marker in ids$marker) {     -->
<!--     ## for (res in grep('RNA_snn', colnames(d[[run]]$genes@meta.data), value = TRUE)) { -->
<!--     res <- 'RNA_snn_res.0.2' -->
<!--     ## print(FeaturePlot(object = d[[run]]$genes, -->
<!--     ##                   features = ids$marker, -->
<!--     ##                   cols = c("green", "blue"),  -->
<!--     ##                   combine = TRUE, -->
<!--     ##                   pt.size = 1, -->
<!--     ##                   shape.by = res, -->
<!--     ##                   reduction = 'umap', -->
<!--     ##                   label = TRUE)) ## + ggtitle(paste(run, marker, res))) -->
<!--         ## } -->
<!--     ## } -->

<!--     for (type in unique(ids$type)) { -->
<!--         cat("### ", type, " {.tabset}\n\n") -->
<!--         ## print(paste(run, type, res)) -->
<!--         print(RidgePlot(object = d[[run]]$genes, -->
<!--                   features = ids[ids$type == type, 'marker'], -->
<!--                   sort = FALSE, -->
<!--                   assay = NULL, -->
<!--                   group.by = res, -->
<!--                   y.max = NULL, -->
<!--                   same.y.lims = FALSE , -->
<!--                   log = FALSE, -->
<!--                   ncol = NULL, -->
<!--                   slot = "data", -->
<!--                   combine = FALSE)) -->

<!--         cat("\n\n") -->
<!--     } -->
<!--     cat("\n\n") -->
<!-- } -->
<!-- ``` -->

# Most variable repeats cellranger {.tabset}

Similarly for repeats expression (avg +- sd of repeats CPM by gene expression cluster?)

What's the point here?


```{r, fig.height = 5, fig.width = 15, results = 'asis'}
res <- 'RNA_snn_res.0.2'

## selected <- c('AluY', 'AluSx', 'MER49', 'MIR', 'L2a', 'L2b', 'L1PA11')
selected <- c(names(head(sort(apply(GetAssayData(d[[run]]$repeats),1, coef_var), decreasing = TRUE),
                       5)), c('AluY', 'AluSx', 'L1PA11', 'MIR', 'L2a'))

for (run in names(d)) {
    cat("## ", run, "\n\n")
    repeats <- as.data.frame(FetchData(d[[run]]$repeats, vars = selected, slot = 'data'))
    repeats$cell <- rownames(repeats)

    d[[run]]$genes@meta.data$cell <- rownames(d[[run]]$genes@meta.data)
    repeats <- merge(d[[run]]$genes@meta.data[, c(res, 'cell') ],
                     repeats, by = 'cell', all.x = TRUE)
    
    rownames(repeats) <- repeats$cell
    
    repeats <- melt(repeats, id.vars = c('cell', res))
    
    colnames(repeats) <- c('cell', res, 'repfamily', 'CPM')

    print(ggplot(repeats,
           aes(x = get(res), y = CPM)) + 
          geom_boxplot(outlier.size=1) +
          scale_y_continuous(trans='log10') +
        scale_fill_discrete(drop=FALSE) +
        facet_wrap(~ factor(repfamily), ncol = 10) +
        theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
        ylab('repeat expression [CPM]') +
        xlab('gene-based clustering') + ggtitle(sprintf('%s %s\n%s', run, 'repeats', res)))
    cat('\n\n')
}

```


# Most variable repeats cellranger by identity {.tabset}


```{r, fig.height = 5, fig.width = 15, results = 'asis'}
scina <- 'scina'
## agg <- list()

## selected <- c('AluY', 'AluSx', 'MER49', 'MIR', 'L2a', 'L2b', 'L1PA11')
selected <- c(names(head(sort(apply(GetAssayData(d[[run]]$repeats),1, coef_var), decreasing = TRUE),
                       5)), c('AluY', 'AluSx', 'L1PA11', 'MIR', 'L2a'))

for (run in names(d)) {
    cat("## ", run, "\n\n")
    repeats <- as.data.frame(FetchData(d[[run]]$repeats, vars = selected, slot = 'data'))
    repeats$cell <- rownames(repeats)

    d[[run]]$genes@meta.data$cell <- rownames(d[[run]]$genes@meta.data)
    repeats <- merge(d[[run]]$genes@meta.data[, c(scina, 'cell') ],
                     repeats, by = 'cell', all.x = TRUE)
    
    rownames(repeats) <- repeats$cell
    
    repeats <- melt(repeats, id.vars = c('cell', scina))
    
    colnames(repeats) <- c('cell', scina, 'repfamily', 'CPM')

    print(ggplot(repeats,
           aes(x = scina, y = CPM)) + 
          geom_boxplot(outlier.size=1) +
          scale_y_continuous(trans='log10') +
        scale_fill_discrete(drop=FALSE) +
        facet_wrap(~ factor(repfamily), ncol = 10) +
        theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
        ylab('repeat expresssion [CPM]') +
        xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats', scina)))
    cat('\n\n')

    ## agg[[run]] <- data.frame(run = run,
    ##                          repeats)
}

```

# All integrated (repeats, scina)

```{r fig.width = 14, fig.height = 14}

## first aggregate all repeats stuff
sds <- lapply(d, function(x) return(apply(as.data.frame(GetAssayData(x$repeats,
                                                        slot = 'data')), 1, sd)))

## get the ones with the highest average sd

mean_sd <- apply(do.call(cbind.data.frame, sds), 1, mean)
selected <- names(head(sort(mean_sd, decreasing = TRUE), 20))
## selected


## aggregate(agg['CPM'], by = list(agg$repfamily, agg$run),  FUN = function(x) c(mean = mean(x, na.rm = TRUE), n = length(x), sd = sd(x, na.rm = TRUE) ))

## now get these data for selected features, aggregate and plot

tmp <- list()
for (run in names(d)) {
    foo <- data.frame(run = run,
                             FetchData(d[[run]]$repeats, vars = selected, slot = 'data'))

    ## add annotations
    bar <- data.frame(cell = rownames(foo),
                      scina = d[[run]]$repeats@meta.data[rownames(foo), 'scina'])

    stopifnot(rownames(foo) == bar$cell)
    
    
    tmp[[run]] <- data.frame(foo, scina = bar$scina)
    rm(foo, bar)
}

tmp <- do.call(rbind.data.frame, tmp)
str(tmp)

## melt
repeats <- melt(tmp, id.vars = c('run', 'scina'))
head(repeats)    
colnames(repeats) <- c('run', 'scina', 'repfamily', 'CPM')

```

```{r, fig.width = 12, fig.height = 12}
print(ggplot(repeats,
             aes(x = scina, y = CPM, fill = run)) + 
      geom_boxplot(outlier.size=1) +
      scale_y_continuous(trans='log10') +
      scale_fill_discrete(drop=FALSE) +
      facet_wrap(~ factor(repfamily), ncol = 4) +
      theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
      ylab('repeat expresssion [CPM]') +
      xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats', scina)))
```


```{r, fig.width = 8, fig.height = 14}
print(ggplot(repeats,
             aes(x = repfamily, y = CPM, fill = run)) + 
      geom_boxplot(outlier.size=1) +
      scale_y_continuous(trans='log10') +
      scale_fill_discrete(drop=FALSE) +
      facet_grid(factor(scina)~ factor(run)) +
      theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
      ylab('repeat expresssion [CPM]') +
      xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats', scina)))




```

# Cluster by average expression (repeats, SCINA)

Avoiding non-determined cell types

```{r, results = 'asis', fig.height = 10, fig.width = 10}
repeats <- repeats[repeats$scina != 'unknown',]
foo <- reshape2::dcast(repeats, run + scina ~ repfamily, mean, value.var = 'CPM')


rownames(foo) <- paste(foo$run, foo$scina, sep = '_')

annot <- foo[,c('run', 'scina')]
rownames(annot) <- paste(annot$run, annot$scina, sep = '_')
foo <- foo[,-(1:2)]

pheatmap(foo, annotation_row = annot, show_rownames = FALSE)

pheatmap(t(foo), annotation_col = annot, show_colnames = FALSE)


```


# Most variable repeats featurecounts {.tabset}

Same for featurecounts-based repeats

```{r, fig.height = 5, fig.width = 15, results = 'asis'}
## selected <- c('AluY', 'AluSx', 'MER49', 'MIR', 'L2a', 'L2b', 'L1PA11')

for (run in names(d)) {
    cat("## ", run, "\n\n")
    featurecounts <- as.data.frame(FetchData(d[[run]]$featurecounts, vars = selected, slot = 'data'))
    featurecounts$cell <- rownames(featurecounts)

    d[[run]]$genes@meta.data$cell <- rownames(d[[run]]$genes@meta.data)
    featurecounts <- merge(d[[run]]$genes@meta.data[, c(res, 'cell') ],
                     featurecounts, by = 'cell', all.x = TRUE)
    
    rownames(featurecounts) <- featurecounts$cell
    
    featurecounts <- melt(featurecounts, id.vars = c('cell', res))
    
    colnames(featurecounts) <- c('cell', res, 'repfamily', 'CPM')

    print(ggplot(featurecounts,
           aes(x = get(res), y = CPM)) + 
          geom_boxplot(outlier.size=1) +
          scale_y_continuous(trans='log10') +
        scale_fill_discrete(drop=FALSE) +
        facet_wrap(~ factor(repfamily), ncol = 10) +
        theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
        ylab('repeat expression [CPM]') +
        xlab('gene-based clustering') + ggtitle(sprintf('%s %s\n%s', run, 'repeats', res)))

    cat('\n\n')
}

```

# Most variable repeats featurecounts by identity (scina) {.tabset}

Same for featurecounts-based repeats

```{r, fig.height = 5, fig.width = 15, results = 'asis'}
## selected <- c('AluY', 'AluSx', 'MER49', 'MIR', 'L2a', 'L2b', 'L1PA11')

for (run in names(d)) {
    cat("## ", run, "\n\n")
    featurecounts <- as.data.frame(FetchData(d[[run]]$featurecounts, vars = selected, slot = 'data'))
    featurecounts$cell <- rownames(featurecounts)

    d[[run]]$genes@meta.data$cell <- rownames(d[[run]]$genes@meta.data)
    featurecounts <- merge(d[[run]]$genes@meta.data[, c(scina, 'cell') ],
                     featurecounts, by = 'cell', all.x = TRUE)
    
    rownames(featurecounts) <- featurecounts$cell
    
    featurecounts <- melt(featurecounts, id.vars = c('cell', scina))
    
    colnames(featurecounts) <- c('cell', scina, 'repfamily', 'CPM')

    print(ggplot(featurecounts,
           aes(x = scina, y = CPM)) + 
          geom_boxplot(outlier.size=1) +
          scale_y_continuous(trans='log10') +
        scale_fill_discrete(drop=FALSE) +
        facet_wrap(~ factor(repfamily), ncol = 10) +
        theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
        ylab('repeat expression [CPM]') +
        xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'repeats', scina)))

    cat('\n\n')
}

```


# All integrated (featurecounts, scina)

```{r fig.width = 14, fig.height = 14}

## first aggregate all featurecounts stuff
sds <- lapply(d, function(x) return(apply(as.data.frame(GetAssayData(x$featurecounts,
                                                        slot = 'data')), 1, sd)))

## get the ones with the highest average sd

mean_sd <- apply(do.call(cbind.data.frame, sds), 1, mean)
selected <- names(head(sort(mean_sd, decreasing = TRUE), 20))
## selected
selected

## aggregate(agg['CPM'], by = list(agg$repfamily, agg$run),  FUN = function(x) c(mean = mean(x, na.rm = TRUE), n = length(x), sd = sd(x, na.rm = TRUE) ))

## now get these data for selected features, aggregate and plot

tmp <- list()
for (run in names(d)) {
    foo <- data.frame(run = run,
                             FetchData(d[[run]]$featurecounts, vars = selected, slot = 'data'))

    ## add annotations
    bar <- data.frame(cell = rownames(foo),
                      scina = d[[run]]$featurecounts@meta.data[rownames(foo), 'scina'])

    stopifnot(rownames(foo) == bar$cell)
    
    
    tmp[[run]] <- data.frame(foo, scina = bar$scina)
    rm(foo, bar)
}

tmp <- do.call(rbind.data.frame, tmp)
str(tmp)

## melt
featurecounts <- melt(tmp, id.vars = c('run', 'scina'))
head(featurecounts)    
colnames(featurecounts) <- c('run', 'scina', 'repfamily', 'CPM')

```

```{r, fig.width = 12, fig.height = 12}
print(ggplot(featurecounts,
             aes(x = scina, y = CPM, fill = run)) + 
      geom_boxplot(outlier.size=1) +
      scale_y_continuous(trans='log10') +
      scale_fill_discrete(drop=FALSE) +
      facet_wrap(~ factor(repfamily), ncol = 4) +
      theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
      ylab('repeat expresssion [CPM]') +
      xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'featurecounts', scina)))
```


```{r, fig.width = 8, fig.height = 14}
print(ggplot(featurecounts,
             aes(x = repfamily, y = CPM, fill = run)) + 
      geom_boxplot(outlier.size=1) +
      scale_y_continuous(trans='log10') +
      scale_fill_discrete(drop=FALSE) +
      facet_grid(factor(scina)~ factor(run)) +
      theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
      ylab('repeat expresssion [CPM]') +
      xlab('scina identity') + ggtitle(sprintf('%s %s\n%s', run, 'featurecounts', scina)))




```


# Cluster by average expression (featurecounts, SCINA)

Avoiding non-determined cell types

```{r, results = 'asis', fig.height = 10, fig.width = 10}
featurecounts <- featurecounts[featurecounts$scina != 'unknown',]
foo <- reshape2::dcast(featurecounts, run + scina ~ repfamily, mean, value.var = 'CPM')


rownames(foo) <- paste(foo$run, foo$scina, sep = '_')

annot <- foo[,c('run', 'scina')]
rownames(annot) <- paste(annot$run, annot$scina, sep = '_')
foo <- foo[,-(1:2)]

pheatmap(foo, annotation_row = annot, show_rownames = FALSE)

pheatmap(t(foo), annotation_col = annot, show_colnames = FALSE)


```



# Pheatmaps

Note the cell sampling: same cells in all plots; `annotation_col` comes from genes' clustering.

```{r, results = 'asis', dpi=50, fig.width = 15, fig.height = 15}
NCELLS <- 1000
NFEATURES <- 500

## stop()
for (run in names(d)) {
    cat("## ", run, "{.tabset}\n\n")
    
    set.seed(654)
    idx_cell <- sample(colnames(GetAssayData(d[[run]]$genes)), NCELLS, replace = FALSE)
    
    for (item in names(d[[run]])) {
        cat("### ", item, "{.tabset}\n\n")

        set.seed(464)
        idx_feat <- sample(VariableFeatures(d[[run]][[item]]), NFEATURES, replace = FALSE)
        
        print(pheatmap(GetAssayData(
            object = subset(d[[run]][[item]],
                            features = idx_feat,
                            cells = intersect(idx_cell, colnames(GetAssayData(d[[run]][[item]])))),
                            slot = 'scale.data'),
            color = viridis(100),
            cluster_rows = TRUE,
            cluster_cols = TRUE,
            show_rownames = FALSE,
            show_colnames = FALSE,
            clustering_distance_cols = "euclidean",
            clustering_method = "ward.D2",
            clustering_distance_rows = "euclidean",
            fontsize_row = 12,
            annotation_col = as.data.frame(d[[run]]$genes@meta.data[idx_cell, c(res),
                                                                    drop = FALSE]),
            scale = "none"))
        cat('\n\n')
    }
    cat('\n\n')
}
```

Maybe add the markers of some celltypes on top?


# ARIS

## Data load

```{r}

aris_df <- NULL

for (rds in params) {
    ## the params are RDS objects from Seurat
    ## aris[[sapply(strsplit(dirname(rds), .Platform$file.sep),
    ##       function(x) return(x[6]))]] <- list()
    fns <- list.files(dirname(rds), pattern = 'aris.*rds')
    for (fn in fns) {
        curr_nest <- readRDS(file.path(dirname(rds), fn))
        ## curr <- do.call(function(x) rbind.data.frame(x, setNames = c('cat1' , 'cat2', 'ari')),
        ##                 curr_nest)
        ## aris_df <- NULL
        for (item in names(curr_nest)) {
            curr <- curr_nest[[item]]
            colnames(curr) <- c('set1', 'set2', 'ari')
            curr$origin <- item

            parsed <- strcapture(
                pattern = "aris_regress_nCount_(.*)_nFeature_(.*).rds",
                x = fn,
                proto = list(ncount = logical(), nfeature = logical()))
            
            curr$ncount_scaling <- parsed$ncount
            curr$nfeature_scaling <- parsed$nfeature
            
            curr$sample <- sapply(strsplit(dirname(rds), .Platform$file.sep),
                                  function(x) return(x[6]))
            aris_df <- rbind(aris_df, curr)

        }

        ## aris[[sapply(strsplit(dirname(rds), .Platform$file.sep),
        ##              function(x) return(x[6]))]][[basename(fn)]] <- curr_nest
    }
}

summary(aris_df)
rm(parsed, curr)
```

Extract resolutions and n

```{r}
parsed <- strcapture(
    pattern = "(.*)_RNA_snn_res.(.*) \\(n=(.*)\\)",
    x = aris_df$set1,
    proto = list(id1 = character(),
                 res1 = numeric(),
                 size1 = numeric()))

aris_df <- cbind(aris_df, parsed)
rm(parsed)

parsed <- strcapture(
    pattern = "(.*)_RNA_snn_res.(.*) \\(n=(.*)\\)",
    x = aris_df$set2,
    proto = list(id2 = character(),
                 res2 = numeric(),
                 size2 = numeric()))

aris_df <- cbind(aris_df, parsed)
rm(parsed)

## head(aris_df)
```

## ARIs comparison {.tabset}

Plot ARIs decay with resolution, scaling etc

```{r, fig.width=12, fig.height=8, results = 'asis'}

aris_df$regressed_out <- 'none'

aris_df$regressed_out[aris_df$ncount_scaling & aris_df$nfeature_scaling] <- 'both'
aris_df$regressed_out[aris_df$ncount_scaling & !aris_df$nfeature_scaling] <- 'genes_ncount'
aris_df$regressed_out[!aris_df$ncount_scaling & aris_df$nfeature_scaling] <- 'genes_nfeature'

aris_df$regressed_out <- factor(aris_df$regressed_out,
                                   levels = c('none', 'genes_ncount', 'genes_nfeature', 'both'))

```

```{r, fig.width=12, fig.height=8, results = 'asis'}

aris_df$label1 <- sprintf('res%s_(n=%s)', aris_df$res1, aris_df$size1)
aris_df$label2 <- sprintf('res%s_(n=%s)', aris_df$res2, aris_df$size2)

cat('### genes vs repeats cr\n\n')

print(ggplot(aris_df[aris_df$id1 == 'genes' & aris_df$id2 == 'repeats_cr',],
             aes(x = res1, y = ari, col = as.factor(regressed_out))) +
      geom_point() +
      labs(title = "genes cellranger vs repeats cellranger (from repeats bam)",
           subtitle = "by repeats resolution", y = "ARI", x = "genes resolution",
           color = 'regressed out') +
      facet_grid(sample~paste0('repeats res.\n', res2)) +
      theme(axis.text.x = element_text(angle = 90, hjust = 1)))

cat('\n\n')

cat('### genes vs repeats fc\n\n')

print(ggplot(aris_df[aris_df$id1 == 'genes' & aris_df$id2 == 'repeats_fc',],
             aes(x = res1, y = ari, col = as.factor(regressed_out))) +
      geom_point() +
      labs(title = "genes cellranger vs repeats featurecounts (from genes bam)",
           subtitle = "by repeats resolution", y = "ARI", x = "genes resolution",
           color = 'regressed out') +
      facet_grid(sample~paste0('repeats res.\n', res2)) +
      theme(axis.text.x = element_text(angle = 90, hjust = 1)))

cat('\n\n')

cat('### repeats fc vs repeats cr\n\n')

print(ggplot(aris_df[aris_df$id1 == 'repeats_fc' & aris_df$id2 == 'repeats_cr',],
             aes(x = res1, y = ari, col = as.factor(regressed_out))) +
      geom_point() +
      labs(title = "repeats featurecounts (from genes bam) vs repeats cellranger (repeats bam)",
           subtitle = "by repeats cellranger resolution", y = "ARI", x = "repeats fc resolution",
           color = 'regressed out') +
      facet_grid(sample~paste0('repeats cr res.\n', res2)) +
      theme(axis.text.x = element_text(angle = 90, hjust = 1)))

cat('\n\n')
```


Maybe add the number of clusters? as a size param?


Rather, plot both size and res

```{r, eval = TRUE}
## print(ggplot(aris_df[aris_df$id1 == 'repeats_fc' & aris_df$id2 == 'repeats_cr',],
##              aes(x = res1, y = ari, col = as.factor(regressed_out))) +
##       geom_point() +
##       labs(title = "repeats featurecounts (from genes bam) vs repeats cellranger (repeats bam)",
##            subtitle = "by repeats cellranger resolution", y = "ARI", x = "repeats fc resolution",
##            color = 'regressed out') +
##       facet_grid(sample~paste0('repeats cr res.\n', res2)) +
##       theme(axis.text.x = element_text(angle = 90, hjust = 1)))

print(ggplot(aris_df,
             aes(x = res1, y = size1, col = as.factor(regressed_out))) +
      geom_point() +
      facet_grid(sample~id1) +
      theme(axis.text.x = element_text(angle = 90, hjust = 1)))

print(ggplot(aris_df,
             aes(x = res2, y = size2, col = as.factor(regressed_out))) +
      geom_point() +
      facet_grid(sample~id2) +
      theme(axis.text.x = element_text(angle = 90, hjust = 1)))
```


# Timestamp

```{r sessionInfo, cache = FALSE}
date()
sessionInfo()
devtools::session_info()

```

