---
title: "Within sample report: PBMC, smartseq2"
author: "Izaskun Mallona at Robinsonlab"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    code_download: true
    number_sections: true
    df_print: kable
    theme: lumen
params:
  identifier: ""   
  repeats_bowtie: ""
  repeats_star: ""
  genes_star: ""
  seurat_output: ""
  aris_output: ""
  regress_genes_nCount: ""
  regress_genes_nFeature: ""
---


# Config

## Parameters

Rendered for sample:

```{r}
print(params)
```

## Libraries

```{r}
suppressPackageStartupMessages({
library(Matrix)

library(knitr)
library(Seurat)
library(dplyr)
## library(scater)
library(dplyr)
library(data.table)
## library(SingleCellExperiment)
## library(scran)
## library(org.Mm.eg.db)
## suppressPackageStartupMessages(library(dplyr))
## suppressPackageStartupMessages(library(DT))
## library(pheatmap)
## library(Cairo)
## library(RColorBrewer)
## library(corrplot)
library(ggplot2)
library(reshape2)
## library(xtable)
library(mclust)
library(pheatmap)
library(SCINA)
library(DT)
library(biomaRt)
})
```

```{r}
opts_chunk$set(fig.width = 5,
               fig.height = 5,
               dev = 'png',
               cache = FALSE,
               include = TRUE,
               cache.lazy = FALSE,
               warning = TRUE,
               echo = TRUE,
               message = FALSE)


options(bitmapType='cairo')
getOption('bitmapType')
```


## Functions

```{r functions}
## add alpha
ac <- function(col, alpha=1){
  apply(sapply(col, col2rgb)/255, 2, 
                     function(x) 
                       rgb(x[1], x[2], x[3], alpha=alpha))  
}

coef_var <- function(x){
        (sd(x)/mean(x))*100
}

```

# Data load

```{r justfordebugging, eval = FALSE}
## params <- list(genes_cellranger = "/home/imallona/repeats_sc/runs/frozen_pbmc_donor_a/cellranger_standard/outs/filtered_feature_bc_matrix/barcodes.tsv.gz",
##                repeats_cellranger= "/home/imallona/repeats_sc/runs/frozen_pbmc_donor_a/cellranger_repeats/outs/filtered_feature_bc_matrix/barcodes.tsv.gz",
##                repeats_featurecounts= "/home/imallona/repeats_sc/runs/frozen_pbmc_donor_a/count_repeats_on_cellranger_standard/frozen_pbmc_donor_a_repeats.counts.gz")

## SRR10974769

## params <- list(genes_cellranger = "/home/imallona/repeats_sc/runs/SRR10974769/cellranger_standard/outs/filtered_feature_bc_matrix/barcodes.tsv.gz",
##                repeats_cellranger= "/home/imallona/repeats_sc/runs/SRR10974769/cellranger_repeats/outs/filtered_feature_bc_matrix/barcodes.tsv.gz",
##                repeats_featurecounts= "/home/imallona/repeats_sc/runs/SRR10974769/count_repeats_on_cellranger_standard/SRR10974769_repeats.counts.gz",
##                identifier = 'test')

params <- list(identifier = "",
               genes_star = '/home/imallona/repeats_sc/runs/pbmcs_smartseq2_ding/star_transcriptome/genes/pbmcs_smartseq2_ding_star_genes.counts.gz',
               repeats_star = '/home/imallona/repeats_sc/runs/pbmcs_smartseq2_ding/star_transcriptome/repeats/pbmcs_smartseq2_ding_star_transcriptome_repeats.counts.gz',
               repeats_bowtie = '',
               seurat_output = "/home/imallona/tmp/test.rds",
               regress_genes_nCount = FALSE,
               regress_genes_nFeatures = FALSE)

```

```{r}
d <- list()
```


```{r}

featurecounts_to_seurat <- function(fn, identifier) {
    ## Featurecounts-based repeats
    tmp <- data.table::fread(fn, sep = '\t', header = TRUE,
                             drop = c('Chr', 'Start', 'End', 'Strand', 'Length'))
    rownames(tmp) <- tmp$Geneid ## here

    colnames(tmp) <- gsub(".*:([A-Z]+)-1.bam","\\1", colnames(tmp))
    ## tmp <- tmp[,7:ncol(tmp)]

    tmp <- as.data.frame(tmp)
    rownames(tmp) <- tmp$Geneid
    tmp <- tmp[,-1]

    return(CreateSeuratObject(
        tmp,
        project = identifier,
        assay = "RNA",
        min.cells = 0,
        min.features = 0,
        names.field = 1,
        names.delim = "_",
        meta.data = NULL))

}
```

```{r}
d <- list()

d$genes <- featurecounts_to_seurat(
    params$genes_star,
    identifier = sprintf('%s_%s', params$identifier, 'genes_star'))
d$repeats_star <- featurecounts_to_seurat(
    params$repeats_star,
    identifier = sprintf('%s_%s', params$identifier, 'repeats_star'))
d$repeats_bowtie <- featurecounts_to_seurat(params$repeats_bowtie)

```

# Basic Seurat processing

## QC before filtering {.tabset}

```{r, results = 'asis', fig.width = 8, fig.height = 4}
for (item in names(d)) {
    cat("### ", item, "\n\n")
    ## stop(grep("^MT-", rownames(d[[item]]), value = TRUE))
    if (item == 'genes') {
        d[[item]][["percent.mt"]] <- PercentageFeatureSet(d[[item]], pattern = "^MT-")
        print(FeatureScatter(d[[item]], feature1 = "nCount_RNA", feature2 = "percent.mt"))
        print(VlnPlot(d[[item]], features = c("nFeature_RNA", "nCount_RNA", "percent.mt"),
                      ncol = 3))

    } else {
        print(VlnPlot(d[[item]], features = c("nFeature_RNA", "nCount_RNA"), ncol = 2))
    
    }
    print(FeatureScatter(d[[item]], feature1 = "nCount_RNA", feature2 = "nFeature_RNA"))
    cat('\n\n')
}
```


## Add genesymbols to genes

Assumes this is human!

```{r}
if (!file.exists('feature_info.RData')) {
    
    tmp <- list()

    tmp$ids = rownames(d$genes)
    length(tmp$ids)


    tmp$filters = "ensembl_gene_id_version"
    tmp$attributes = c(tmp$filters, "external_gene_name", "chromosome_name", 
                       "gene_biotype", "start_position", "end_position")
    tmp$biomart = "ENSEMBL_MART_ENSEMBL" 
    tmp$dataset = "hsapiens_gene_ensembl"
    tmp$host = "asia.ensembl.org"
    tmp$bmart <- biomaRt::useMart(biomart = tmp$biomart, dataset = tmp$dataset, 
                                  host = tmp$host)

    feature_info <- biomaRt::getBM(attributes = tmp$attributes, filters = tmp$filters, 
                                   values = tmp$ids, mart = tmp$bmart)

    dim(feature_info)
    save(feature_info, tmp, file = 'feature_info.RData')
} else {
    load('feature_info.RData')
}

```

```{r}

tmp <- merge(data.frame(ensembl_gene_id_version = rownames(d$genes)),
             by = 'ensembl_gene_id_version', feature_info, all.x = TRUE)

rownames(tmp) <- tmp$ensembl_gene_id_version
tmp <- tmp[rownames(d$genes),]
stopifnot (nrow(d$genes) == nrow(tmp))

stopifnot(rownames(d$genes) == tmp$ensembl_gene_id_version)

rownames(d$genes) <- sprintf('%s_%s', tmp$ensembl_gene_id_version, tmp$external_gene_name)

stop('not feasible in a seurat object! move it before seurat')
```

## QC after filtering {.tabset}

Filtering out cells according to the repeats expression, and forwarding the selection to other layers

```{r}
d$genes <- subset(d$genes, subset = nFeature_RNA > 200 & nFeature_RNA < 5000 & (percent.mt < 10 | is.na(percent.mt)))
selected_cells <- colnames(d$genes)

for (item in names(d)) {
    d[[item]] <- subset(d[[item]], cells = selected_cells)
}

```

```{r, results = 'asis', fig.width = 8, fig.height = 4}
for (item in names(d)) {
    cat("### ", item, "\n\n")
    ## stop(grep("^MT-", rownames(d[[item]]), value = TRUE))
    if (item == 'genes') {
        d[[item]][["percent.mt"]] <- PercentageFeatureSet(d[[item]], pattern = "^MT-")
        print(FeatureScatter(d[[item]], feature1 = "nCount_RNA", feature2 = "percent.mt"))
        print(VlnPlot(d[[item]], features = c("nFeature_RNA", "nCount_RNA", "percent.mt"),
                      ncol = 3))

    } else {
        print(VlnPlot(d[[item]], features = c("nFeature_RNA", "nCount_RNA"), ncol = 2))
    
    }
    print(FeatureScatter(d[[item]], feature1 = "nCount_RNA", feature2 = "nFeature_RNA"))
    cat('\n\n')
}
```




## Variable features {.tabset}

```{r, message = FALSE, fig.width =8, fig.height = 4, results= 'asis'}
for (item in names(d)) {
    cat("### ", item, "\n\n")

    d[[item]] <- NormalizeData(object = d[[item]],
                               normalization.method = "LogNormalize",
                               scale.factor = 10000)
    
    d[[item]] <- FindVariableFeatures(d[[item]], selection.method = "vst", nfeatures = 2000)

    ## Identify the 10 most highly variable genes
    top10 <- head(VariableFeatures(d[[item]]), 10)

    ## plot variable features with and without labels
    plot1 <- VariableFeaturePlot(d[[item]])
    plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
    print(plot2 + ggtitle(item))
    cat('\n\n')
}



```

## Dimred PCA {.tabset}

```{r, message = FALSE, results ='asis'}
NUM_PCS <- 1:20
for (item in names(d)) {
    cat("### ", item, "\n\n")

    to_regress <- NULL
    if (as.logical(params$regress_genes_nCount)) {
        to_regress <- c(to_regress, 'nCount_RNA')
    }
    if (as.logical(params$regress_genes_nFeature)) {
        to_regress <- c(to_regress, 'nFeature_RNA')
        
    }
    
    if (item == 'genes') {
        d[[item]] <- ScaleData(d[[item]], vars.to.regress = to_regress)
    } else {
        ## in this case, the variables to regress come from the genes!
        ## e.g. is the genes library size the one to normalize the repets'
        tmp <- d$genes@meta.data
        tmp$cell <- rownames(tmp)
        d[[item]]@meta.data$cell <- rownames(d[[item]]@meta.data)
        d[[item]]@meta.data <- merge(d[[item]]@meta.data, tmp, by = 'cell',
                                     suffixes = c('', '_genes'))
        rownames(d[[item]]@meta.data) <- d[[item]]@meta.data$cell


        d[[item]] <- ScaleData(d[[item]],
                               vars.to.regress = if(is.null(to_regress)) NULL else paste0(to_regress, '_genes'))
        
        rm(tmp) 
    }
    
    d[[item]] <- RunPCA(d[[item]], features = VariableFeatures(object = d[[item]]))
    
    DimPlot(d[[item]], reduction = "pca")
    
    print(ElbowPlot(d[[item]]) + ggtitle(item))


    cat('\n\n')
}


```

```{r, message = FALSE, results = 'asis'}
for (item in names(d)) {

    d[[item]] <- FindNeighbors(d[[item]], dims = NUM_PCS)
    for (res in seq(0.2, 2, by = 0.2)) {
        d[[item]] <- FindClusters(d[[item]], resolution = res, verbose = FALSE)

    }

    d[[item]] <- RunUMAP(d[[item]], dims = NUM_PCS)
        
    d[[item]] <- RunTSNE(d[[item]], dims = NUM_PCS, check_duplicates = FALSE )
}


```


## Dimred UMAP and tSNE {.tabset}

```{r fig.width = 8, fig.height = 5, results = 'asis', cache = FALSE}

for (item in names(d)) {
    
    cat("### ", item, "\n\n")
    
    p1 <- DimPlot(d[[item]], reduction = "umap")
    p2 <- DimPlot(d[[item]], reduction = "tsne")

    print(CombinePlots(plots = list(p1, p2)) + ggtitle(item))
    rm(p1, p2)
    cat('\n\n')

}
```

# SCINA-based PBMC annotations

Markers from supplementary table s10 from [here](https://www.biorxiv.org/content/10.1101/632216v1.supplementary-material?versioned=true) [supp data](https://static-content.springer.com/esm/art%3A10.1038%2Fs41587-020-0465-8/MediaObjects/41587_2020_465_MOESM3_ESM.xlsx)


```{r}
markers2 <- read.table(text = 'cell_type	gene	sign
CD4+_T_cell	CD3D	+
CD4+_T_cell	CD3E	+
CD4+_T_cell	CD3G	+
CD4+_T_cell	TRAC	+
CD4+_T_cell	CD4	+
CD4+_T_cell	TCF7	+
CD4+_T_cell	CD27	+
CD4+_T_cell	IL7R	+
CD4+_T_cell	CD8A	-
CD4+_T_cell	CD8B	-
CD4+_T_cell	GNLY	-
CD4+_T_cell	NKG7	-
CD4+_T_cell	CST7	-
Cytotoxic_T_cell	CD3D	+
Cytotoxic_T_cell	CD3E	+
Cytotoxic_T_cell	CD3G	+
Cytotoxic_T_cell	TRAC	+
Cytotoxic_T_cell	CD8A	+
Cytotoxic_T_cell	CD8B	+
Cytotoxic_T_cell	GZMK	+
Cytotoxic_T_cell	CCL5	+
Cytotoxic_T_cell	NKG7	+
Cytotoxic_T_cell	CD4	-
Cytotoxic_T_cell	FCER1G	-
B_cell	CD19	+
B_cell	MS4A1	+
B_cell	CD79A	+
B_cell	CD79B	+
B_cell	MZB1	+
B_cell	IGHD	+
B_cell	IGHM	+
Natural_killer_cell	NCAM1	+
Natural_killer_cell	NKG7	+
Natural_killer_cell	KLRB1	+
Natural_killer_cell	KLRD1	+
Natural_killer_cell	KLRF1	+
Natural_killer_cell	KLRC1	+
Natural_killer_cell	KLRC2	+
Natural_killer_cell	KLRC3	+
Natural_killer_cell	KLRC4	+
Natural_killer_cell	CD3D	-
Natural_killer_cell	CD3E	-
Natural_killer_cell	CD3G	-
Natural_killer_cell	CD14	-
Natural_killer_cell	FCGR3A	+
Natural_killer_cell	FCGR3B	+
Natural_killer_cell	ITGAL	+
Natural_killer_cell	ITGAM	+
Natural_killer_cell	FCER1G	+
Natural_killer_cell	TRAC	-
CD14+_monocyte	VCAN	+
CD14+_monocyte	FCN1	+
CD14+_monocyte	S100A8	+
CD14+_monocyte	S100A9	+
CD14+_monocyte	CD14	+
CD14+_monocyte	ITGAL	+
CD14+_monocyte	ITGAM	+
CD14+_monocyte	CSF3R	+
CD14+_monocyte	CSF1R	+
CD14+_monocyte	CX3CR1	+
CD14+_monocyte	FCGR3A	-
CD14+_monocyte	FCGR3B	-
CD14+_monocyte	TYROBP	+
CD14+_monocyte	LYZ	+
CD14+_monocyte	S100A12	+
CD14+_monocyte	CD3D	-
CD14+_monocyte	CD3E	-
CD14+_monocyte	CD3G	-
CD14+_monocyte	TRAC	-
CD14+_monocyte	NKG7	-
CD14+_monocyte	KLRB1	-
CD14+_monocyte	KLRD1	-
CD16+_monocyte	FCN1	+
CD16+_monocyte	FCGR3A	+
CD16+_monocyte	FCGR3B	+
CD16+_monocyte	ITGAL	+
CD16+_monocyte	ITGAM	+
CD16+_monocyte	CSF3R	+
CD16+_monocyte	CSF1R	+
CD16+_monocyte	CX3CR1	+
CD16+_monocyte	CDKN1C	+
CD16+_monocyte	MS4A7	+
CD16+_monocyte	S100A8	-
CD16+_monocyte	S100A9	-
CD16+_monocyte	S100A12	-
CD16+_monocyte	CD14	-
CD16+_monocyte	CD3D	-
CD16+_monocyte	CD3E	-
CD16+_monocyte	CD3G	-
CD16+_monocyte	TRAC	-
CD16+_monocyte	NKG7	-
CD16+_monocyte	KLRB1	-
CD16+_monocyte	KLRD1	-
Dendritic_cell	HLA-DPB1	+
Dendritic_cell	HLA-DPA1	+
Dendritic_cell	HLA-DQA1	+
Dendritic_cell	ITGAX	+
Dendritic_cell	CD3D	-
Dendritic_cell	CD3E	-
Dendritic_cell	CD3G	-
Dendritic_cell	NCAM1	-
Dendritic_cell	CD19	-
Dendritic_cell	CD14	-
Dendritic_cell	CD1C	+
Dendritic_cell	CD1E	+
Dendritic_cell	FCER1A	+
Dendritic_cell	CLEC10A	+
Dendritic_cell	FCGR2B	+
Dendritic_cell	MS4A1	-
Dendritic_cell	CD79A	-
Dendritic_cell	CD79B	-
Plasmacytoid_dendritic_cell	IL3RA	+
Plasmacytoid_dendritic_cell	GZMB	+
Plasmacytoid_dendritic_cell	JCHAIN	+
Plasmacytoid_dendritic_cell	IRF7	+
Plasmacytoid_dendritic_cell	TCF4	+
Plasmacytoid_dendritic_cell	LILRA4	+
Plasmacytoid_dendritic_cell	CLEC4C	+
Plasmacytoid_dendritic_cell	ITGAX	-
Plasmacytoid_dendritic_cell	CD3D	-
Plasmacytoid_dendritic_cell	CD3E	-
Plasmacytoid_dendritic_cell	CD3G	-
Plasmacytoid_dendritic_cell	NCAM1	-
Plasmacytoid_dendritic_cell	CD19	-
Plasmacytoid_dendritic_cell	CD14	-
Plasmacytoid_dendritic_cell	MS4A1	-
Plasmacytoid_dendritic_cell	CD79A	-
Plasmacytoid_dendritic_cell	CD79B	-
Plasma_cell	CD38	+
Plasma_cell	XBP1	+
Plasma_cell	CD27	+
Plasma_cell	SLAMF7	+
Plasma_cell	CD19	-
Plasma_cell	MS4A1	-
Plasma_cell	CD3D	-
Plasma_cell	CD3E	-
Plasma_cell	CD3G	-
Plasma_cell	IGHA1	+
Plasma_cell	IGHA2	+
Plasma_cell	IGHG1	+
Plasma_cell	IGHG2	+
Plasma_cell	IGHG3	+
Plasma_cell	IGHG4	+
Platelet	PF4	+
Platelet	PPBP	+
Platelet	GP5	+
Platelet	ITGA2B	+
Platelet	NRGN	+
Platelet	TUBB1	+
Platelet	SPARC	+
Platelet	RGS18	+
Platelet	MYL9	+
Platelet	GNG11	+', header = TRUE, sep = '\t', stringsAsFactors = FALSE)
```

```{r}
pos_markers2_tmp <- markers2[markers2$sign == '+',]
pos_markers2 <-  list()

for (ct in unique(pos_markers2_tmp$cell_type)) {
    pos_markers2[[ct]] <- pos_markers2_tmp[pos_markers2_tmp$cell_type ==ct, 'gene']
}

```

```{r}
named <-  SCINA(GetAssayData(d$genes, slot = 'data'),
                signatures = pos_markers2,
                max_iter = 200,
                convergence_n = 20, 
                convergence_rate = 0.999, sensitivity_cutoff = 0.9)

table(named$cell_labels)
## plotheat.SCINA(exp = GetAssayData(d$genes, slot = 'data'),
##                 signatures = pos_markers2, results = named)
```

## Confusion matrices with clustering results {.tabset}

First, add the SCINA outputs to the different SO objects

```{r, cache = FALSE}
labelling <- data.frame(cell = colnames(d$genes),
                  scina = named$cell_labels)

rownames(labelling) <- labelling$cell
d$genes@meta.data$cell <- rownames(d$genes@meta.data)
d$genes@meta.data <- merge(d$genes@meta.data, labelling, by = 'cell')
rownames(d$genes@meta.data) <- d$genes@meta.data$cell

## still, the annotation can also be transfered to repeats/featurecounts

for (item in c('repeats', 'featurecounts')) {
    rownames(d[[item]]@meta.data) <- d[[item]]@meta.data$cell

    tmp <- merge(data.frame(cell = d[[item]]@meta.data$cell),
                 labelling, all.x = TRUE)
    rownames(tmp) <- tmp$cell

    d[[item]] <- AddMetaData(object = d[[item]],
                             metadata = tmp['scina'],
                             col.name = 'scina')
    rownames(d[[item]]@meta.data) <- d[[item]]@meta.data$cell
}

rm(labelling, tmp)

```

```{r problemrendering, results = 'asis', df_print = 'kable', cache = FALSE}
for (res in grep('RNA_snn', colnames(d$genes@meta.data), value = TRUE)) {
    cat('### ', res, '\n\n')

    tmp <- as.data.frame.matrix(table(d$genes@meta.data[,res],
                                      d$genes@meta.data$scina))

    ## print(t(tmp))
    print(tmp)
    
    cat('\n\n\n')
    rm(tmp)
}

```

<!-- ## ARIs celltypes and SCINA outputs -->

<!-- ```{r, cache = FALSE} -->
<!-- for (res in grep('RNA_snn', colnames(d$genes@meta.data), value = TRUE)) { -->
<!--     print(res) -->
<!--     print(as.data.frame.matrix(table(d$genes@meta.data[,res], d$genes@meta.data$scina))) -->
<!-- } -->
<!-- ``` -->



# Repeats and genes expression profiles by SCINA annotation {.tabset}

tSNEs/UMAPs

```{r, fig.width = 7, fig.height = 4, results = 'asis', cache = FALSE}
for (item in names(d)) {
    cat("## ", item, "\n\n")
    ## d[[item]]@meta.data$scina <- as.factor(d[[item]]@meta.data$scina)
    Idents(d[[item]]) <- 'scina'
    print(DimPlot(d[[item]], reduction = "umap", group.by = 'ident'))
    print(DimPlot(d[[item]], reduction = "tsne", group.by = 'ident'))

    ## print(CombinePlots(plots = list(p1, p2)) + ggtitle(item))
    cat('\n\n')

}
```

# boxplots by identity {.tabset}

```{r fig.height = 6, fig.width = 16, results = 'asis'}
## selected <- c('AluY', 'AluSx', 'MER49', 'MIR', 'L2a', 'L2b', 'L1PA11')
for (slot in c('repeats_star', 'repeats_bowtie')) {
    cat("## ", slot, "\n\n")
    selected <- c(names(head(sort(apply(GetAssayData(d[[slot]]),1, coef_var),
                                  decreasing = TRUE),
                             5)),
                  c('AluY', 'AluSx', 'L1PA11', 'MIR', 'L2a'))


    repeats <- as.data.frame(FetchData(d[[slot]], vars = selected, slot = 'data'))
    repeats$cell <- rownames(repeats)

    
    d$genes@meta.data$cell <- rownames(d$genes@meta.data)
    repeats <- merge(d$genes@meta.data[, c('scina', 'cell') ],
                     repeats, by = 'cell', all.x = TRUE)

    rownames(repeats) <- repeats$cell
    
    repeats <- melt(repeats, id.vars = c('cell', 'scina'))

    colnames(repeats) <- c('cell', 'scina', 'repfamily', 'data')
    
    print(ggplot(repeats,
           aes(x = scina, y = data)) + 
          geom_boxplot(outlier.size=1) +
          scale_y_continuous(trans='log10') +
        scale_fill_discrete(drop=FALSE) +
        facet_wrap(~ factor(repfamily), ncol = 10) +
        theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
        ylab('repeat expression [ncount]') +
        xlab('SCINA identity') + ggtitle(sprintf('%s %s\n%s', slot, 'repeats', 'scina')))

    rm(repeats)
    cat('\n\n')
}
```

# Descriptive (clustering) {.tabset}

Tabset: coloring according to the clustering results.

```{r descriptive, results = 'asis', fig.width =8, fig.height = 8}

res <- 'RNA_snn_res.0.2'
common_cells <- intersect(intersect(colnames(d$genes), colnames(d$repeats_star), d$repeats_bowtie)


for (item_clust in names(d)) {
    tag <- sprintf('%s\n%s', item_clust, res)
    cat('## ', tag, '\n\n')
    
    merged <- data.frame(cell = common_cells,
                         percent.mt = d$genes@meta.data[common_cells, 'percent.mt'])

    merged[,tag] <- d[[item_clust]]@meta.data[common_cells,res]

    for (item in names(d)) {
        d[[item]]@meta.data$cell <- rownames(d[[item]]@meta.data)

        merged <- merge(merged,
                        d[[item]]@meta.data[c('cell', 'nCount_RNA', 'nFeature_RNA')],
                        by = 'cell',
                        all.x = TRUE,
                        suffixes = c('', sprintf('\n%s', item)))

        ## cat("### ", item, "\n\n")

    }
    rownames(merged) <- merged$cell
    merged <- merged[,-1]
    colnames(merged) <- gsub('_RNA', '', colnames(merged))
    plot(merged, pch = 20, cex = 0.5, col = ac(as.numeric(merged[,tag]), 0.1))
    cat('\n\n')
}

rm(res)
rm(merged)
```

# Descriptive (with scina)

Tabset: coloring according to the clustering results.

```{r descriptive2, results = 'asis', fig.width =8, fig.height = 8}

## res <- 'RNA_snn_res.0.2'
common_cells <- intersect(intersect(colnames(d$genes), colnames(d$repeats)),
                                    colnames(d$featurecounts))

tag <- 'scina'

merged <- data.frame(cell = common_cells,
                     percent.mt = d$genes@meta.data[common_cells, 'percent.mt'])

## merged[,tag] <- d[[item_clust]]@meta.data[common_cells,res]
merged[,tag] <- d[[item]]@meta.data[common_cells, tag]

for (item in names(d)) {
        ## cat('## ', tag, '\n\n')
        d[[item]]@meta.data$cell <- rownames(d[[item]]@meta.data)

        merged <- merge(merged,
                        d[[item]]@meta.data[c('cell', 'nCount_RNA', 'nFeature_RNA')],
                        by = 'cell',
                        all.x = TRUE,
                        suffixes = c('', sprintf('\n%s', item)))

        ## cat("### ", item, "\n\n")
}

rownames(merged) <- merged$cell
merged <- merged[,-1]
colnames(merged) <- gsub('_RNA', '', colnames(merged))
plot(merged, pch = 20, cex = 0.5, col = ac(as.numeric(as.factor(merged[,tag])), 0.1))
## plot(merged, pch = 20, cex = 0.5, col = ac(as.numeric(as.factor(merged[,tag])), 1))
## cat('\n\n')

rm(merged)
```

# Correlation values distribution

Repeats by cellranger vs repeats by featurecounts on cellranger-standard bams.

Correlations per cell

```{r correl_first, warning = FALSE}
ccell <- list()
     
for (cell in intersect( colnames(GetAssay(d$repeats, 'RNA')),
                       colnames(GetAssay(d$featurecounts, 'RNA')))) {
    reps <- intersect(rownames(d$repeats), rownames(d$featurecounts))
    
    ccell[[cell]] <- cor(as.numeric(GetAssay(d$repeats, 'RNA')[reps,cell]),
                         as.numeric(GetAssay(d$featurecounts, 'RNA')[reps,cell]),
                         method = 'spearman')
}
```

```{r}
hist(unlist(ccell),
     main = "correlation across cells\nrepeats cellranger vs repeats (featurecounts)")
## lines(density(unlist(ccell)))

```

Correlations per repeat

```{r, warning = FALSE}
crepeat <- list()

## GetAssay(d$repeats, 'RNA')[1:10,1:10]

for (re in intersect( rownames(GetAssay(d$repeats, 'RNA')),
                     rownames(GetAssay(d$featurecounts, 'RNA')))) {

    cells <-  intersect(colnames(d$repeats), colnames(d$featurecounts))
    crepeat[[re]] <- cor(as.numeric(GetAssay(d$repeats, 'RNA')[re,cells]),
                         as.numeric(GetAssay(d$featurecounts, 'RNA')[re, cells]),
                         method = 'spearman')
}
```

```{r}
hist(unlist(crepeat), main = "correlation across repeats\nrepeats cellranger vs repeats (featurecounts)")
## lines(density(unlist(ccell)))

```

# Confusion matrix clusters

At a given resolution, compare both clusterings.

Color scale depicts Rand Index.

## Genes vs repeats star

```{r, results = 'asis', fig.width = 4, fig.height = 4, cache = FALSE}
aris <- list()
i <- 1
for (genes_res in seq(0.2, 2, 0.2)) {
    genes_res <- sprintf('RNA_snn_res.%s', genes_res)
    
    for (repeats_res in seq(0.2, 2, 0.2)) {
        repeats_res <- sprintf('RNA_snn_res.%s', repeats_res)

        cells <- intersect(rownames(d$genes@meta.data), rownames(d$repeats_star@meta.data))
        
   

        aris[[i]] <- c(sprintf('genes_%s (n=%s)', genes_res,
                               length(unique(d$genes@meta.data[cells, genes_res]))),
                       sprintf('repeats_star_%s (n=%s)', repeats_res,
                               length(unique(d$repeats_star@meta.data[cells, repeats_res]))),
                       adjustedRandIndex(d$genes@meta.data[cells,genes_res],
                                         d$repeats_star@meta.data[cells,repeats_res]))
        i <- i +1
    }    
}

aris <- do.call(rbind.data.frame, aris)
colnames(aris) <- c('genes_res', 'repeats_cr_res', 'ari')
aris$ari <- as.numeric(as.character(aris$ari))
```


```{r, fig.width = 6, fig.height = 6}
aris_wide <- reshape2::dcast(aris, genes_res~repeats_cr_res, value.var="ari")

rownames(aris_wide)<- aris_wide$genes_res
aris_wide <- aris_wide[,-1]

pheatmap(aris_wide, cluster_rows = FALSE, cluster_cols = FALSE,
         color = colorRampPalette(c("white","blue"))(20))
```

```{r store_aris}
aris_stored <- list()
stop('tillhere')
aris_stored[['genes__repeats_cr']] <- aris
```

## Genes vs repeats (bowtie)


```{r, results = 'asis', fig.width = 4, fig.height = 4, cache = FALSE}
aris <- list()
i <- 1
for (genes_res in seq(0.2, 2, 0.2)) {
    genes_res <- sprintf('RNA_snn_res.%s', genes_res)
    
    for (repeats_res in seq(0.2, 2, 0.2)) {
        repeats_res <- sprintf('RNA_snn_res.%s', repeats_res)

        cells <- intersect(rownames(d$genes@meta.data), rownames(d$repeats_bowtie@meta.data))
        
   

        aris[[i]] <- c(sprintf('genes_%s (n=%s)', genes_res,
                               length(unique(d$genes@meta.data[cells, genes_res]))),
                       sprintf('repeats_bowtie_%s (n=%s)', repeats_res,
                               length(unique(d$repeats_bowtie@meta.data[cells, repeats_res]))),
                       adjustedRandIndex(d$genes@meta.data[cells,genes_res],
                                         d$repeats_bowtie@meta.data[cells,repeats_res]))
        i <- i +1
    }    
}

aris <- do.call(rbind.data.frame, aris)
colnames(aris) <- c('genes_res', 'repeats_fc_res', 'ari')
aris$ari <- as.numeric(as.character(aris$ari))
```

```{r, fig.width = 6, fig.height = 6}
aris_wide <- reshape2::dcast(aris, genes_res~repeats_fc_res, value.var="ari")

rownames(aris_wide)<- aris_wide$genes_res
aris_wide <- aris_wide[,-1]

pheatmap(aris_wide, cluster_rows = FALSE, cluster_cols = FALSE,
         color = colorRampPalette(c("white","blue"))(20))
```

```{r}
aris_stored[['genes__repeats_fc']] <- aris
```


## Repeats vs repeats

```{r, results = 'asis', fig.width = 4, fig.height = 4, cache = FALSE}
aris <- list()
i <- 1
for (repeats_star_res in seq(0.2, 2, 0.2)) {
    repeats_star_res <- sprintf('RNA_snn_res.%s', repeats_star_res)
    
    for (repeats_res in seq(0.2, 2, 0.2)) {
        repeats_res <- sprintf('RNA_snn_res.%s', repeats_res)

        cells <- intersect(rownames(d$repeats_star@meta.data), rownames(d$repeats@meta.data))
        
   

        aris[[i]] <- c(sprintf('repeats_fc_%s (n=%s)', repeats_star_res,
                               length(unique(d$repeats_star@meta.data[cells, repeats_star_res]))),
                       sprintf('repeats_cr_%s (n=%s)', repeats_res,
                               length(unique(d$repeats@meta.data[cells, repeats_res]))),
                       adjustedRandIndex(d$repeats_star@meta.data[cells,repeats_star_res],
                                         d$repeats@meta.data[cells,repeats_res]))
        i <- i +1
    }    
}

aris <- do.call(rbind.data.frame, aris)
colnames(aris) <- c('repeats_star_res', 'repeats_res', 'ari')
aris$ari <- as.numeric(as.character(aris$ari))
```


```{r, fig.width = 6, fig.height = 6}
aris_wide <- reshape2::dcast(aris, repeats_star_res~repeats_res, value.var="ari")

rownames(aris_wide)<- aris_wide$repeats_star_res
aris_wide <- aris_wide[,-1]

pheatmap(aris_wide, cluster_rows = FALSE, cluster_cols = FALSE,
         color = colorRampPalette(c("white","blue"))(20))
```

```{r}
aris_stored[['repeats_cr__repeats_fc']] <- aris

```


## Genes vs SCINA

Removed the unassigned class in SCINA

```{r}
aris <- list()
i <- 1
for (genes_res in seq(0.2, 2, 0.2)) {
    genes_res <- sprintf('RNA_snn_res.%s', genes_res)
    rownames(d$genes@meta.data) <- d$genes@meta.data$cell
    known <- rownames(d$genes@meta.data[d$genes@meta.data$scina != 'unknown',])

    aris[[i]] <- c(sprintf('genes_%s (n=%s)', genes_res,
                           length(unique(d$genes@meta.data[known, genes_res]))),
                   sprintf('scina_%s (n=%s)', 'scina',
                           length(unique(d$genes@meta.data[known, 'scina']))),
                   adjustedRandIndex(d$genes@meta.data[known, genes_res],
                                     d$genes@meta.data[known, 'scina']))
                                     
    i <- i +1
}    

aris <- do.call(rbind.data.frame, aris)
colnames(aris) <- c('genes_res', 'scina', 'ari')
aris$ari <- as.numeric(as.character(aris$ari))
print(aris)
```

# Export objects

List of seurat objects
```{r}
saveRDS(object = d, file = params$seurat_output)
saveRDS(object = aris_stored, file = params$aris_output)
```

# Timestamp

```{r sessionInfo, cache = FALSE}
date()
sessionInfo()
devtools::session_info()

```

