---
title: "Simulations chromium"
author: "Izaskun Mallona at Robinsonlab"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    code_download: true
    number_sections: true
    df_print: kable
    theme: lumen
    always_allow_html: yes
---

```
params:
    repeats_featurecounts_unique: ""
    repeats_only_unique: ""
    repeats_featurecounts_multi: ""
    repeats_only_multi: ""
    repeats_alevin_locus: ""
    repeats_alevin_no_decoy_locus: ""
    repeats_bowtie_unique: ""
    repeats_bowtie_multi: ""
    scte: ""
    shao_counts: ""
    shao_saf: ""
    shao_sam: ""
    truth: ""
    reps_gtf: ""
    nthreads: ""
```

Memory intensive, caution (see future memory options)

```{r, include = FALSE, eval = TRUE}
suppressPackageStartupMessages({
    library(Matrix)
    library(knitr)
    library(pheatmap)
    library(ggplot2)
    library(tximeta)
    library(SummarizedExperiment)
    library(forcats)    
    library(dplyr)
    library(reshape2)
    library(future.apply)
    library(data.table)
    library(GenomicRanges)
})

## if (!requireNamespace("BiocManager", quietly = TRUE))
##     install.packages("BiocManager")

## install.packages(c('waldo',
##                    'rematch2',
##                    'testthat',
##                    'ggplot2'), dependencies = TRUE)
                 
## BiocManager::install(c("knitr","tximeta",
##                        "pheatmap",
##                        lib = "/home/imallona/R/bioc-312"), dependencies = TRUE)

```

```{r, include = FALSE, eval = TRUE}
BASE <- file.path('/home', 'imallona', 'repeats_sc')
```

```{r, include = FALSE, eval = TRUE}
opts_chunk$set(fig.width = 5,
               fig.height = 5,
               dev = 'png',
               cache = TRUE,
               echo = TRUE,
               include = TRUE,
               cache.lazy = FALSE,
               warning = TRUE,
               error = TRUE,
               message = FALSE)


options(bitmapType='cairo')
## getOption('bitmapType')

## data.table

```


```{r justfordebugging, include = FALSE, eval = TRUE}
params <- list(repeats_alevin_locus = "/home/imallona/repeats_sc/runs/sim_5k_pbmc_v3/alevin/repeats_per_locus/alevin/quants_mat.gz",
               repeats_alevin_no_decoy_locus = "/home/imallona/repeats_sc/runs/sim_5k_pbmc_v3/alevin/repeats_no_decoy_per_locus/alevin/quants_mat.gz",
               truth = "/home/imallona/repeats_sc/data/sim_5k_pbmc_v3/truth.tsv.gz",
               repeats_featurecounts_unique = "/home/imallona/repeats_sc/runs/sim_5k_pbmc_v3/count_repeats_loci_on_cellranger_standard/unique_reads/sim_5k_pbmc_v3_repeats.counts.gz",
               repeats_only_unique = '~/repeats_sc/runs/sim_5k_pbmc_v3/count_repeats_loci_on_cellranger_standard_not_overlapping_genes/unique_reads/sim_5k_pbmc_v3_repeats_not_overlapping_genes.counts.gz',
               repeats_featurecounts_multi = "/home/imallona/repeats_sc/runs/sim_5k_pbmc_v3/count_repeats_loci_on_cellranger_standard/multimappers/sim_5k_pbmc_v3_repeats.counts.gz",
               repeats_only_multi = '~/repeats_sc/runs/sim_5k_pbmc_v3/count_repeats_loci_on_cellranger_standard_not_overlapping_genes/multimappers/sim_5k_pbmc_v3_repeats_not_overlapping_genes.counts.gz',
               repeats_bowtie_multi = "/home/imallona/repeats_sc/runs/sim_5k_pbmc_v3/bowtie_repeatome/multimappers/by_locus_repeats.counts.gz",
               repeats_bowtie_unique = "/home/imallona/repeats_sc/runs/sim_5k_pbmc_v3/bowtie_repeatome/unique_reads/by_locus_repeats.counts.gz",
               scte = "/home/imallona/repeats_sc/runs/sim_5k_pbmc_v3/scTE/scTE.out.csv.gz",
               shao_counts = '~/repeats_sc/runs/sim_5k_pbmc_v3/zumis/sim_5k_zumis_count_dt.txt',
               shao_saf = '~/repeats_sc/annotation/shao/shao_GRCh38_te.saf',
               shao_sam = '~/repeats_sc/runs/sim_5k_pbmc_v3/zumis/sim_5k_zumis_featurecount.sam.txt',
               reps_gtf = '~/repeats_sc/annotation/GRCh38_rmsk_TE.gtf.gz',
               nthreads = 40)
```

```{r}
NTHREADS <- params$nthreads
setDTthreads(threads = params$nthreads)
```

```{r}
d <- list()
truth <- read.table(params$truth, header = TRUE)
truth <- cbind(truth, t(sapply(strsplit(truth$locus, ';'), function(x) return(x))))
colnames(truth)[5:9] <- c('name', 'id', 'family', 'class', 'locus')
```

```{r}
d$repeats_alevin <- tximeta::tximeta(
     coldata = data.frame(
         names = "genes",
         files = params$repeats_alevin_locus,
         stringsAsFactors = FALSE
     ), type = "alevin")

d$repeats_alevin <- assays(d$repeats_alevin)$counts
```

```{r}
d$repeats_alevin_no_decoy <- tximeta::tximeta(
     coldata = data.frame(
         names = "genes",
         files = params$repeats_alevin_no_decoy_locus,
         stringsAsFactors = FALSE
     ), type = "alevin")

d$repeats_alevin_no_decoy <- assays(d$repeats_alevin_no_decoy)$counts
```


```{r}
read_featurecounts <- function(fn){
    ## without multimappers
    tmp <- data.table::fread(fn,
                             sep = '\t', header = TRUE,
                             drop = c('Chr', 'Start', 'End', 'Strand', 'Length'))
    rownames(tmp) <- tmp$Geneid ## here

    colnames(tmp) <- gsub(".*:([A-Z]+)-1.*.bam","\\1", colnames(tmp))
    ## tmp <- tmp[,7:ncol(tmp)]

    tmp <- as.data.frame(tmp)
    rownames(tmp) <- tmp$Geneid
    tmp <- tmp[,-1]

    return(tmp)
}
```

```{r}
featurecounts_runs <- c('repeats_featurecounts_unique',
                        'repeats_only_unique',
                        'repeats_featurecounts_multi',
                        'repeats_only_multi',
                        ## 'repeats_bowtie_multi',
                        'repeats_bowtie_unique')
```


```{r}
for (item in featurecounts_runs) {
    d[[item]] <- read_featurecounts(params[[item]])
    ## colnames(d[[item]]) <-  gsub(".*:([A-Z]+)-1.*.bam","\\1", colnames(d[[item]]))
}
```


Read scTE

```{r scte}
tmp <- data.table::fread(params$scte,
                         sep = ',',
                         header = TRUE)

repnames <- setdiff(colnames(tmp), 'barcodes')
barcodes <- tmp$barcodes

tmp <- as.data.frame(data.table::transpose(tmp[,-'barcodes']))

colnames(tmp) <- barcodes

rownames(tmp) <- repnames
print(tmp[1:3,1:3])

## filter out genes here maybe?

d$scte <- tmp
rm(tmp)

```


# read shao

```{r}
d$shao <- list()
d$shao$counts <- fread(params$shao_counts)
d$shao$saf <- fread(params$shao_saf)
d$shao$sam <- fread(params$shao_sam)
## d$shao <- merge(d$shao$counts, d$shao$saf, by.x= 'feature', by.y = 'GeneID', all = TRUE)
```

```{r}

truth$chr <- gsub(".*::(.+):([0-9]+)-([0-9]+)", "\\1", truth$locus)
truth$start <- gsub("(.+):([0-9]+)-([0-9]+)", "\\2", truth$locus)
truth$end <- gsub("(.+):([0-9]+)-([0-9]+)", "\\3", truth$locus)

```

```{r, eval = FALSE}
saveRDS(file = 'd.rds', object = d)
```

# Locus level 


```{r, eval = FALSE}
## heavy downsampling

d <- lapply(d, function(x) return(head(x, 1e4)[,1:50]))

## scte does not provide outputs at the locus level
sapply(d, dim)
sapply(d, class)
d$scte[1:3,1:3]

sapply(d, function(x) head(colnames(x)))

```

Proportion of correctly mapped, incorrectly mapped, unaligned

```{r}
props <- list()
```

For alevin

```{r}

res <- list()
for (cell in unique(truth$cell)) {

    if (cell %in% colnames(d$repeats_alevin)) {
        true <- truth[truth$cell == cell, 'id']

        tmp <- d$repeats_alevin[,cell]
        observed <- tmp[tmp > 0]

        ## among these, some are observed 'fractionally'
        correct <- observed[observed == 1]

        ## under- or overestimation
        multi <- observed[observed != 1]

        missing <- setdiff(true, names(observed))
        wrong <- setdiff(names(observed), true)

        res[[cell]] <- setNames(c(cell,
                                  length(true), length(correct), length(multi),
                                  length(missing),
                                  length(wrong)),
                                c('true', 'correct', 'multi', 'missing', 'wrong'))
    }
    else {
        res[[cell]] <- setNames( c(cell, length(true), 0, 0, length(true), 0),
                                c('cell', 'true', 'correct', 'multi', 'missing', 'wrong'))
    }
}

res <- do.call(rbind.data.frame, res)
colnames(res) <-  c('cell', 'true', 'correct', 'multi', 'missing', 'wrong')
props[['repeats_alevin']] <- res
props[['repeats_alevin']]$flavour <- 'repeats_alevin'

```

For alevin without decoy

```{r}

res <- list()
for (cell in unique(truth$cell)) {

    if (cell %in% colnames(d$repeats_alevin_no_decoy)) {
        true <- truth[truth$cell == cell, 'id']

        tmp <- d$repeats_alevin_no_decoy[,cell]
        observed <- tmp[tmp > 0]

        ## among these, some are observed 'fractionally'
        correct <- observed[observed == 1]

        ## under- or overestimation
        multi <- observed[observed != 1]

        missing <- setdiff(true, names(observed))
        wrong <- setdiff(names(observed), true)

        res[[cell]] <- setNames(c(cell,
                                  length(true), length(correct), length(multi),
                                  length(missing),
                                  length(wrong)),
                                c('true', 'correct', 'multi', 'missing', 'wrong'))
    }
    else {
        res[[cell]] <- setNames( c(cell, length(true), 0, 0, length(true), 0),
                                c('cell', 'true', 'correct', 'multi', 'missing', 'wrong'))
    }
}

res <- do.call(rbind.data.frame, res)
colnames(res) <-  c('cell', 'true', 'correct', 'multi', 'missing', 'wrong')
props[['repeats_alevin_no_decoy']] <- res
props[['repeats_alevin_no_decoy']]$flavour <- 'repeats_alevin_no_decoy'
```


<!-- For feature counts -->

<!-- ```{r} -->
<!-- for (fc in c('repeats_featurecounts_unique', -->
<!--                'repeats_only_unique', -->
<!--                'repeats_featurecounts_multi', -->
<!--                'repeats_only_multi')) { -->
<!--     res <- list() -->
<!--     for (cell in unique(truth$cell)) { -->

<!--         if (cell %in% colnames(d[[fc]])) { -->
<!--             true <- truth[truth$cell == cell, 'id'] -->

<!--             observed <- rownames(d[[fc]])[d[[fc]][,cell] > 0] -->

<!--             ## among these, some are observed 'fractionally' -->
<!--             correct <- observed[d[[fc]][observed,cell] == 1] -->

<!--             ## under- or overestimation -->
<!--             multi <- observed[d[[fc]][observed,cell] != 1] -->

<!--             missing <- setdiff(true, observed) -->
<!--             wrong <- setdiff(observed, true) -->

<!--             res[[cell]] <- setNames(c(cell, -->
<!--                                       length(true), length(correct), length(multi), -->
<!--                                       length(missing), -->
<!--                                       length(wrong)), -->
<!--                                     c('cell', 'true', 'correct', 'multi', 'missing', 'wrong')) -->
<!--         } -->
<!--         else { -->
<!--             res[[cell]] <- setNames( c(cell, length(true), 0, 0, length(true), 0), -->
<!--                                     c('cell', 'true', 'correct', 'multi', 'missing', 'wrong')) -->
<!--         } -->
<!--     } -->

<!--     res <- do.call(rbind.data.frame, res) -->
<!--     colnames(res) <-  c('cell', 'true', 'correct', 'multi', 'missing', 'wrong') -->
<!--     props[[fc]] <- res -->
<!--     props[[fc]]$flavour <- fc -->
<!-- } -->
<!-- ``` -->


<!-- make this faster -->

<!-- ```{r} -->

<!-- y <- list() -->
<!-- for (ii in seq_along(X)) { -->
<!--   y[[ii]] <- local({ -->
<!--     x <- X[[ii]] -->
<!--     tmp <- sqrt(x) -->
<!--     tmp            ## same as return(tmp) -->
<!--   }) -->
<!-- } -->

<!-- library(future.apply) -->
<!-- plan(multiprocess) ## => parallelize on your local computer -->

<!-- X <- 1:5 -->

<!-- y <- future_lapply(X, function(x) { -->
<!--   tmp <- sqrt(x) -->
<!--   tmp -->
<!-- }) -->

<!-- ``` -->


```{r}
## 130 GB
options(future.globals.maxSize = 130 * 1024^3)

plan(multiprocess, workers = NTHREADS) 

## no scte nor assembly because one uses repName aggregated data
##   and the other does not give a meaningful identifier
## fc  <- c(featurecounts_runs, 'scte')
fc  <- featurecounts_runs

props2 <- future_lapply(fc, function(x) {
    res <- list()
    for (cell in unique(truth$cell)) {

        if (cell %in% colnames(d[[x]])) {
            true <- truth[truth$cell == cell, 'id']

            observed <- rownames(d[[x]])[d[[x]][,cell] > 0]

            ## among these, some are observed 'fractionally'
            correct <- observed[d[[x]][observed,cell] == 1]

            ## under- or overestimation
            multi <- observed[d[[x]][observed,cell] != 1]

            missing <- setdiff(true, observed)
            wrong <- setdiff(observed, true)

            res[[cell]] <- setNames(c(cell,
                                      length(true), length(correct), length(multi),
                                      length(missing),
                                      length(wrong)),
                                    c('cell', 'true', 'correct', 'multi', 'missing', 'wrong'))
        }
        else {
            res[[cell]] <- setNames( c(cell, length(true), 0, 0, length(true), 0),
                                    c('cell', 'true', 'correct', 'multi', 'missing', 'wrong'))
        }
    }

    res <- do.call(rbind.data.frame, res)
    colnames(res) <-  c('cell', 'true', 'correct', 'multi', 'missing', 'wrong')
    res$flavour <- x
    return(res)
})

```

```{r}
sapply(d, function(x) head(colnames(x)))
sapply(props2, function(x) head(colnames(x)))
```


```{r}
saveRDS(file = 'props.rds', object = c(props, props2))

```


```{r}
fd <- do.call(rbind.data.frame, c(props, props2))
tail(fd)
```


```{r}
## stolen from https://stackoverflow.com/questions/53241804/error-bars-incorrectly-positioned-in-a-stacked-bar-graph-r

fd <- reshape2::melt(fd,  id.vars = c('cell', 'flavour'))
fd$value <- as.numeric(fd$value)
fd <- fd[fd$variable != 'true',]
## discarded multimapping
fd$variable <- as.character(fd$variable)
fd$variable[fd$variable == 'multi'] <- 'multimapping'

fd$variable <- factor(fd$variable, levels = rev(c('correct', 'multimapping', 'missing', 'wrong')))

## group_by(fd, flavour, variable) %>%
##     summarize(mean = mean(value),
##               sd = sd(value),
##               lower = mean(value) - sd(value),
##               upper = mean(value) + sd(value)) %>%
##     arrange(flavour) %>%
##     group_by(flavour) %>%

##     ungroup() %>%
##     ggplot(aes(x = flavour, y = mean, fill = variable),
##            xLabels = NA) +
##     geom_bar(stat="identity") +
##     geom_errorbar(aes(ymin = lower, ymax = upper),
##                                          width = .2, col = "red") +
##     theme_bw() + scale_x_discrete(limits = unique(fd$flavour)) +
##     xlab("") +
##     ylab ("") +
##     scale_fill_grey() +
##     theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

```{r}
fd2 <- as.data.frame(group_by(fd, flavour, variable) %>%
              summarize(mean = mean(value),
                        sd = sd(value),
                        lower = mean(value) - sd(value),
                        upper = mean(value) + sd(value)) %>%
              arrange(flavour) %>%
              group_by(flavour)) 

fd2$variable <- factor(fd2$variable, levels = c('correct', 'multimapping', 'wrong', 'missing'))

# fd2 <- fd2[order(fd2$flavour, fd2$variable),]

## ggplot(fd2, aes(x = flavour, y = mean, fill = variable),
##            xLabels = NA) +
##     geom_bar(stat="identity") +
##     geom_errorbar(aes(ymin = lower, ymax = upper),
##                                          width = .2, col = "red") +
##     theme_bw() + scale_x_discrete(limits = unique(fd$flavour)) +
##     xlab("") +
##     ylab ("") +
##     scale_fill_grey() +
##     theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

```

```{r}
ggplot(fd2, aes(x=flavour, weight=mean, ymin=lower, ymax=upper, fill=variable)) +
    geom_bar      (position=position_dodge(), aes(y=mean), stat="identity") +
    geom_errorbar (position=position_dodge(width=0.9), colour="black", width = 0.2) +
    theme_bw() +
    scale_x_discrete(limits = unique(fd2$flavour)) +
    scale_fill_grey(name = 'mapping\ncategory') +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    xlab('strategy') +
    ylab('alignments per cell (counts)')
```

```{r}
ggplot(fd2, aes(x=variable, weight=mean, ymin=lower, ymax=upper, fill=flavour)) +
    geom_bar      (position=position_dodge(), aes(y=mean), stat="identity") +
    geom_errorbar (position=position_dodge(width=0.9), colour="black", width = 0.2) +
    theme_bw() +
    scale_x_discrete(limits = unique(fd2$variable)) +
    scale_fill_grey(name = 'mapping\ncategory') +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    xlab('strategy') +
    ylab('alignments per cell (counts)')

```

Maybe the same with percentages of the simulated reads?


```{r}
fd <- do.call(rbind.data.frame, c(props, props2))

for (item in c('correct', 'multi', 'missing', 'wrong'))
    fd[,item] <- as.numeric(fd[,item])

pt <- prop.table(as.matrix(fd[,c('correct', 'multi', 'missing', 'wrong')]),1 )

fd <- data.frame(cell = fd$cell,
                 flavour = fd$flavour,
                 pt)

fd <- reshape2::melt(fd,  id.vars = c('cell', 'flavour'))
fd$value <- as.numeric(fd$value)

## discarded multimapping
fd$variable <- as.character(fd$variable)
fd$variable[fd$variable == 'multi'] <- 'multimapping'

fd$variable <- factor(fd$variable, levels = rev(c('correct', 'multimapping', 'missing', 'wrong')))



fd2 <- as.data.frame(group_by(fd, flavour, variable) %>%
              summarize(mean = mean(value),
                        sd = sd(value),
                        lower = mean(value) - sd(value),
                        upper = mean(value) + sd(value)) %>%
              arrange(flavour) %>%
              group_by(flavour)) 

fd2$variable <- factor(fd2$variable, levels = c('correct', 'multimapping', 'wrong', 'missing'))

```


```{r}
ggplot(fd2, aes(x=flavour, weight=mean, ymin=lower, ymax=upper, fill=variable)) +
    geom_bar      (position=position_dodge(), aes(y=mean), stat="identity") +
    geom_errorbar (position=position_dodge(width=0.9), colour="black", width = 0.2) +
    theme_bw() +
    scale_x_discrete(limits = unique(fd$flavour)) +
    scale_fill_grey(name = 'mapping\ncategory') +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    xlab('strategy') +
    ylab('alignments per cell (proportion)')
```


```{r}
ggplot(fd2, aes(x=variable, weight=mean, ymin=lower, ymax=upper, fill=flavour)) +
    geom_bar      (position=position_dodge(), aes(y=mean), stat="identity") +
    geom_errorbar (position=position_dodge(width=0.9), colour="black", width = 0.2) +
    theme_bw() +
    scale_x_discrete(limits = unique(fd$variable)) +
    scale_fill_grey(name = 'mapping\ncategory') +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    xlab('strategy') +
    ylab('alignments per cell (proportion)')

```


From incorrectly mapped, get those that at least mach the repName, repFamily, or repClass

# Not averages, but rather total amounts

```{r}
fd <- do.call(rbind.data.frame, c(props$alevin, props2))

grouped <- aggregate(list(correct = as.numeric(fd$correct),
                          truth = as.numeric(fd$true),
                          wrong = as.numeric(fd$wrong),
                          multi = as.numeric(fd$multi),
                          missing = as.numeric(fd$missing)),
                     by = list(fd$flavour), sum)

grouped <- reshape2::melt(grouped, id.vars = 'Group.1') 
grouped$variable <- factor(grouped$variable, levels = c('truth', 'correct',
                                                        'multi', 'missing', 'wrong'))

levels(grouped$variable)
grouped %>% ggplot(aes(x=Group.1, y=value, fill =variable)) +
    geom_bar      (position=position_dodge(), stat="identity") +
    theme_bw() +
    scale_x_discrete(limits = unique(fd$flavour)) +
    scale_fill_grey(name = 'mapping\ncategory') +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    xlab('strategy') +
    ylab('alignments (number of)')


```

# Grouping by repname etc, to be able to include scTE

Proportion of families?

```{r}

aggs <- list()

for (item in c('name', 'family', 'class')) {

    ## truth contains a dictionary
    ## we weight half counts for multimappers (truth doesn't have them, but other tools might)
    weighted <- as.data.frame(table(truth[,item], truth$count))
    head(weighted)
    weighted$weight <- as.numeric(as.character(weighted$Var2)) * weighted$Freq
    
    weighted <- aggregate(weighted$weight, by = list(weighted$Var1), sum)
    colnames(weighted) <- c(item, 'truth')
    aggs[[item]] <- list(truth = weighted)

    for (flavour in setdiff(names(d), 'shao')) {
        if (flavour == 'scte') {
            d[[flavour]][] <- lapply(d[[flavour]], function(x) {
                if(is.factor(x)) as.numeric(as.character(x))
                else if (is.character(x)) as.numeric(x)
                else x
            })
            
            summed <- rowSums(d[[flavour]])
            summed <- data.frame(name = names(summed),
                                 sum = summed)

            annot <- truth[,c('name', item)]
            annot <- annot[!duplicated(annot),]

            summed <- merge(summed,
                            annot,
                            by.x = 'name',
                            by.y = 'name')

            weighted <- aggregate(summed$sum, by = list(summed[,item]), sum)
            colnames(weighted) <- c(item, flavour)
            aggs[[item]][[flavour]] <- weighted       
        }
        else {
            summed <- rowSums(d[[flavour]])
            summed <- data.frame(id = names(summed),
                                 sum = summed)
            summed <- merge(summed, truth[,c('id', item)],
                            by.x = 'id',
                            by.y = 'id')
            
            weighted <- aggregate(summed$sum, by = list(summed[,item]), sum)
            colnames(weighted) <- c(item, flavour)
            aggs[[item]][[flavour]] <- weighted            
        }
    }
}
```

Add Shao

<!-- ```{r} -->

<!-- d$shao$saf <- as.data.frame(d$shao$saf) -->

<!-- overlaps <- mergeByOverlaps(query = GRanges(seqnames = d$shao$saf$Chr, -->
<!--                                      ranges = IRanges(d$shao$saf$Start, d$shao$saf$End), -->
<!--                                      feature = d$shao$saf$GeneID), -->
<!--                      subject = GRanges(seqnames = paste0('chr', truth$chr), -->
<!--                                         ranges = IRanges(as.numeric(truth$start), -->
<!--                                                          as.numeric(truth$end)), -->
<!--                                        locus = truth$locus), -->
<!--                      minoverlap = 30) -->

<!-- ## still, these are not unique -->
<!-- overlap <- as.data.frame(overlaps[c('feature', 'locus')]) -->
<!-- dim(overlap) -->


<!-- ``` -->

<!-- Removing Shao's te identifiers with multiple assignments. Skipped, or the total mapping is very low. -->

<!-- ```{r, eval = FALSE} -->
<!-- overlap <- overlap[!duplicated(overlap$feature),] -->
<!-- dim(overlap) -->
<!-- ``` -->

<!-- Getting the aggregated counts for Shao's -->

<!-- ```{r} -->

<!-- d$shao <- merge(d$shao$counts, overlap, by.x = 'feature', by.y='feature', all.x = TRUE) -->

<!-- d$shao$name <- gsub("(.+);(.+);(.+);(.+)::.*", "\\1", d$shao$locus) -->
<!-- d$shao$family <- gsub("(.+);(.+);(.+);(.+)::.*", "\\3", d$shao$locus) -->
<!-- d$shao$class <- gsub("(.+);(.+);(.+);(.+);::.*", "\\4", d$shao$locus) -->

<!-- flavour <- 'shao_EM' -->
<!-- d$shao <- as.data.frame(d$shao) -->

<!-- for (item in c('name', 'family', 'class')) { -->
<!--     weighted <- aggregate(d$shao$EM_distri, by = list(d$shao[,item]), -->
<!--                           function(x) sum(na.omit(x))) -->
    
    
<!--     colnames(weighted) <- c(item, flavour) -->
<!--     aggs[[item]][[flavour]] <- weighted        -->
<!-- } -->

<!-- ``` -->


```{r}

does_overlap <- function(x) {
    return (x$Chr == paste0('chr', x$sim_chr) | abs(x$Start - x$sim_start) < 100)
}

d$shao$saf <- as.data.frame(d$shao$saf)
d$shao$counts <- as.data.frame(na.omit(d$shao$counts))
d$shao$sam <- as.data.frame(d$shao$sam)

d$shao$sam$sim_locus <-  gsub("(.+);(.+);(.+);(.+)::.*", "\\2", d$shao$sam$V1)
d$shao$sam$sim_chr <-  gsub("(.*)_([0-9XY]+)_([0-9]+)_([0-9]+)_([+-]+)", "\\2",
                            d$shao$sam$sim_locus)
d$shao$sam$sim_start <-  as.numeric(gsub("(.*)_([0-9XY]+)_([0-9]+)_([0-9]+)_([+-]+)", "\\3",
                              d$shao$sam$sim_locus))
d$shao$sam$sim_end <-  as.numeric(gsub("(.*)_([0-9XY]+)_([0-9]+)_([0-9]+)_([+-]+)", "\\4",
                            d$shao$sam$sim_locus))
d$shao$sam$sim_family <-  gsub("(.+);(.+);(.+);(.+)::.*", "\\3", d$shao$sam$V1)
d$shao$sam$sim_class <-  gsub("(.+);(.+);(.+);(.+);::.*", "\\4", d$shao$sam$V1)
d$shao$sam$sim_name <- gsub("(.*)_([0-9XY]+)_([0-9]+)_([0-9]+)_([+-]+)", "\\1",
                            d$shao$sam$sim_locus)


str(d$shao)
```

For those TE ids assigned to CB, retrieve the SAF coordinates
```{r}
d$shao$merged <- merge(d$shao$counts, d$shao$saf, by.x = 'feature', by.y = 'GeneID')
head(d$shao$merged)
```

```{r}
table(table(d$shao$saf$GeneID))
table(d$shao$saf$type)

## the alignments reported, some with multimappings
dim(d$shao$sam)

## the count table, with the counts and the SAF, depicting
dim(d$shao$merged)
head(d$shao$merged)
head(d$shao$sam)
head(truth)

```


<!-- Then, check if the coordinates match the ones from the mapping (they provide a SAM file) -->


<!-- ```{r} -->
<!-- d$shao$merged <- merge(d$shao$merged, d$shao$sam, by.x = 'feature', by.y = 'V4') -->
<!-- d$shao$merged$overlaps <- does_overlap(d$shao$merged) -->


<!-- ``` -->

<!-- ```{r} -->
<!-- tail(d$shao$sam) -->

<!-- ## all(d$shao$sam$family == d$shao$sam$class) -->
<!-- table(d$shao$sam$mapped_name == d$shao$sam$sim_name) -->

<!-- tmp <- merge(d$shao$counts, as.data.frame(d$shao$sam), by.x = 'feature', by.y = 'V4') -->

<!-- ``` -->

<!-- Removing Shao's te identifiers with multiple assignments. Skipped, or the total mapping is very low. -->

<!-- Getting the aggregated counts for Shao's -->

<!-- ```{r} -->

<!-- d$shao <- merge(d$shao$counts, overlap, by.x = 'feature', by.y='feature', all.x = TRUE) -->

<!-- d$shao$name <- gsub("(.+);(.+);(.+);(.+)::.*", "\\1", d$shao$locus) -->
<!-- d$shao$family <- gsub("(.+);(.+);(.+);(.+)::.*", "\\3", d$shao$locus) -->
<!-- d$shao$class <- gsub("(.+);(.+);(.+);(.+);::.*", "\\4", d$shao$locus) -->

<!-- flavour <- 'shao_EM' -->
<!-- d$shao <- as.data.frame(d$shao) -->

<!-- for (item in c('name', 'family', 'class')) { -->
<!--     weighted <- aggregate(d$shao$EM_distri, by = list(d$shao[,item]), -->
<!--                           function(x) sum(na.omit(x))) -->
    
    
<!--     colnames(weighted) <- c(item, flavour) -->
<!--     aggs[[item]][[flavour]] <- weighted        -->
<!-- } -->

<!-- ``` -->


```{r}
## head(d$shao$merged)
head(unique(truth$cell))
head(unique(d$shao$merged$RG))
intersect(unique(truth$cell), unique(d$shao$merged$RG))
for (RG in unique(truth$cell)) {
    ## subsetting the the current cell
    curr <- d$shao$merged[d$shao$merged$RG == RG,]
}
```


by Class

```{r}
tmp <- data.frame(class = aggs$class$truth$class)
for (flavour in aggs$class) {
    tmp <- merge(tmp, flavour, by = 'class', all.x = TRUE)
}


tmp <- melt(tmp,  id.vars = c('class'))
colnames(tmp) <- c('class', 'flavour', 'counts')

ggplot(tmp, aes(fill = class, y = counts, x = flavour)) + 
    geom_bar(position="stack", stat="identity")  +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```


```{r}
ggplot(tmp, aes(fill = flavour, y = counts, x = class)) + 
    geom_bar(position="dodge", stat="identity")  +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

```


```{r}
ggplot(tmp, aes(fill = class, y = counts, x = flavour)) + 
    geom_bar(position="dodge", stat="identity")  +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

```


by family, only the top 10 families in the truth are shown


```{r}
tmp <- data.frame(family = aggs$family$truth$family)
for (flavour in aggs$family) {
    tmp <- merge(tmp, flavour, by = 'family', all.x = TRUE)
}

selected <- head(tmp[order(tmp$truth, decreasing = TRUE),'family'], 10)

tmp <- reshape2::melt(tmp,  id.vars = c('family'))
colnames(tmp) <- c('family', 'flavour', 'counts')

tmp <- tmp[tmp$family %in% selected,]


ggplot(tmp, aes(fill = family, y = counts, x = flavour)) + 
    geom_bar(position="stack", stat="identity")  +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

```{r}
ggplot(tmp, aes(fill = flavour, y = counts, x = family)) + 
    geom_bar(position="dodge", stat="identity")  +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    scale_y_continuous(trans = 'log10')

```


```{r}
ggplot(tmp, aes(fill = family, y = counts, x = flavour)) + 
    geom_bar(position="dodge", stat="identity")  +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    scale_y_continuous(trans = 'log10')

```


by name, only the top 10 names are shown


```{r}
tmp <- data.frame(name = aggs$name$truth$name)
for (flavour in aggs$name) {
    tmp <- merge(tmp, flavour, by = 'name', all.x = TRUE)
}

selected <- head(tmp[order(tmp$truth, decreasing = TRUE),'name'], 10)

tmp <- melt(tmp,  id.vars = c('name'))
colnames(tmp) <- c('name', 'flavour', 'counts')

tmp <- tmp[tmp$name %in% selected,]


ggplot(tmp, aes(fill = name, y = counts, x = flavour)) + 
    geom_bar(position="stack", stat="identity")  +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

```{r}
ggplot(tmp, aes(fill = flavour, y = counts, x = name)) + 
    geom_bar(position="dodge", stat="identity")  +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    scale_y_continuous(trans = 'log10')

```

```{r}
ggplot(tmp, aes(fill = name, y = counts, x = flavour)) + 
    geom_bar(position="dodge", stat="identity")  +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    scale_y_continuous(trans = 'log10')

```

```{r}
knit::exit()
```

# Cells corr


Some pairwise correlations for randomly picked cells


```{r}
set.seed(64554)
sampled <- sample(x = truth$cell, size = 10, replace = FALSE)


for (cell in sampled) {
    tmp <- list()
    tmp$truth <- truth[truth$cell == cell, 'id']
    for (item in grep('repeats', names(d), value = TRUE)) {
        tmp[[item]] <- rownames(d[[item]])[d[[item]][,cell] > 0]

    }

    ## build a matrix with binary distances?
    
    
}

```

```{r, eval = FALSE}
print('knit me')
knit::exit()
```

# are the wrong mappings things from the same family?


```{r}



gtf <- fread(params$reps_gtf, colClasses = c(rep('NULL', 8), 'character'))

reg <- 'gene_id "(.+)"; transcript_id "(.+)"; family_id "(.+)"; class_id "(.+)";'

gtf$id <- gsub(pattern = reg, replacement = '\\2', x = gtf$V9)
gtf$name <- gsub(pattern = reg, replacement = '\\1', x = gtf$V9)
gtf$family <- gsub(pattern = reg, replacement = '\\3', x = gtf$V9)
gtf$class <- gsub(pattern = reg, replacement = '\\4', x = gtf$V9)

gtf <- as.data.frame(gtf[,c('id', 'name', 'family', 'class')])

fds <- list()
for (item in setdiff(names(params), c('scte', 'shao_saf', 'truth', 'shao_counts', 'nthreads'))) {
    for (fn in list.files(dirname(params[[item]]), pattern = "*featureCounts")){
        print(fn)
        foo <- as.data.frame(fread(fn, sep = '\t'))
        foo <- merge(foo, gtf, by.x= 'V4', by.y = 'id', all.x = TRUE)
        colnames(foo) <- c('mapped.id', 'read_name', 'status', 'count', 'mapped.name', 'mapped.family',
                           'mapped.class')

        foo$sim.id <- gsub(pattern = '(.+);(.+);(.+);(.+);.*', replacement = '\\2', x = foo$read_name)
        foo <- merge(foo, gtf, by.x= 'sim.id', by.y = 'id', all.x = TRUE)
        ## head(foo)

        ## spawn multioverlaps?
        ## strsplit(foo$V4, split = ",")
        ## data.frame(V1 = rep(df$V1, sapply(s, length)), V2 = unlist(s))
        
        colnames(foo)[9:11] <- c('sim.name', 'sim.family', 'sim.class') 

        
        fds[[fn]] <- list(fn = fn,
                          family.correct = sum(na.omit(foo$mapped.family == foo$sim.family)),
                          family.wrong = sum(na.omit(foo$mapped.family != foo$sim.family)),
                          name.correct = sum(na.omit(foo$mapped.name == foo$sim.name)),
                          name.wrong = sum(na.omit(foo$mapped.name != foo$sim.name)),
                          class.correct = sum(na.omit(foo$mapped.class == foo$sim.class)),
                          class.wrong = sum(na.omit(foo$mapped.class != foo$sim.class)),
                          Unassigned_MultiMapping = sum(foo$status == 'Unassigned_MultiMapping'),
                          Unassigned_NoFeatures = sum(foo$status == 'Unassigned_NoFeatures'),
                          Unassigned_Unmapped = sum(foo$status == 'Unassigned_Unmapped'),
                          Unassigned_Ambiguity = sum(foo$status == 'Unassigned_Ambiguity'),
                          Assigned = sum(foo$status == 'Assigned'))
                          
        
        

        data.frame(cbind(table(foo$mapped.family == foo$sim.family, useNA = 'always')))
                              
        table(foo$mapped.name == foo$sim.name, useNA = 'always')
        table(foo$mapped.class == foo$sim.class, useNA = 'always')
        table(foo$status)


    }
        
}

```

# Timestamp

```{r sessionInfo, cache = FALSE}
date()
sessionInfo()
devtools::session_info()

```

```{r}
knitr::knit_exit()
```


